
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Geonomics &#8212; geonomics 0.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="highlight-python"><div class="highlight"><pre><span></span>               <span class="c1">## :: ::    :::            ##</span>
         <span class="c1">##:::   :::::    :::   ::    :: :: :::##</span>
      <span class="c1">## ::::     ::           ::   ::::::::::::::##</span>
    <span class="c1">##::::::                       ::::::::: :::::: :##</span>
  <span class="c1">## :    :::                    :::    ::    :::::::::##</span>
 <span class="c1">##ggggg eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
<span class="c1">##g     e    o   o  nn  n  o   o  m   m   i   c     s    ##</span>
<span class="c1">##g     eee o     o n n n o     o mm mm   i   c     sssss##</span>
<span class="c1">##g ggg eee o     o n  nn o     o m m m   i   c         s##</span>
<span class="c1">##g   g e    o   o  n   n  o   o  m   m   i   c        ss##</span>
 <span class="c1">##gggg  eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
  <span class="c1">##  ::::::::        ::::::::::::  :       ::  ::   : ##</span>
    <span class="c1">##  ::::              :::::::  ::     ::::::::  :##</span>
      <span class="c1">## :::               :::::: ::       ::::::  ##</span>
         <span class="c1">##:                ::::                ##</span>
               <span class="c1">##                         ##</span>
</pre></div>
</div>
<img alt="_images/yosemite_header_img.png" class="align-center" src="_images/yosemite_header_img.png" />
<div class="section" id="geonomics">
<h1>Geonomics<a class="headerlink" href="#geonomics" title="Permalink to this headline">¶</a></h1>
<div class="section" id="a-python-package-for-easy-construction-of-individual-based-spatially-explicit-forward-time-and-highly-customizable-landscape-genomics-simulations">
<h2>A Python package for easy construction of individual-based, spatially explicit, forward-time, and highly customizable landscape-genomics simulations<a class="headerlink" href="#a-python-package-for-easy-construction-of-individual-based-spatially-explicit-forward-time-and-highly-customizable-landscape-genomics-simulations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="provides">
<h3>Provides<a class="headerlink" href="#provides" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ol class="arabic simple">
<li><code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>,
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> classes</li>
<li>A <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> class that builds objects of the aforementioned classes
according to the scenario stipulated in a parameters file,
then uses them to run numerous simulation iterations</li>
<li>Classes to help the user to save data and basic statistics
from those iterations at all desired timesteps</li>
</ol>
<ol class="arabic simple" start="3">
<li>Classes to help the user model arbitrarily complex landscape- and
demographic-change scenarios</li>
<li>Numerous visualization methods, to help the user design, run, explore,
present, and explain their models’ behavior and results</li>
</ol>
</div></blockquote>
</div>
<div class="section" id="intro">
<h3>Intro<a class="headerlink" href="#intro" title="Permalink to this headline">¶</a></h3>
<p>Backward-time (i.e. coalescent) simulators abound.
But they are inadequate for simulation of many scenarios of
interest, including: natural selection on traits with arbitrary genomic
architectures; spatially variable natural selection; simulation of species or
populations distributed continuously and moving realistically across
complex landscapes; complex demographic change simultaneous with ongoing,
often non-stationary environmental change; and coevolutionary interactions
between multiple species or incipient species. Few existing forward-time
simulators can model all of these phenomena, and the few that can often
impose a high cost of entry (e.g. learning a new, non-standard programming
language in order to write one’s desired simulations). Geonomics aims to fill
this empty niche by combining ease of use with broad extensibility.
If it succeeds at doing this, Geonomics should prove uniquely useful
for a wide range of purposes, from intro-level educational use to
high-quality theoretical, methodological, empirical, and
applied research.</p>
<p>Geonomics is written in Python, a full-fledged scripting language
that is relatively easy to learn (and fun!). So it can be pretty quick for a
new user to get up to speed and start doing useful work. For work with
Geonomics, this turnaround time should be even quicker. Geonomics aims to
require minimal Python knowledge (yet maintain high extensibility for
interested, expert users). Essentially, anyone should be able to build their
own, arbitrarily complex Geonomics models as long as they know how to install
the package, open a Python console, call Python functions, and edit some
default values in a pre-packaged script.</p>
<p>Geonomics will allow you to:</p>
<blockquote>
<div><ul class="simple">
<li>create a <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> with any number of <code class="code py python docutils literal"><span class="name"><span class="pre">Layers</span></span></code> in it;</li>
<li>create any number of <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> living on that
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, each of which is composed of a bunch of
independent <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s, and each of which will have a bunch of
parameters dsecribing what it’s like and how it lives;</li>
<li>optionally give the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s of any <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
genomes, which can optionally determine phenotypes for any number
of <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s (all of this is controlled by the
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> that you would create for
the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>);</li>
<li>simulate any number of timsesteps of the evolution of those
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> on that <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, where each timestep can include
movement, mating, mortality (by density-dependence and optionally also by
natural selection), and demographic, life-history, or
environmental changes</li>
</ul>
</div></blockquote>
<hr class="docutils" />
<div class="section" id="object-orientation">
<h4>Object-orientation<a class="headerlink" href="#object-orientation" title="Permalink to this headline">¶</a></h4>
<p>For a more technical understanding of the model, it may be helpful to
understand the concept of <strong>object-oriented programming</strong>.  Here is a very
brief tutorial for the unacquainted:</p>
<p>Python is a very handy language for object-oriented programming,
and this is the primary programming paradigm in which Geonomics is written.
Essentially, object-orientation involves:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Defining certain types of data structures, or <strong>classes</strong> (e.g.
<code class="code docutils literal"><span class="pre">Car</span></code>), and assigning them various behaviors, or <strong>methods</strong>
(e.g. <code class="code docutils literal"><span class="pre">honk</span></code>, <code class="code docutils literal"><span class="pre">turn</span></code>, <code class="code docutils literal"><span class="pre">brake</span></code>);</li>
<li>Using particular data values to create individual instances, or
<strong>objects</strong> belonging to those classes (e.g. <code class="code docutils literal"><span class="pre">my_1986_jeep</span></code>, or
<code class="code docutils literal"><span class="pre">batmobile</span></code>);</li>
<li>Instructing those <strong>objects</strong> to carry out their behaviors by ‘calling’
their <strong>methods</strong> (e.g. <code class="code docutils literal"><span class="pre">my_1986_jeep.honk()</span></code> might return “Beepity
beep!”, wheras <code class="code docutils literal"><span class="pre">batmobile.honk()</span></code> might
return “&lt;Batman theme song&gt;”).</li>
</ol>
</div></blockquote>
<p>Geonomics defines a number of <strong>classes</strong>
(such as the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>,
<code class="code py python docutils literal"><span class="name"><span class="pre">Popualtion</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> classes mentioned
above. The user will use the values they specfiy in a parameters file to
create <strong>objects</strong> belongining to these classes. Then the user will call
key <strong>methods</strong> that belong to these <strong>objects</strong>, to get them
to carry out certain behaviors (which are what constitute the simulation).</p>
<p>The subsequent documentation will present the <strong>classes</strong> definined in
Geonomics and their key <strong>methods</strong>. It will explain exactly what those methods
do, and how what they do fits into the overall structure and function of
Geonomics models.</p>
</div>
<hr class="docutils" />
<div class="section" id="getting-started">
<h4>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h4>
<dl class="docutils">
<dt><strong>For the beginner</strong>, we recommend the following steps:</dt>
<dd><ol class="first last arabic simple">
<li>Review the following three sections (‘Model organization’, ‘Data
structures and classes’, and ‘Operations’), to get a general
undertsanding of the logic, components, and necessary and optional
behaviors of a Geonomics model;</li>
<li>Skim the subsequent section (‘Parameters’), to understand the structure
of a Geonomics parameters file;</li>
<li>Pip-install Geonomics (<code class="code bash docutils literal"><span class="pre">$</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">geonomics</span></code>);</li>
<li>Open Python and run <code class="code py python docutils literal"><span class="keyword namespace"><span class="pre">import</span></span> <span class="name namespace"><span class="pre">geonomics</span></span> <span class="keyword namespace"><span class="pre">as</span></span> <span class="name namespace"><span class="pre">gnx</span></span></code>;</li>
<li>Use the <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> function, to begin
creating template parameters files that they can tweak as desired;</li>
<li>Use the <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_model</span></span></code> function and then the
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> or <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span></code> methods to instantiate and run
the models they’ve parameterized;</li>
<li>Use the various <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot</span></span></code> methods to visualize
the behavior and results of their models.</li>
</ol>
</dd>
</dl>
<p><strong>For the `impatient` beginner</strong>, as soon as Geonomics has been
installed, you should be able to run the following code:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">geonomics</span> <span class="kn">as</span> <span class="nn">gnx</span>

<span class="n">gnx</span><span class="o">.</span><span class="n">run_default_model</span><span class="p">()</span>
</pre></div>
</div>
<p>This will load the Geonomics package, create a default Geonomics
parameters file in your current working directory,
then use that file to instantiate and run a <code class="code docutils literal"><span class="pre">Model</span></code> using the default
parameter values.</p>
</div>
<hr class="docutils" />
<div class="section" id="the-documentation">
<h4>The documentation<a class="headerlink" href="#the-documentation" title="Permalink to this headline">¶</a></h4>
<p>Finally, some brief notes about the structure and style of this documentation:</p>
<blockquote>
<div><p>It is designed to be read from from the top down; information generally
becomes increasingly detailed as you scroll down). However, given the
interrelationships between all the components of a Geonomics
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, there are inevitably places where you’ll run
into material that relates to material from another section.
To the extent possible, we attempt to cross-reference rather than duplicate
information.</p>
<p>We assume, throughout, that Genomics has been imported <code class="code py python docutils literal"><span class="keyword"><span class="pre">as</span></span> <span class="name"><span class="pre">gnx</span></span></code> and
that Numpy has been imported <code class="code py python docutils literal"><span class="keyword"><span class="pre">as</span></span> <span class="name"><span class="pre">np</span></span></code>.</p>
</div></blockquote>
<p>Merry modeling!</p>
<hr class="docutils" />
</div>
</div>
<hr class="docutils" />
<div class="section" id="model-organization-and-worflow">
<h3>Model organization and worflow<a class="headerlink" href="#model-organization-and-worflow" title="Permalink to this headline">¶</a></h3>
<img alt="_images/flow_diagram.png" src="_images/flow_diagram.png" />
</div>
<hr class="docutils" />
<div class="section" id="data-structures-and-classes">
<h3>Data structures and classes<a class="headerlink" href="#data-structures-and-classes" title="Permalink to this headline">¶</a></h3>
<p>The following sections discuss the structure and function of the key
Geonomics classes. Users will interface with these classes more or less
directly when running Geonomics models, so a fundamental understanding of how
they’re organized and how they work will be useful.</p>
<div class="section" id="landscape-and-layer-objects">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> objects<a class="headerlink" href="#landscape-and-layer-objects" title="Permalink to this headline">¶</a></h4>
<p>One of the core components of a Geonomics model is the land. The land is
modeled by the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> class. This class is an
integer-keyed <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> composed of numerous instances of the
class <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>. Each <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> represents a separate
environmental variable (or ‘layer’, in GIS terminology),
which is modeled a 2d Numpy array (or raster; in
attribute ‘rast’), of identical dimensions to each
other <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> in the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>
object, and with the values of its environmental variable ‘e’ constrained to
the interval [0 &lt;= e &lt;= 1]. Each <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> can be initialized from its own
parameters subsection within the ‘land’ parameters section of a Geonomics
parameters file.</p>
<p>For each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> (see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">Individuals</span></span></code>
and <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’, below), the different <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>
layers in the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> can be used to model habitat
viability, habitat connectivity, or variables imposing spatially varying
natural selection. <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> objects
also contain some metatdata (as public attributes), including
the resolution (attribute ‘res’), upper-left corner (‘ulc’),
and projection (‘prj’), which default to 1, (0,0), and None but
will be set otherwise if some or all of the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> layers are read in from
real-world GIS rasters.</p>
</div>
<hr class="docutils" />
<div class="section" id="genomes-genomicarchitecture-and-trait-objects">
<h4>Genomes, <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> objects<a class="headerlink" href="#genomes-genomicarchitecture-and-trait-objects" title="Permalink to this headline">¶</a></h4>
<p><code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> objects (see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">Individuals</span></span></code>
and <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’, below) can optionally be assigned genomes.
If they are, each <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s genome is modeled as a
2-by-L Numpy array (where 2 is the ploidy, currently fixed at
diploidy, and L is genome length) containing 0s and 1s (because
Geonomics strictly models biallelic SNPs, i.e SNPs with ‘0’- and ‘1’-alleles).</p>
<p>The parameter L, as well as numerous other genomic parameters (including
locus-wise starting frequencies of the 1 alleles; locus-wise dominance effects;
locus-wise recombination rates; and genome-wide mutation rates for neutral,
globally deleterious, and adaptive loci), are controlled by the
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> object pertaining to the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> to which an
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> belongs. (For the full and detailed list of attributes in a
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> object, see its class documentation, below.)
The genomes of the initial <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
in a simulation, as well as those of
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s in subsequent generations, are either drawn
or recombined, and are mutated, according to the values stipulated
by the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> of
their <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. The user can create a species with a
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> and with corresponding
genomes by including a ‘genome’ subsection in that
species’ section of the Geonomics parameters file (and
setting the section’s various parameters to their desired values).</p>
<p>Geonomics can model <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ phenotypes.
It does this by allowing the
user to create an arbitrary number of distinct <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s
for each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. Each trait is
represented by a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> object, which
maps genomic loci onto that trait, maps effect sizes (‘alpha’) onto those loci,
and sets the trait’s polygenic selection
coefficient (‘phi’). An <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s
phenotype for a given trait is calculated as the ‘null phenotype’ plus a
weighted sum of the products of its ‘effective genotypes’ at all loci
underlying that <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> and the effect sizes (i.e. ‘alpha’) of those loci:</p>
<div class="math">
\[z_{i,t} = null\_genotype + \sum_{l = 0}^{n} \alpha_{t,l} g_{i,l}\]</div>
<p>where <span class="math">\(z_{i,t}\)</span> is the phenotype of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> i for trait t,
<span class="math">\(g_{i, l}\)</span> is the genotype of the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> at that locus, and
<span class="math">\(\alpha_{t,l}\)</span> is the effect size of that locus for that trait.</p>
<p>The ‘null phenotype’ refers determines what would be the phenotypic value that
an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> who is homozygyous for
the 0 allele at all loci for a trait.
For monogenic traits the null phenotype is 0 and the effect size is fixed at
0.5 (such that individuals can have phenotypes of 0, 0.5, or 1);
for polygenic traits the null phenotype is 0.5 and effect sizes can be fixed
at or distributed around a mean value (which is controlled in the
parameters file).</p>
<p>The ‘effective genotype’ refers to how the genotype is calculated based on the
dominance at a locus, as indicated by the following table of genotypes:</p>
<table border="1" class="docutils">
<colgroup>
<col width="36%" />
<col width="32%" />
<col width="32%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Biallelic genotype</th>
<th class="head">Codominant</th>
<th class="head">Dominant</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0 : 0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="row-odd"><td>0 : 1</td>
<td>0.5</td>
<td>1</td>
</tr>
<tr class="row-even"><td>1 : 1</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>(For the full and detailed list of attributes in a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> object,
see its class documentation, below.)</p>
<p>Note that for maximal control over the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>
of a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, the user can set the value of the ‘gen_arch_file’
parameter in the parameters file to the name of a separate CSV file
stipulating the locus numbers, starting 1-allele frequencies, dominance
effects, traits, and inter-locus recombination rates (as columns) of
all loci (rows) in the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>;
these values will override any other values provided in the ‘genome’
subsection of the species’ parameters.</p>
</div>
<hr class="docutils" />
<div class="section" id="individual-species-and-community-objects">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> objects<a class="headerlink" href="#individual-species-and-community-objects" title="Permalink to this headline">¶</a></h4>
<p>Being that Geonomics is an individual-based model, individuals serve as
the fundamental units (or agents) of all simulations. They are represented by
objects of the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> class.
Each <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> has an index (saved
as attribute ‘idx’), a sex (attribute ‘sex’), an age (attribute ‘age’),
an x,y position (in continuous space; attributes ‘x’ and ‘y’), and a
<code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of environment values (attribute ‘e’), extracted from the
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s current cell on each <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>
of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> on which the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> lives.</p>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> class is an <code class="code py python docutils literal"><span class="name"><span class="pre">OrderedDict</span></span></code>
(defined by the <code class="code py python docutils literal"><span class="name"><span class="pre">collections</span></span></code>
package) containing all <code class="code py python docutils literal"><span class="name"><span class="pre">Individaul</span></span></code>s, (with
their ‘idx’ attributes as keys). If a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
has a <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> then the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
in the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> will also each have genomes (attribute ‘genome’),
and the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> includes <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s
then each individual will also have a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of
phenotype values (one per <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>; attribute ‘z’) and a
single fitness value (attribute ‘fit’). (These attributes all otherwise
default to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.)</p>
<p>Each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> also has a number of other attributes of interest. Some
of these are universal (i.e. they are created regardless of the
parameterization of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> to which a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> inheres). These
include: the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ name (attribute ‘name’); its current density
raster (a Numpy array attribute called ‘N’); and the number of births,
number of deaths, and terminal population size (i.e. total number of
individuals in the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>) of each timestep (which are
<code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> attributes called ‘n_births’, ‘n_deaths’, and ‘Nt’). If the
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> was parameterized with a
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> then that will
be created as the ‘gen_arch’ attribute (otherwise this attribute will be
<code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>).</p>
<p>All of the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> in a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>
are collected in the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>’s
<code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> object. The <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> class
is simply an integer-keyed <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code>
of <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. For the time being, the <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> object allows a
Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> to simulate multiple <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> simultaneously on
the same <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, but otherwise affords no additional functionality
of interest. However, its implementation will facilitate the potential
future development of methods for interaction between <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>.
(e.g. to simulate coevolutionary, speciation, or hybridization scenarios).</p>
</div>
<hr class="docutils" />
<div class="section" id="model-objects">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> Objects<a class="headerlink" href="#model-objects" title="Permalink to this headline">¶</a></h4>
<p>Objects of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> class serve as the main interface between the user
and the Geonomics program. (While it is certainly possible for a user
to work directly with the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>
and <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> or <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> objects to
script their own custom models, the typical user should find that the
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object allows them accomplish their goals with minimal toil.)
The main affordance of a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object is the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span></code> method,
which, as one could guess, will run the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>. The typical workflow
for creating and running a  <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object is as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Create a template paramters file containing the desired sections,
by calling <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> with all revelant arguments;</li>
<li>Define the scenario to be simulated, by opening and editing that
parameters file (and optionally, creating/editing corresponding
files, e.g. genomic-architecture CSV files;
or raster or numpy-array files to be used as <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s);</li>
<li>Instantiate a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object from that parameters file, by calling
<code class="code py python docutils literal"><span class="name"><span class="pre">mod</span></span> <span class="operator"><span class="pre">=</span></span> <span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_model</span></span><span class="punctuation"><span class="pre">(</span></span><span class="literal string single"><span class="pre">‘/path/to/params_filename.py’</span></span><span class="punctuation"><span class="pre">)</span></span></code>;</li>
<li>Run the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, by calling <code class="code py python docutils literal"><span class="name"><span class="pre">mod</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span><span class="punctuation"><span class="pre">()</span></span></code>.</li>
</ol>
</div></blockquote>
<p>For detailed information on usage of these functions, see their docstrings.
When a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is run, it will:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Run the burn-in (until the mininmal burn-in length stipulated in the
parameters file and the built-in stationarity statistics
determine that the burn-in is complete);</li>
<li>Run the main model for the stipulated number of timesteps;</li>
<li>Repeat this for the stipulated number of iterations (retaining or
refreshing the first run’s initial <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
objects as stipulated).</li>
</ol>
</div></blockquote>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object offers one other method, however, <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code>,
which allows the user to run a model, in either ‘burn’ or ‘main’ mode,
for an arbitrary number of timesteps within a single iteration (see its
docstring for details). This is particularly useful for running
Geonomics within an interactive Python session. Thus, <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> is
primarily designed for passively running numerous iterations of a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>,
to generate data for analysis, whereas <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> is primarily designed
for the purposes of learning, teaching, or debugging the package, or
developing, exploring, introspecting, or visaulizing particular <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>s.</p>
</div>
<hr class="docutils" />
<div class="section" id="secondary-classes">
<h4>Secondary classes<a class="headerlink" href="#secondary-classes" title="Permalink to this headline">¶</a></h4>
<p>The typical user will not need to access or interact with the following
classes in any way. They will, however, parameterize them in the
parameters file by either leaving or altering their default values. Geonomics
sets generally sensible default parameter values wherever possible,
but for some scenarios they may not be adequate, and for some parameters
(e.g. the window-width used by the _DensityGridStack; see below), there is
no “one-size-fits-most” option. Thus, it is important that the user
have a basic acquaintance with the purpose and operation of these classes.</p>
<div class="section" id="conductancesurface">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code><a class="headerlink" href="#conductancesurface" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> class allows Geonomics
to model a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’
realistic movement across a spatially varying landscape. It does this by
creating an array of circular probability distributions (i.e. VonMises
distributions), one for each cell on the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, from which
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s choose their directions each time they move. To create the
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>,
the user must indicate the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>
that should be used to create it (i.e. the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that represents
landscape permeability for that <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>).
The <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>’s
distributions can be <strong>simple (i.e. unimodal)</strong>, such that the
maximum value of the distribution at each cell will point toward the
maximum value in the 8-cell neighborhood; this works best for permeability
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s with shallow, monotonic gradients, because the differences
between permeability values of neighboring cells can be minor (e.g. a
gradient representing the directionality of a prevalent current).
Alternatively, the distributions can be <strong>mixture (i.e. multimodal)</strong>
distributions, which are weighted sums of 8 unimodal distributions, one
for each neighboring cell, where the weights are the relative cell
permeabilities (i.e. the relative probabilities that an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> would
move into each of the 8 neighboring cells); this works best for non-monotonic,
complex permeability <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s (e.g. a DEM of a mountainous region that is
used as a permeability <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>).
(The <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> is surrounded by a margin of 0-permeability
cells before the <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is calculated, such
that <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> edges are treated
as barriers to movement.) The class consists
principally of a 3d Numpy array (x by y by z, where x and y are the
dimensions of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> and z is the length of the vector of values
used to approximate the distributions in each cell.</p>
</div>
<div class="section" id="densitygridstack">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_DensityGridStack</span></span></code><a class="headerlink" href="#densitygridstack" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_DensityGridStack</span></span></code> class implements an algorithm for rapid estimating
an array of the local density of a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. The density is estimated
using a sliding window approach, with the window-width determining the
neighborhood size of the estimate. The resulting array has a spatial
resolution equivalent to that of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, and is used in all
density-dependent operations.</p>
</div>
<div class="section" id="kdtree">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_KDTree</span></span></code><a class="headerlink" href="#kdtree" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_KDTree</span></span></code> class is just a wrapper around <code class="code py python docutils literal"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">spatial</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">cKDTree</span></span></code>.
It provides an optimized algorithm (the kd-tree) for finding
neighboring points within a given search radius.
This class is used for all neighbor-searching operations (e.g. mate-search).</p>
</div>
<div class="section" id="recombinationpaths">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_RecombinationPaths</span></span></code><a class="headerlink" href="#recombinationpaths" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_RecombinationPaths</span></span></code> class contains a large (and customizable)
number of <code class="code py python docutils literal"><span class="name"><span class="pre">bitarray</span></span></code>s, each of which indicates the genome-length
diploid chromatid numbers (0 or 1) for a
recombinant gamete produced by an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> of a given <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
(henceforth referred to as ‘recombination paths’). These recombination
paths are generated using the genome-wide recombination rates specified by
the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ <code class="code py python docutils literal"><span class="name"><span class="pre">GeonomicArchitecture</span></span></code>. They are generated during
construction of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, then drawn randomly as needed (i.e.
each time an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> produces a gamete). This provides a
reasonable trade-off between realistic modelling of recombination and runtime.</p>
</div>
<div class="section" id="landscapechanger-and-specieschanger">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_LandscapeChanger</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">_SpeciesChanger</span></span></code><a class="headerlink" href="#landscapechanger-and-specieschanger" title="Permalink to this headline">¶</a></h5>
<p>These classes manage all of the landscape changes and demographic changes
that were parameterized for the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> and
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> objects to which they inhere.
The functions creating these changes are defined at the outset,
then queued and called at their scheduled timesteps.</p>
</div>
<div class="section" id="datacollector-and-statscollector">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_DataCollector</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">_StatsCollector</span></span></code><a class="headerlink" href="#datacollector-and-statscollector" title="Permalink to this headline">¶</a></h5>
<p>These classes manage all of the data and statistics that should be collected
and written to file for the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object to which they inhere
(as determined by the parameters file used the create the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>).
The types of data to be collected, or statistics to be calculated, as
well as the timesteps at which and methods by which they’re
collected/calculated and determined at the outset, then the
appropriate functions called at the appropriate timesteps.</p>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="operations">
<h3>Operations<a class="headerlink" href="#operations" title="Permalink to this headline">¶</a></h3>
<p>The following sections discuss the mechanics of core Geonomics operations.
The material here is inevitably intertwined with some of the material in
the “Data structures and classes” section. To the extent possible, we
attempt to cross-reference rather than duplicate information (with
the exception of this sentence).</p>
<div class="section" id="movement-and-dispersal">
<h4>Movement and Dispersal<a class="headerlink" href="#movement-and-dispersal" title="Permalink to this headline">¶</a></h4>
<p>Movement is optional, such that turning off movement will allow the user
to simulate sessile organisms (which will reproduce and disperse,
but not move after dispersal; this distinction is of course irrelevant
for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> with a maximum age of 1). For <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
with movement, <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s can
move by two distinct mechanisms. <strong>Spatially random movement</strong>
is the default behavior; in this case, <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
move to next locations that are determined by a random distance drawn
from a Wald distribution and a random direction drawn from a uniform
circular (i.e. Von Mises) distribution.  As with most distributions used
in Geonomics, the parameters of these distributions have sensible
default values but can be customized in a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>’s parameters file
(see section ‘Parameters’, below).</p>
<p>The alternative movement mechanism that is available is
<strong>movement across a permeability surface</strong>,
using a <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> object.
To parameterize a <code class="code py python docutils literal"><span class="name"><span class="pre">_MovemementSurface</span></span></code> for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, the user
must create a template parameters file that includes the
necessary parameters section for the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> (i.e.
the user must set ‘movement’ to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">True</span></span></code> and ‘movement_surface’ to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">True</span></span></code>
in the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ arguments to the <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code>
function (see the docstring for that function for details and an example).
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s move to next locations determined by a random distance drawn
from a Wald distribution and a random direction drawn from the distribution
at the  <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> cell in which which the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
are currently located. For details about <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> creation,
see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>’ above, or the class’ docstring.</p>
<p>Dispersal is currently implemeneted identically to spatially random movement
(with the caveat that the an offspring’s new location is determined
relative its parents’ midpoint). But the option to use a
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for dispersal will be offered soon.</p>
</div>
<hr class="docutils" />
<div class="section" id="reproduction">
<h4>Reproduction<a class="headerlink" href="#reproduction" title="Permalink to this headline">¶</a></h4>
<p>Each timestep, for each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, all pairs of individuals within
a certain distance of each other (i.e. the mating radius,
which is set in the parameters file) are identified.
These pairs are subsetted if necessary (i.e. if the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
requires that <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s be above a certain reproductive age,
or that they be of opposite sexes, in order to mate; these values
can also be changed from their defaults in the parameters file).
Remaining pairs mate probabilistically (according to a Bernoulli
random draw with probability equal to the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ birth
rate, which is also set in the parameters file).</p>
<p>Pairs that are chosen to mate will produce a number of new
offspring drawn from a Poisson distribution (with lambda set in the
parameters file). For each offspring, sex is chosen probablistically
(a Bernoulli random draw with probability equal to the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’
sex ratio), age set to 0, and location chosen by dispersal from
the parents’ midpoint (see section ‘Movement and Dispersal’). For
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> that have genomes, offspring genomes will be a
fusion of two recombinant genomes from each of the two parents (where
each recombinant is indexed out a parent’s genome using a recombination
path; see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">_RecombinationPaths</span></span></code>’). For <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
with <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s in their
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>s, offspring phenotypes are
determined at birth. Mutations are also drawn and introduced at this
point (see section ‘Mutation for details).</p>
</div>
<hr class="docutils" />
<div class="section" id="mortality">
<h4>Mortality<a class="headerlink" href="#mortality" title="Permalink to this headline">¶</a></h4>
<p>Mortality can occur as a combination of two factors: <strong>density dependence</strong>
and <strong>natural selection</strong>. Each <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> has a death decision drawn
as a Bernoulli random variable with
<span class="math">\(P(d_{i}) = 1 - P(s_{i_{dens}})P(s_{i_{fit}})\)</span>, where <span class="math">\(P(d_{i})\)</span>
is the probability of death of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span>, and
<span class="math">\(P(s_{i_{dens}})\)</span> and <span class="math">\(P(s_{i_{fit}})\)</span> are the probabilities of
survival of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span> given density-dependence and
fitness. The probability of density-dependent death is contingent on an
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s x,y location
(i.e. the cell in which they’re currently located.
And an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s probability of survival due to fitness
is just equal to the product of their absolute fitness (<span class="math">\(\omega\)</span>)
for each of the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s <span class="math">\(m\)</span> <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s.
Thus the equation for an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s probability of death becomes:</p>
<div class="math">
\[P(d_{i}) = 1 - (1 - P(d_{x,y})) \prod_{p = 1}^{m}\omega_{i,p}\]</div>
<p>The following two sections explain in detail the implementation and
calculation of the two halves of the right side of this equation.</p>
<div class="section" id="density-dependence">
<h5>Density dependence<a class="headerlink" href="#density-dependence" title="Permalink to this headline">¶</a></h5>
<p>Density dependence is implemented using a spatialized form of the class
logistic growth equation (<span class="math">\(\frac{\mathrm{d}
N_{x,y}}{\mathrm{d}t}=rN_{x,y}(1-\frac{N_{x,y}}{K_{x,y}})\)</span>,
where the x,y subscripts refer to
values for a given cell on the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>).
Each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> has a carrying-capacity raster (a 2d Numpy array;
attribute ‘K’), which is defined in the parameters file to be
one of the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s in the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>.
The comparison between this raster and
the population-density raster calculated at each timestep serves as the
basis for the spatialized logistic growth equation, because both
equations can be calculated cell-wise for the entire extent of the
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> (using the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’
intrinsic growth rate, the attribute
‘R’, which is set in the parameters file).</p>
<p>The logistic equation returns an array of instantaneous population growth
rates within each cell. We can derive from this the density-dependent
probability of death at each cell by subtracting an array of the expected
number of births at each cell, then dividing by the array of
population density:</p>
<div class="math">
\[P(d_{x,y}) = E[N_{d;x,y}]/N_{x,y} = \frac{E[N_{b;x,y}]
 - \frac{\mathrm{d}N_{x,y}}{\mathrm{d}t}}{N_{x,y}}\]</div>
<p>The expected number of births at each cell is calculated as a density
raster of the number of succesful mating pairs, multiplied by the expected
number of births per pair (i.e. the expectation of the Poisson
distribution of the number of offspring per mating pair, which
is just the distribution’s paramater lambda).</p>
</div>
<div class="section" id="selection">
<h5>Selection<a class="headerlink" href="#selection" title="Permalink to this headline">¶</a></h5>
<p>Selection on a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> can exhibit three regimes: <strong>spatially divergent</strong>,
<strong>universal</strong>, and <strong>spatially contingent</strong>. <strong>Spatially divergent</strong> selection
is the default behavior, and the most commonly used; in this form of
selection, an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s fitness depends on the absolute difference
between the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s phenotypic value and the environmental
value of the relevant <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> (i.e. the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that represents the
environmental variable acting as the selective force) in the cell where
the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> is located.</p>
<p><strong>Universal</strong> selection (which can be toggled using the ‘univ_adv’
parameter with a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>’s section in the parameters file) occurs
when a phenotype of 1 is optimal everywhere on the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>. In other
words, it represents directional selection on an entire <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>,
regardless of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ spatial contexts. (Note that this can
be thought of as operating the same as spatially divergent selection,
but with the environmental variable driving natural selection being
represented by an array in which all cells are equal to 1.)</p>
<p>Under <strong>spatially contingent</strong> selection, the selection coefficient of a
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> varies across space, such that the strength of selection
is environmentally determined in some way. Importantly, this selection regime
is <em>not mutually exclusive</em> with the other two; in other words,
selection on a certain <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> be both spatially contingent
and either spatially divergent or universal. Spatially contingent selection
can be implemented by providing an array of values (equal in dimensions
to the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>) to the ‘phi’ value of a
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>, rather than a scalar
value (which could be done within the parameters file itself, but may be
more easily accomplished as a step between reading in a parameters file and
instantiating a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object from it). (Note that non-spatailly
cotingent selection could in fact be thought of as a special case of
spatially contingent selection, but where the array of selection-coefficients
has the same value at each cell.)</p>
<p>All possible combinations of the three selection regimes of selection can all
be thought of as special cases of the following equation for the fitness of
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span> for <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> <span class="math">\(p\)</span> (<span class="math">\(\\omega_{i,p}\)</span>):</p>
<div class="math">
\[\omega_{i,p}= 1 - \phi_{p;x,y} (\mid e_{p;x,y} - z_{i;p} \mid)^{\gamma_{p}}\]</div>
<p>where <span class="math">\(\\phi_{p;x,y}\)</span> is the selection coefficient of trait
<span class="math">\(p\)</span>; <span class="math">\(e_{p;x,y}\)</span> is the environmental variable of the
relevant <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> at <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span>‘s x,y location
(which can also be thought of as the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s optimal
phenotype); <span class="math">\(z_{i;p}\)</span> is <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span>‘s (actual)
phenotype for <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> <span class="math">\(p\)</span>; and <span class="math">\(gamma_{p}\)</span> controls
the curvature of the fitness function (i.e. how fitness decreases as
the absolute difference between an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s
optimal and actual phenotypes increases; the default value of 1 causes
fitness to decrease linearly around the optimal phenotypic value).</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="mutation">
<h4>Mutation<a class="headerlink" href="#mutation" title="Permalink to this headline">¶</a></h4>
<p>Geonomics can model mutations of three different types: <strong>neutral</strong>,
<strong>deleterious</strong>, and <strong>trait</strong> mutations. These terms don’t map
precisely onto the traditional population-genetic
lingo of “neutral”, “deleterious”, and “beneficial”, but they
are more or less analogous:</p>
<ul class="simple">
<li><strong>Neutral</strong> mutations are the same conceptually in Geonomics as
they are in the field of population genetics in general:
They are mutations that have no effect on the fitness of
the individuals in which they occur.</li>
<li><strong>Deleterious</strong> mutations in Geonomics are also conceptually the
same in Geonomics and in population genetics: They negatively impact
the fitness of the individuals in which they occur.</li>
<li><strong>Trait</strong> mutations are the place where the Geonomics concept and
the population-genetic concept diverge: In Geonomics, natural selection
acts on the phenotype, not the genotype (although these concepts are
identical if a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> in monogenic), and it is (by default,
but not always; see section ‘Selection’, above) divergent. For this reason
it would be a misnomer to call mutations that influence a given
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>’s phenotypes ‘beneficial’ – even though that term is the closest
population-genetic concept to this concept as it is employed in Geonomics –
because the same mutant genotype in the same <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>
could have opposite effects on that <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s fitness
in different environmental contexts (i.e. it could behave as
a beneficial mutation is one region of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>
but as a deleterious mutation in another).</li>
</ul>
</div>
<hr class="docutils" />
<div class="section" id="species-interactions">
<h4>Species interactions<a class="headerlink" href="#species-interactions" title="Permalink to this headline">¶</a></h4>
<p>This functionality is not yet included available. But the Community class was
created in advance recognition that this functionality could be desirable
for future versions (e.g. to simulate coevolutionary, speciation, or
hybridization scenarios).</p>
</div>
<hr class="docutils" />
<div class="section" id="landscape-and-species-change">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> change<a class="headerlink" href="#landscape-and-species-change" title="Permalink to this headline">¶</a></h4>
<p>For a given <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>, any number of change events
can be planned.
In the parameters file, for each event, the user stipulates the initial
timestep; the final timestep; the end raster (i.e. the array
of the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that will exist after the event is complete, defined using
the <strong>end_rast</strong> parameter); and the
interval at which intermediate changes will occur.  When the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is
created, the stepped series of intermediate <code class="code py python docutils literal"><span class="name"><span class="pre">Layers</span></span></code> (and
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> objects,
if the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that is changing serves as the basis for a
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for any <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>) will be
created and queued, so that they will swap out accordingly at the appropriate
timesteps.</p>
<p>For a given <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, any number of demographic change events can
also be planned. In the parameters file, for each event, the user
stipulates the type of the event (‘monotonic’, ‘cyclical’, ‘random’, or
‘custom’) as well as the values of a number of associated
parameters (precisely which parameters depdends on the type of event chosen).
As with <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> change events, all necessary stepwise changes will be
planned and queued when the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is created, and will be
executed at the appropriate timesteps.</p>
<p>It is also possible to schedule any number of instantaneous changes
to some of the life-history parameters of a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> (e.g. birth rate;
the lambda parameter of the Poisson distribution determining the number of
offspring of mating events). This functionality is currently minimalistic,
but will be more facilitated in future versions.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="visualization">
<h3>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">¶</a></h3>
<p>Each <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object has a variety of visualization methods
(<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_fitness</span></span></code>, etc.),
which aim to help users design, run, explore, present,
and explain their <code class="code py python docutils literal"><span class="name"><span class="pre">Models</span></span></code>’ behavior and results.
These methods can be called at any time (e.g. as
soon as the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> has been created, or after it has
run for any number of timesteps); but it is worth mentioning that some
methods may be invalid depending on the point in model-time at
which they’re called (e.g. <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_genotype</span></span></code>,
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_phenotype</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_fitness</span></span></code>
cannot be run for <code class="code py python docutils literal"><span class="name"><span class="pre">Models</span></span></code> that have not yet been burned in,
as they will not yet have genomes assigned), or on
the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> for which they’re called
(e.g. the aforementioned methods cannot create plots for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
that has no <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>; and likewise, the
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_demographic_changes</span></span></code> method cannot be called for a
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> for which demographic changes were not parameterized).</p>
<p>Below is a list of the visualization methods available, with example
output for each (generated from the default Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>):</p>
<p>:: VIZ FUNCTION</p>
<p>:: PLOT</p>
<p>:: VIZ FUNCTION</p>
<p>:: PLOT</p>
<p>:: …</p>
</div>
<hr class="docutils" />
<div class="section" id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h3>
<p>In order to create and run a Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, you will need a valid
Geonomics parameters file. No worry though – this is very easy to create!
To generate a new, template parameters file, you will simply call the
<code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> function, feeding it the appropriate
arguments (to indicate how many <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s you
want to include in your <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>; which parameters sections you want
included in the file, both for those
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s and <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> and for
other components of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>; and the path and filename for your new
parameters file). Geonomics will then automatically create the file for you,
arranged as you requested and saved where you requested.</p>
<p>When you then open that file, you will see the following:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#&lt;your_filename&gt;.py</span>

<span class="c1">#This is a default parameters file generated by Geonomics</span>
<span class="c1">#(by the gnx.params.make_parameters_file() function).</span>


                      <span class="c1">## :: ::    :::            ##</span>
                <span class="c1">##:::   :::::    :::   ::    :: :: :::##</span>
             <span class="c1">## ::::     ::           ::   ::::::::::::::##</span>
           <span class="c1">##::::::                       ::::::::: :::::: :##</span>
         <span class="c1">## :    :::                    :::    ::    :::::::::##</span>
        <span class="c1">##ggggg eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
       <span class="c1">##g     e    o   o  nn  n  o   o  m   m   i   c     s    ##</span>
       <span class="c1">##g     eee o     o n n n o     o mm mm   i   c     sssss##</span>
       <span class="c1">##g ggg eee o     o n  nn o     o m m m   i   c         s##</span>
       <span class="c1">##g   g e    o   o  n   n  o   o  m   m   i   c        ss##</span>
        <span class="c1">##gggg  eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
         <span class="c1">##  ::::::::        ::::::::::::  :       ::  ::   : ##</span>
           <span class="c1">##  ::::              :::::::  ::     ::::::::  :##</span>
             <span class="c1">## :::               :::::: ::       ::::::  ##</span>
                <span class="c1">##:                ::::                ##</span>
                      <span class="c1">##                         ##</span>


<span class="n">params</span> <span class="o">=</span> <span class="p">{</span>

<span class="c1">##############</span>
<span class="c1">#### LAND ####</span>
<span class="c1">##############</span>
    <span class="s1">&#39;land&#39;</span><span class="p">:</span> <span class="p">{</span>

    <span class="c1">##############</span>
    <span class="c1">#### main ####</span>
    <span class="c1">##############</span>
        <span class="s1">&#39;main&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># dimensions of the Landscape</span>
            <span class="s1">&#39;dim&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>

   <span class="c1">#.</span>
   <span class="c1">#.</span>
   <span class="c1">#.</span>
</pre></div>
</div>
<p>This is the beginning of a file that is really just a long but simple Python
script (hence the ‘.py’ extension); this whole file just defines a single,
long, nested <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> (i.e. a Python ‘dictionary’) containing all of your
parameter values. It may look like a lot, but don’t be concerned! For two
reasons:</p>
<blockquote>
<div><ol class="arabic simple">
<li>All the hard work is already done for you. You’ll just need to change
the default values where and how you want to, to set up your particular
simulation scenario.</li>
<li>You will probably leave a good number of the parameters defined in this
file untouched. Geonomics does its best to set sensible default values
for all its parameters. Though of course, you’ll want to think clearly
nonetheless about whether the default value for each parameter
is satisfactory for your purposes.</li>
</ol>
</div></blockquote>
<p>Each parameter in the parameters value is preceded by a terse comment, to
remind you what the parameter does. But for detailed information about each
parameter, you’ll want to refer to the following information.
What follows is a list of all of the Geonomics parameters (in the sections and
the top-to-bottom order in which they’ll appear in your parameters files).
For each parameter, you will see a section with the following information:</p>
<blockquote>
<div><ul class="simple">
<li>a snippet of the context (i.e. lines of
Python code) in which it appears in a parameters file;</li>
<li>the valid Python data type(s) the parameter can take</li>
<li>the default value of the parameter</li>
<li>a ranking score, indicating how likely it is that you will want to reset
this parameter (i.e. change it from its default value), and
encoded as follows:<ul>
<li>‘Y’: almost certainly, <em>or</em> must be reset for your <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> to run</li>
<li>‘P’: it is quite possible that you will want to reset this
parameter, but this will depend on your use and scenario</li>
<li>‘N’: almost certainly not, <em>or</em> no need to reset because it should be
set intelligently anyhow (Note: this does <em>not</em> mean that you cannot
reset the parameter! if that is the case for any value then it does not
appear in the parameters file)</li>
</ul>
</li>
<li>other relevant, detailed information about the parameter, including
an explanation of what it defines, how its value is used, where to look
for additioanl information about parameters related to other Python
packages, etcetera</li>
</ul>
</div></blockquote>
<p>These section will be formatted as follows:</p>
<p><strong>&lt;param_name&gt;</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#brief comment about the parameter</span>
<span class="s1">&#39;&lt;param_name&gt;&#39;</span><span class="p">:</span>               <span class="o">&lt;</span><span class="n">default_param_value</span><span class="o">&gt;</span><span class="p">,</span>
</pre></div>
</div>
<p>&lt;valid Python data type(s)&gt;</p>
<p>default: &lt;default value&gt;</p>
<p>reset? &lt;ranking&gt;</p>
<blockquote>
<div>&lt;Explanation of what the parameter defines, how its value is used,
and any other relevant information.&gt;</div></blockquote>
<p>This section should serve as your primary point of reference
if you confront any uncertainty while creating your own parameters files.
We’ll start with the section of parameters that
pertains to the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> object.</p>
<div class="section" id="landscape-parameters">
<h4>Landscape parameters<a class="headerlink" href="#landscape-parameters" title="Permalink to this headline">¶</a></h4>
<div class="section" id="main">
<h5>Main<a class="headerlink" href="#main" title="Permalink to this headline">¶</a></h5>
<hr class="docutils" />
<p><strong>dim</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># dimensions of the Landscape</span>
<span class="s1">&#39;dim&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">(</span></span><span class="literal number integer"><span class="pre">20</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">20</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset: P</p>
<blockquote>
<div>This defines the x and y dimensions of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>,
in units of cells. As you might imagine, these values are used
for a wide variety of basic operations throughout Geonomics. Change the
default value to the dimensions of the landscape you wish to simulate on.</div></blockquote>
<hr class="docutils" />
<p><strong>res</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># resolution of the Landscape</span>
<span class="s1">&#39;res&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">(</span></span><span class="literal number integer"><span class="pre">1</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">1</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset: N</p>
<blockquote>
<div>This defines the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> resolution (or cell-size) in the x and y
dimensions. This information is only used if GIS rasters of <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>
layers are to be written out as GIS raster files (as parameterized in the
‘Data’ parameters). Defaults to the meaningless value (1,1), and this value
generally needn’t be changed in your parameters file, because it will
be automatically updated to the resolution of any GIS rasters that
are read in for use as <code class="code py python docutils literal"><span class="name"><span class="pre">Layers</span></span></code> (assuming they all share the same
resolution; otherwise, an Error is thrown).</div></blockquote>
<hr class="docutils" />
<p><strong>ulc</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># upper-left corner of the Landscape</span>
<span class="s1">&#39;ulc&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">(</span></span><span class="literal number integer"><span class="pre">0</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">0</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset: N</p>
<blockquote>
<div>This defines the upper-left corner (ULC) of the
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> (in the units of
some real-world coordinate reference system, e.g. decimal degrees, or
meters). This information is only used if GIS rasters of
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> layers are to be written out as GIS raster files.
Defaults to the meaningless value
(0,0), and this value usually needn’t be changed in your parameters file,
because it will be automatically updated to match the ULC value
of any GIS rasters that are read in for use as <code class="code py python docutils literal"><span class="name"><span class="pre">Layers</span></span></code> (assuming
they all share the same ULC; otherwise, an Error is thrown).</div></blockquote>
<hr class="docutils" />
<p><strong>prj</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#projection of the Landscape</span>
<span class="s1">&#39;prj&#39;</span><span class="p">:</span>                      <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>; (WKT projection string)</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset: N</p>
<blockquote>
<div>This defines the projection of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, as a
string of Well Known Text (WKT).
This information is only used if GIS rasters of <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> layers are
to be written out as GIS raster files. Defaults to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, which is fine,
because this value will be automatically updated to match the projection
of any GIS rasters that are read in for us as <code class="code py python docutils literal"><span class="name"><span class="pre">Layers</span></span></code> (assuming they
all share the same projection; otherwise, an Error is thrown)</div></blockquote>
</div>
<div class="section" id="layers">
<h5>Layers<a class="headerlink" href="#layers" title="Permalink to this headline">¶</a></h5>
<hr class="docutils" />
<p><strong>layer_&lt;n&gt;</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">################</span>
<span class="c1">#### layers ####</span>
<span class="c1">################</span>
    <span class="s1">&#39;layers&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="c1">#layer name (LAYER NAMES MUST BE UNIQUE!)</span>
        <span class="s1">&#39;layer_0&#39;</span><span class="p">:</span> <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name"><span class="pre">layer_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>. (Note that unlike most
parameters, this parameter is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> key,
the value for which is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> must have a unique name
(so that a parameterized <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> isn’t overwritten in the
<code class="code py python docutils literal"><span class="name"><span class="pre">ParametersDict</span></span></code> by a second, identically-named <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> represents. Example valid values include: 0, 0.1, ‘layer_0’, 1994,
‘1994’, ‘mean_ann_tmp’. Names default to <code class="code py python docutils literal"><span class="name"><span class="pre">layer_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>,
where n is a series of integers starting from 0 and counting the number
of <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s.</p>
<div class="section" id="init">
<h6>Init<a class="headerlink" href="#init" title="Permalink to this headline">¶</a></h6>
<p>There are four different types of <code class="code py python docutils literal"><span class="name"><span class="pre">Layers</span></span></code> that can be created. The
parameters for each are explained in the next four subsections.</p>
<div class="section" id="random">
<h7>random<a class="headerlink" href="#random" title="Permalink to this headline">¶</a></h7>
<hr class="docutils" />
<p><strong>n_pts</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;random&#39;-type Layer</span>
<span class="s1">&#39;rand&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#number of random points</span>
    <span class="s1">&#39;n_pts&#39;</span><span class="p">:</span>                        <span class="mi">500</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 500</p>
<p>reset? P</p>
<p>This defines the number of randomly located, randomly valued points
from which the random <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> will be interpolated. (Locations drawn
from uniform distributions between 0 and the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> dimensions on
each axis. Values drawn from a uniform distribution between 0 and 1.)</p>
<hr class="docutils" />
<p><strong>interp_method</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#interpolation method (&#39;linear&#39;, &#39;cubic&#39;, or &#39;nearest&#39;)</span>
<span class="s1">&#39;interp_method&#39;</span><span class="p">:</span>                <span class="s1">&#39;cubic&#39;</span><span class="p">,</span>
<span class="p">},</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘linear’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cubic’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘nearest’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cubic’</span></span></code></p>
<p>reset? N</p>
<p>This defines the method to use to interpolate random points to the array that
will serve as the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>’s raster. Whichever of the three valid values
is chosen (<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘linear’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cubic’</span></span></code>, or <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘nearest’</span></span></code>) will be passed
on as an argument to <code class="code py python docutils literal"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">interpolate</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">griddata</span></span></code>. Note that the
<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘nearest’</span></span></code> method will generate a random categorical array, such as
might be used for modeling habitat types.</p>
</div>
<div class="section" id="defined">
<h7>defined<a class="headerlink" href="#defined" title="Permalink to this headline">¶</a></h7>
<hr class="docutils" />
<p><strong>pts</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;defined&#39;-type Layer</span>
<span class="s1">&#39;defined&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#point coordinates</span>
    <span class="s1">&#39;pts&#39;</span><span class="p">:</span>                    <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>nx2 <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? Y</p>
<p>This defines the coordinates of the points that will be used to
interpolate this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>.</p>
<hr class="docutils" />
<p><strong>vals</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#point values</span>
<span class="s1">&#39;vals&#39;</span><span class="p">:</span>                  <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code>, 1xn <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? Y</p>
<p>This defines the values of the points that will be used to
interpolate this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>.</p>
<hr class="docutils" />
<p><strong>interp_method</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#interpolation method {&#39;linear&#39;, &#39;cubic&#39;, &#39;nearest&#39;}</span>
<span class="s1">&#39;interp_method&#39;</span><span class="p">:</span>                <span class="s1">&#39;cubic&#39;</span><span class="p">,</span>
<span class="p">},</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘linear’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cubic’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘nearest’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cubic’</span></span></code></p>
<p>reset? N</p>
<p>This defines the method to use to interpolate random points to the array that
will serve as the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>’s raster. Whichever of the three valid values
is chosen (<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘linear’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cubic’</span></span></code>, or <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘nearest’</span></span></code>) will be passed
on as an argument to <code class="code py python docutils literal"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">interpolate</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">griddata</span></span></code>. Note that the
<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘nearest’</span></span></code> method will generate a random categorical array, such as
might be used for modeling habitat types.</p>
</div>
<div class="section" id="file">
<h7>file<a class="headerlink" href="#file" title="Permalink to this headline">¶</a></h7>
<hr class="docutils" />
<p><strong>filepath</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;file&#39;-type Layer</span>
<span class="s1">&#39;file&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#&lt;/path/to/file&gt;.&lt;ext&gt;</span>
    <span class="s1">&#39;filepath&#39;</span><span class="p">:</span>                     <span class="s1">&#39;/PATH/TO/FILE.EXT&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘/PATH/TO/FILE.EXT’</span></span></code></p>
<p>reset? Y</p>
<p>This defines the location and name of the file that should be read in as the
raster-array for this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>. Valid file types include a ‘.txt’ file
containing a 2d <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>, or any GIS raster file that can be read
by <code class="code py python docutils literal"><span class="name"><span class="pre">osgeo</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">gdal</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">Open</span></span></code>. In all cases, the raster-array read in from the
file must have dimensions equal to the stipulated dimensions of the
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> (as defined in the <strong>dims</strong> parameter, above); otherwise,
Geonomics will throw an Error.</p>
<hr class="docutils" />
<p><strong>scale_min_val</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#minimum value to use to rescale the Layer to [0,1]</span>
<span class="s1">&#39;scale_min_val&#39;</span><span class="p">:</span>                <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This defines the minimum value (in the units of the variable represented by
the file you are reading in) to use when rescaling the file’s array to
values between 0 and 1. (This is done to satisfy the requirement that all
Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s have arrays in that interval). Defaults to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>
(in which case Geonomics will set it to the minimum value observed in this
file’s array). But note that you should put good thought into
this parameter, because it <em>won’t</em> necessarily be the minimum value
observed in the file; for example, if this file is being used
to create a <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that will undergo environmental change
in your <cite>Model</cite>, causing its real-world values to drop
below this file’s minimum value, then you will probably want to set
this value to the minimum real-world value that will occur for this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>
during your <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> scenario, so that low values
that later arise on this <cite>Layer</cite> don’t get truncated at 0.</p>
<hr class="docutils" />
<p><strong>scale_max_val</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#maximum value to use to rescale the Layer to [0,1]</span>
<span class="s1">&#39;scale_max_val&#39;</span><span class="p">:</span>                <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This defines the maximum value (in the units of the variable represented by
the file you are reading in) to use when rescaling the file’s array to
values between 0 and 1. (This is done to satisfy the requirement that all
Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s have arrays in that interval). Defaults to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>
(in which case Geonomics will set it to the maximum value observed in this
file’s array). But note that you should put good thought into
this parameter, because it <em>won’t</em> necessarily be the maximum value
observed in the file; for example, if this file is being used
to create a <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that will undergo environmental change
in your <cite>Model</cite>, causing its real-world values to increase
above this file’s maximum value, then you will probably want to set
this value to the maximum real-world value that will occur for this
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> during your <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> scenario, so that high values that
later arise on this <cite>Layer</cite> don’t get truncated at 1.</p>
</div>
<div class="section" id="nlmpy">
<h7>nlmpy<a class="headerlink" href="#nlmpy" title="Permalink to this headline">¶</a></h7>
<hr class="docutils" />
<p><strong>function</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#parameters for an &#39;nlmpy&#39;-type Layer</span>
<span class="s1">&#39;nlmpy&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#nlmpy function to use the create this Layer</span>
    <span class="s1">&#39;function&#39;</span><span class="p">:</span>                 <span class="s1">&#39;mpd&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code> that is the name of an <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> function</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘mpd’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> function that should be used to generate
this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>’s array. (<code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> is a Python package for
generating neutral landscape models; NLMs.) Defaults to <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘mpd’</span></span></code> (the
function for creating a midpoint-displacement NLM). Can be set to any other
<code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code> that identifies a valid <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> function, but then the
remaining parameters in this section must be changed to the parameters
that that function needs, and <em>only</em> those parameters
(because they will be unpacked into this function,
i.e. passed on to it, at the time it is called.
(Visit the <a class="reference external" href="https://pypi.org/project/nlmpy/">Cheese Shop</a> for more
information about the <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> package and available functions).</p>
<hr class="docutils" />
<p><strong>nRow</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#number of rows (MUST EQUAL LAND DIMENSION y!)</span>
<span class="s1">&#39;nRow&#39;</span><span class="p">:</span>                     <span class="mi">20</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 20</p>
<p>reset? P</p>
<p>This defines the number of rows in the <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> array that is created.
As the capitalized reminder in the parameters file mentions, this must be
equal to the y-dimension of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>; otherwise, an error
will be thrown. Note that this parameter (as for the remaining parameters in
this section, other than the <strong>function</strong> parameter) is valid for the
default <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">mpd</span></span></code> function that is set by the
<strong>function</strong> parameter); if you are using a different <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code>
function to create this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> then this and the remaining parameters
must be changed to the parameters that that function needs,
and <em>only</em> those parameters (because they will be unpacked into that function,
i.e. passed on to it, at the time it is called).</p>
<hr class="docutils" />
<p><strong>nCol</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#number of cols (MUST EQUAL LAND DIMENSION x!)</span>
<span class="s1">&#39;nCol&#39;</span><span class="p">:</span>                     <span class="mi">20</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 20</p>
<p>reset? P</p>
<p>This defines the number of columns in the <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> array that is created.
As the capitalized reminder in the parameters file mentions, this must be
equal to the x-dimension of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>; otherwise, an error
will be thrown. Note that this parameter (as for the remaining parameters in
this section, other than the <strong>function</strong> parameter) is valid for the
default <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">mpd</span></span></code> function that is set by the
<strong>function</strong> parameter); if you are using a different <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code>
function to create this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> then this and the remaining parameters
must be changed to the parameters that that function needs,
and <em>only</em> those parameters (because they will be unpacked into that function,
i.e. passed on to it, at the time it is called).</p>
<hr class="docutils" />
<p><strong>h</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#level of spatial autocorrelation in element values</span>
<span class="s1">&#39;h&#39;</span><span class="p">:</span>                     <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the level of spatial autocorrelation in the element values
of the <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> array that is created.
Note that this parameter (and the remaining parameters in
this section, other than the <strong>function</strong> parameter) is valid for the
default <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code> function (<code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">mpd</span></span></code>, which is set by the
<strong>function</strong> parameter); but if you are using a different <code class="code py python docutils literal"><span class="name"><span class="pre">nlmpy</span></span></code>
function to create this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> then this and the remaining parameters
must be changed to the parameters that that function needs,
and <em>only</em> those parameters (because they will be unpacked into that function,
i.e. passed on to it, at the time it is called).</p>
</div>
</div>
<div class="section" id="change">
<h6>Change<a class="headerlink" href="#change" title="Permalink to this headline">¶</a></h6>
<hr class="docutils" />
<p><strong>end_rast</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#land-change event for this Layer</span>
<span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#end raster for event (DIM MUST EQUAL DIM OF LAND!)</span>
    <span class="s1">&#39;end_rast&#39;</span><span class="p">:</span>         <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">)),</span>
</pre></div>
</div>
<p>{2d <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">zeros</span></span><span class="punctuation"><span class="pre">((</span></span><span class="literal number integer"><span class="pre">20</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">20</span></span><span class="punctuation"><span class="pre">))</span></span></code></p>
<p>reset? Y</p>
<p>This defines the end raster for a <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>-change event (i.e. the array
this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> will change into over the course of the event). As the
capitalized reminder in the parameters file mentions, this raster must of
course have the same dimensions as the <cite>Landscape</cite> to which the <cite>Layer</cite> belongs;
otherwise, Genomics will throw an Error. Defaults a placeholder array
of zeros, so should be parameterized to meet your needs. Valid values are
a 2-d <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> object (stipulating the raster itself), or a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>
pointing to a file containing the raster (where valid files can be a ‘.txt’
file holding a 2-d <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> or any GIS raster files that can be
read by <code class="code py python docutils literal"><span class="name"><span class="pre">osgeo</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">gdal</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">Open</span></span></code>).</p>
<hr class="docutils" />
<p><strong>start_t</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#starting timestep of event</span>
<span class="s1">&#39;start_t&#39;</span><span class="p">:</span>          <span class="mi">49</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 49</p>
<p>reset? P</p>
<p>This indicates the timestep on which the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>-change event
will begin. Defaults to 49, but should be set to suit your
specific scenario.</p>
<hr class="docutils" />
<p><strong>end_t</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#ending timestep of event</span>
<span class="s1">&#39;end_t&#39;</span><span class="p">:</span>          <span class="mi">99</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 99</p>
<p>reset? P</p>
<p>This indicates the timestep on which the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>-change event
will finish. Defaults to 99, but should be set to suit your
specific scenario.</p>
<hr class="docutils" />
<p><strong>n_steps</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#number of stepwise changes in event</span>
<span class="s1">&#39;n_steps&#39;</span><span class="p">:</span>          <span class="mi">5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 5</p>
<p>reset? P</p>
<p>This indicates the number of stepwise changes to use to model a
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>-change event. The changes during a <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>-change event
are linearly interpolated (cellwise for the whole <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>) to this
number of discrete, instantaneous <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> changes between
the starting and ending rasters. Thus, the fewer the number of
steps, the larger, magnitudinally, each change will be. So generally, more
steps is ‘better’, as it will better approximate change that is continuous
in time. However, there is a potenitally significant memory trade-off here:
The whole series of stepwise-changed arrays is computed when the
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is created, then saved and used at the appropriate timestep
during each <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> run (and if the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that is changing is used
by any <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> as a <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> then each
intermediate <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is also calculated
when the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is first built.
These objects take up memory, which may be limiting for larger
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>s and/or <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> objects. This will probably not be a
major issue, but is worth considering.</p>
</div>
</div>
</div>
<div class="section" id="community-parameters">
<h4>Community parameters<a class="headerlink" href="#community-parameters" title="Permalink to this headline">¶</a></h4>
<div class="section" id="species">
<h5>Species<a class="headerlink" href="#species" title="Permalink to this headline">¶</a></h5>
<hr class="docutils" />
<p><strong>spp_&lt;n&gt;</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#spp name (SPECIES NAMES MUST BE UNIQUE!)</span>
<span class="s1">&#39;spp_0&#39;</span> <span class="p">:</span>   <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name"><span class="pre">spp_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>.
(Note that unlike most parameters, this parameter is
a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> key, the value for which is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
must have a unique name (so that a parameterized
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> isn’t overwritten in the <code class="code py python docutils literal"><span class="name"><span class="pre">ParametersDict</span></span></code> by a
second, identically-named <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> represents. Example valid values include: 0, ‘spp0’,
‘high-dispersal’, ‘C. fasciata’. Names default to
<code class="code py python docutils literal"><span class="name"><span class="pre">spp_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>, where n is a series of
integers starting from 0 and counting the number of <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>.</p>
<div class="section" id="id1">
<h6>Init<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h6>
<hr class="docutils" />
<p><strong>N</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#starting number of individs</span>
    <span class="s1">&#39;N&#39;</span><span class="p">:</span>                <span class="mi">250</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 250</p>
<p>reset? P</p>
<p>This defines the starting size of this <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. Importantly, this
may or may not be near the stationary size of the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> after
the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> has burned in, because that size will depend on the
carrying-capacity raster (set by the <strong>K</strong> parameter), and on
the dynamics of specific a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> (because of the interaction of
its various parameters).</p>
<hr class="docutils" />
<p><strong>K_layer</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#name of the carrying-capacity Layer</span>
<span class="s1">&#39;K_layer&#39;</span><span class="p">:</span>         <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: ‘layer_0’</p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> to be used as the
carrying-capacity raster for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. The values of this
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> should express the carrying capacity at each cell, in number
of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s. Note that the sum of the values of this <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>
can serve as a rough estimate of the expected stationary
number of individuals of a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>;
however, observed stationary size could vary
substantially depending on various other <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> parameters (e.g. birth
and death rates and mean number of offspring per mating event) as well
as on stochastic events (e.g. failure to colonize, or survive in, all
habitable portions of the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>).</p>
</div>
<div class="section" id="mating">
<h6>Mating<a class="headerlink" href="#mating" title="Permalink to this headline">¶</a></h6>
<hr class="docutils" />
<p><strong>repro_age</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="s1">&#39;mating&#39;</span>    <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#age(s) at sexual maturity (if tuple, female first)</span>
    <span class="s1">&#39;repro_age&#39;</span><span class="p">:</span>            <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="punctuation"><span class="pre">(</span></span><span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">)</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the age at which <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s in the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
can begin to reproduce. If the value provided is a 2-tuple of different
numbers (and the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> uses separate sexes), then the first
number will be used as females’ reproductive age, the second as males’.
If the value is 0, or <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s are capable
of reproduction from time of time.</p>
<hr class="docutils" />
<p><strong>sex</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to assign sexes</span>
<span class="s1">&#39;sex&#39;</span><span class="p">:</span>                  <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s will be assigned separate sexes
that are used to ensure only male-female mating events.</p>
<hr class="docutils" />
<p><strong>sex_ratio</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#ratio of males to females</span>
<span class="s1">&#39;sex_ratio&#39;</span><span class="p">:</span>            <span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1/1</p>
<p>reset? P</p>
<p>This defines the ratio of males to females (i.e. it will be converted to
a probability that an offspring is a male, which is used as the probability
of a Bernoulli draw of that offspring’s sex).</p>
<hr class="docutils" />
<p><strong>distweighted_birth</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether P(birth) should be weighted by parental dist</span>
<span class="s1">&#39;distweighted_birth&#39;</span><span class="p">:</span>  <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p>#NOTE: I WILL PROBABLY GET RID OF THIS PARAMETER…</p>
<hr class="docutils" />
<p><strong>R</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#intrinsic growth rate</span>
<span class="s1">&#39;R&#39;</span><span class="p">:</span>                    <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.5</p>
<p>reset? P</p>
<p>This defines a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ intrinsic growth rate, which is used
as the ‘R’ value in the spatialized logistic growth equation that
regulates population density (<span class="math">\(\frac{\mathrm{d}
N_{x,y}}{\mathrm{d}t}=rN_{x,y}(1-\frac{N_{x,y}}{K_{x,y}})\)</span>).</p>
<hr class="docutils" />
<p><strong>b</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#intrinsic birth rate (MUST BE 0&lt;=b&lt;=1)</span>
<span class="s1">&#39;b&#39;</span><span class="p">:</span>                    <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ intrinsic birth rate, which is
implemented as the probability that an identified potential mating
pair successfully produces offspring. Because this is a probability, as
the capitalized reminder in the parameters file mentions, this value must
be in the inclusive interval [0, 1].</p>
<p>NOTE: this may later need to be re-implemented to allow for spatial
variation in intrinsic rate (i.e.. expression of a birth-rate raster),
and/or for density-dependent birth as well as mortality</p>
<hr class="docutils" />
<p><strong>n_births_dist_lambda</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#expectation of distr of n offspring per mating pair</span>
<span class="s1">&#39;n_births_distr_lambda&#39;</span><span class="p">:</span>      <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the lambda parameter for the Poisson distribution from
which a mating pair’s number of offspring is drawn. Hence it is the
expected value for the number of offspring born in a
successful mating event.</p>
<hr class="docutils" />
<p><strong>mating_radius</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#radius of mate-search area</span>
<span class="s1">&#39;mating_radius&#39;</span><span class="p">:</span>        <span class="mi">1</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1</p>
<p>reset? Y</p>
<p>This defines the radius within which an <code class="code py python docutils literal"><span class="name"><span class="pre">Indvidual</span></span></code> can find a mate.
This radius is provided to queries run on the <code class="code py python docutils literal"><span class="name"><span class="pre">_KDTree</span></span></code> object.</p>
</div>
<div class="section" id="id2">
<h6>Mortality<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h6>
<hr class="docutils" />
<p><strong>max_age</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#maximum age</span>
<span class="s1">&#39;max_age&#39;</span><span class="p">:</span>              <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the maximum age an individual can achieve before being
forcibly culled from the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. Defaults to 1 (which will create
a Wright-Fisher-like simulation, with discrete generations). Can be set
to any other age, or can be set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code> (in which case no maxmimum
age is enforced).</p>
<hr class="docutils" />
<p><strong>d_min</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#min P(death) (MUST BE 0&lt;=d_min&lt;=1)</span>
<span class="s1">&#39;d_min&#39;</span><span class="p">:</span>                     <span class="mf">0.01</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0.01</p>
<p>reset? N</p>
<p>This defines the minimum probabilty of death that an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>
can face each time its Bernoulli death-decision is drawn. Because this
is a probability, as the capitalized reminder in
the parameters file mentions, this value must be in the
inclusive interval [0, 1].</p>
<hr class="docutils" />
<p><strong>d_max</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#max P(death) (MUST BE 0&lt;=d_max&lt;=1)</span>
<span class="s1">&#39;d_max&#39;</span><span class="p">:</span>                    <span class="mf">0.99</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0.99</p>
<p>reset? N</p>
<p>This defines the minimum probabilty of death that an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>
can face each time its Bernoulli death-decision is drawn. Because this
is a probability, as the capitalized reminder in
the parameters file mentions, this value must be in the
inclusive interval [0, 1].</p>
<hr class="docutils" />
<p><strong>density_grid_window_width</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="s1">&#39;mortality&#39;</span>     <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#width of window used to estimate local pop density</span>
    <span class="s1">&#39;dens_grid_window_width&#39;</span><span class="p">:</span>   <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: None</p>
<p>reset? N</p>
<p>This defines the width of the window used by the <code class="code py python docutils literal"><span class="name"><span class="pre">_DensityGridStack</span></span></code>
to estimate a raster of local <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> densities. The user should
feel free to set different values for this parameter (which could be
especially helpful when calling <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_density</span></span></code> to inspect the
resulting surfaces calculated at different window widths, if trying
to heuristically choose a reasonable value to set for a
particular simulation scenario). But be aware that choosing particularly
small window widths (in our experience, windows smaller than ~1/20th of
the larger <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> dimension) will cause dramatic increases in the
run-time of the density calculation (which runs twice per timestep).
Defaults to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, which internally will be set to 1/10th of the
larger <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> dimension; for many purposes this will work, but for
others the user may wish to control this.</p>
</div>
<div class="section" id="movement">
<h6>Movement<a class="headerlink" href="#movement" title="Permalink to this headline">¶</a></h6>
<hr class="docutils" />
<p><strong>direction_distr_mu</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="p">{</span>
     <span class="c1">#mode of distr of movement direction</span>
     <span class="s1">&#39;direction_distr_mu&#39;</span><span class="p">:</span>     <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, :py;`float`}</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This is the <span class="math">\(\mu\)</span> parameter of the VonMises distribution
(a circularized normal distribution) from which
movement directions are chosen when movement is random and isotropic
(rather than
being determined by a <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>;
if a <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>
is being usen this parameter is ignored). The <span class="math">\(\kappa\)</span> value
that is fed into this same distribution (<strong>direction_distr_kappa</strong>)
causes it to be very dispersed,
such that the distribution is effectively a uniform distribution on
the unit circle (i.e. all directions are effectively equally probable).
For this reason, changing this parameter without changing the
<strong>direction_distr_kappa</strong> value also, will make no change in the directions
drawn for movement.  If random, isotropic
movement is what you aim to model then there is probably little reason
to change these parameters.</p>
<p><strong>direction_distr_kappa</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#concentration of distr of movement direction</span>
<span class="s1">&#39;direction_distr_kappa&#39;</span><span class="p">:</span>  <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0</p>
<p>reset? N</p>
<p>This is the <span class="math">\(\kappa\)</span> parameter of the VonMises distribution
(a circularized normal distribution) from which
movement directions are chosen when movement is random and isotropic
(rather than
being determined by a <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>;
if a <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>
is being usen this parameter is ignored). The default value of 0 will
cause this distribution to be very dispersed, approximating a uniform
distribution on the unit circle and rendering the <span class="math">\(\mu\)</span>
value (<strong>direction_distr_mu</strong>) effectively meaningless. However, as this
parameter’s value increases the resulting circular distributions will become
more concentrated around <span class="math">\(\mu\)</span>, making the value fed to
<strong>direction_distr_mu</strong> influential. If random, isotropic
movement is what you aim to model then there is probably little reason
to change these parameters.</p>
<p><strong>distance_distr_mu</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of movement distance</span>
<span class="s1">&#39;distance_distr_mu&#39;</span><span class="p">:</span>      <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math">\(\mu\)</span> parameter of the Wald distribution used to draw
movement distances, expressed in units of raster-cells. This parameter and
<strong>distance_distr_sigma</strong> (the Wald distribution’s <span class="math">\(sigma\)</span>) should be
set to reflect a distribution of movement distances that is appropriate
for your scenario.</p>
<p><strong>distance_distr_sigma</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of movement distance</span>
<span class="s1">&#39;distance_distr_sigma&#39;</span><span class="p">:</span>   <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math">\(\sigma\)</span> parameter of the Wald distribution used to draw
movement distances, expressed in units of raster-cells. This parameter and
<strong>distance_distr_mu</strong> (the Wald distribution’s <span class="math">\(mu\)</span>) should be
set to reflect a distribution of movement distances that is appropriate
for your scenario.</p>
<p><strong>dispersal_distr_mu</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of dispersal distance</span>
<span class="s1">&#39;dispersal_distr_mu&#39;</span><span class="p">:</span>     <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math">\(\mu\)</span> parameter of the Wald distribution used to draw
dispersal distances, expressed in units of raster-cells. This parameter and
<strong>distance_distr_sigma</strong> (the Wald distribution’s <span class="math">\(sigma\)</span>) should be
set to reflect a distribution of dispersal distances that is appropriate
for your scenario.</p>
<p><strong>dispersal_distr_sigma</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of dispersal distance</span>
<span class="s1">&#39;dispersal_distr_sigma&#39;</span><span class="p">:</span>  <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math">\(\sigma\)</span> parameter of the Wald distribution used to draw
dispersal distances, expressed in units of raster-cells. This parameter and
<strong>distance_distr_mu</strong> (the Wald distribution’s <span class="math">\(mu\)</span>) should be
set to reflect a distribution of dispersal distances that is appropriate
for your scenario.</p>
<div class="section" id="movement-and-dispersal-conductancesurfaces">
<h7>Movement and Dispersal _ConductanceSurfaces<a class="headerlink" href="#movement-and-dispersal-conductancesurfaces" title="Permalink to this headline">¶</a></h7>
<p><strong>layer</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="s1">&#39;move_surf&#39;</span>     <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#move-surf Layer name</span>
    <span class="s1">&#39;layer&#39;</span><span class="p">:</span>                <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘layer_0’</span></span></code></p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> to be used as to construct the
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>. Note that this can also
be thought of as the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that should serve as a
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ permeability raster (because <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s moving
on this <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> toward the higher
(if mixture distributions are used) or highest
(if unimodl distributions are used) values in their neighborhoods).</p>
<p><strong>mixture</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to use mixture distrs</span>
<span class="s1">&#39;mixture&#39;</span><span class="p">:</span>              <span class="bp">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> should be built using
VonMises mixture distributions or unimodal VonMises distributions.
If True, each cell in the <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> will have an approximate
circular distribution that is a
weighted sum of 8 unimodal VonMises distributions (one per cell in the 8-cell
neighborhood); each of those summed unimodal distributions will have as its
mode the direction of the neighboring cell on which it is based and as its
weight the relative permeability of the cell on which it is based
(relative to the full neighborhood). If False, each cell in the
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> will have an approximated circular distribution
that is a single
VonMises distribution with its mode being the direction of the maximum-valued
cell in the 8-cell neighborhood and its concentration determined by
<strong>vm_distr_kappa</strong>.</p>
<p><strong>vm_distr_kappa</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#concentration of distrs</span>
<span class="s1">&#39;vm_distr_kappa&#39;</span><span class="p">:</span>       <span class="mi">12</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 12</p>
<p>reset? N</p>
<p>This sets the concentration of the VonMises distributions used to build
the approximated circular distributions in the <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>.
The default value was chosen heuristically as one that provides a reasonable
concentration in the direction of a unimodal VonMises distribution’s mode
without causing VonMises mixture distributions built from an
evenly weighted sum of distributions pointing toward the
8-cell-neighborhood directions to have 8 pronounced modes.
There will probably be little need to change the default value, but if
interested then the user could create <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>s with various values
of this parameter and then use the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_movement_surface</span></span></code>
method to explore the influence of the parameter on the resulting
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>s.</p>
<p><strong>approx_len</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#length of approximation vectors for distrs</span>
<span class="s1">&#39;approx_len&#39;</span><span class="p">:</span>       <span class="mi">5000</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 5000</p>
<p>reset? P</p>
<p>This determines the length of the vector of values used to approximate each
distribution on the <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> (i.e. the size of the z-axis
of the <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> used to hold all the distribution-approximations, where
the x and y axes have the same dimensions as the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>). The default
value of 5000 is fine for many cases, but may need to be
reduced depending on the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> dimensions (because for a larger
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, say 1000x1000 cells, it would create a
<code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> that is roughly 4Gb,
and if the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> on which the <code class="code py python docutils literal"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is based will be
undergoing landscape changes then numerous versions of an object of this size
would need to be generated when the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is built and held in memory).
The value to use for this parameter will depend on the size of the
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>, the exact scenario being simulated, and the memory of the
machine on which the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is to be run.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="genomicarchitecture">
<h6>_GenomicArchitecture<a class="headerlink" href="#genomicarchitecture" title="Permalink to this headline">¶</a></h6>
<p><strong>gen_arch_file</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="s1">&#39;gen_arch&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#/path/to/file.csv defining custom genomic arch</span>
    <span class="s1">&#39;gen_arch_file&#39;</span><span class="p">:</span>            <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: {<code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘&lt;your_model_name&gt;_spp-&lt;n&gt;_gen_arch.csv’</span></span></code></p>
<p>reset? P</p>
<p>This arguments indicates whether a custom genomic architecture file should
be used to create a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and if so,
where that file is located. If the value is <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, no file will be
used and the values of this <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’ other genomic
architecture parameters in the parameters file will be used to create
the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>. If the value is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code> pointing to a
custom genomic-architecture file
(i.e. a CSV file with loci as rows and ‘locus_num’,
‘p’, ‘dom’, ‘r’, ‘trait’, and ‘alpha’ as columns stipulating the starting
allele frequencies, dominance values, inter-locus recombination rates,
trait names, and effect sizes of all loci). Geonomics will create an empty
file of this format for each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> for which the
‘custom_genomic_architecture’ argument is provided the value True when
<code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> is called (which will be saved as
‘&lt;your_model_name&gt;_spp-&lt;n&gt;_gen_arch.csv’).</p>
<p>Note that when Geonomics reads in a custom genomic architecture file
to create a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, it will check
that the length (i.e. number of rows) in this file is equal to the length
stipulated by the <strong>L</strong> parameter, and will also check that the first value
at the top of the ‘r’ column is 0.5 (which is used to implement independent
assortment during gametogenesis). If either of these checks fails,
Geonomics throws an Error.</p>
<p><strong>L</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#num of loci</span>
<span class="s1">&#39;L&#39;</span><span class="p">:</span>                        <span class="mi">1000</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 1000</p>
<p>reset? P</p>
<p>This defines the total number of loci in the genomes in a
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>.</p>
<p><strong>l_c</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#num of chromosomes</span>
<span class="s1">&#39;l_c&#39;</span><span class="p">:</span>                      <span class="p">[</span><span class="mi">750</span><span class="p">,</span> <span class="mi">250</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">[</span></span><span class="literal number integer"><span class="pre">750</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal number integer"><span class="pre">250</span></span><span class="punctuation"><span class="pre">]</span></span></code></p>
<p>reset? P</p>
<p>This defines the lengths (in number of loci) of each of the chromosomes
in the genomes in a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>.  Note that the sum of this <code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code>
must equal <strong>L</strong>, otherwise Geonomics will throw an Error.
Also note that Geonomics models genomes as single <strong>L</strong> x 2
arrays, where separate chromosomes are delineated by points along
the genome where the recombination rate is 0.5;
thus, for a model where recombination rates are often at or near 0.5, this
parameter will have little meaning.</p>
<p><strong>mu_neut</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#genome-wide per-base neutral mut rate (0 to disable)</span>
<span class="s1">&#39;mu_neut&#39;</span><span class="p">:</span>                  <span class="mf">1e-9</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1e-9</p>
<p>reset? P</p>
<p>This defines the genome-wide per-base neutral mutation rate.
This value can be set to 0 to disable neutral mutation.</p>
<p><strong>mu_delet</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#genome-wide per-base deleterious mut rate (0 to disable)</span>
<span class="s1">&#39;mu_delet&#39;</span><span class="p">:</span>                 <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the genome-wide per-base deleterious mutation rate.
This value can be set to 0 to disable deleterious mutation. Note that all
deleterious mutation will fall outside the loci that affect any <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s
a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> may have, and will behave simply as globally
deleterious mutations (i.e. mutations that reduce the mutated
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s fitness regardless of that <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s
spatial location).</p>
<p><strong>delet_alpha_distr_shape</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#shape of distr of deleterious effect sizes</span>
<span class="s1">&#39;delet_alpha_distr_shape&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines the shape parameter of the gamma distribution from which
the effect sizes of deleterious loci are drawn. (Values drawn will be
truncated to the interval [0,1].)</p>
<p><strong>delet_alpha_distr_scale</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#scale of distr of deleterious effect sizes</span>
<span class="s1">&#39;delet_alpha_distr_scale&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines the scale parameter of the gamma distribution from which
the effect sizes of deleterious loci are drawn. (Values drawn will be
truncated to the interval [0,1].)</p>
<p><strong>r_distr_alpha</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#alpha of distr of recomb rates</span>
<span class="s1">&#39;r_distr_alpha&#39;</span><span class="p">:</span>            <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.5</p>
<p>reset? P</p>
<p>This defines the alpha parameter of the beta distribution from which
interlocus recombination rates are drawn. (Values drawn will be truncated to
the interval [0, 0.5].)</p>
<p><strong>r_distr_beta</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#beta of distr of recomb rates</span>
<span class="s1">&#39;r_distr_beta&#39;</span><span class="p">:</span>            <span class="mf">15e9</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 15e9</p>
<p>reset? P</p>
<p>This defines the beta parameter of the beta distribution from which
interlocus recombination rates are drawn. (Values drawn will be truncated to
the interval [0, 0.5].)</p>
<p><strong>dom</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether loci should be dominant (for allele &#39;1&#39;)</span>
<span class="s1">&#39;dom&#39;</span><span class="p">:</span>                      <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether loci should be treated as dominant (if True)
for the ‘1’ allele  or as codominant (if False). Codominance is the default
behavior, because it is assumed that Geonomics will often be used
to model quantitative traits, for which this is a reasonable assumption.</p>
<p><strong>pleiotropy</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to allow pleiotropy</span>
<span class="s1">&#39;pleiotropy&#39;</span><span class="p">:</span>               <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether pleiotropy should be allowed. If True, loci will be
permitted to contribute to more than one <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>.</p>
<p><strong>recomb_rate_custom_fn</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#custom fn for drawing recomb rates</span>
<span class="s1">&#39;recomb_rate_custom_fn&#39;</span><span class="p">:</span>    <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name"><span class="pre">function</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This parameter allows the user to provide a custom function according to which
interlocus recombination rates will be assigned. If set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, the
default behavior (i.e. recombination rates chosen from a beta distribution
using <strong>r_distr_alpha</strong> and <strong>r_distr_beta</strong>) will be used.</p>
<p><strong>n_recomb_paths_mem</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#number of recomb paths to hold in memory</span>
<span class="s1">&#39;n_recomb_paths_mem&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mf">1e4</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">(</span></span><span class="literal number float"><span class="pre">1e4</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset? P</p>
<p>This defines the maximum number of recombination paths for Genomics to hold in
memory at one time. Geonomics models recombination by using the interlocus
recombination rates to draw a large number of recombination ‘paths’
along the 2xL genome array (when the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is first built), and
then shuffling and cycling through those recombination paths as
needed during <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> runs. Of the total number of paths created, some
subset will be held in memory (the number of these is defined by
this parameter), while the remainder will live in a temporary
file (which is occasionally read in whenever the paths in memory are close to
being used up). Thus, to avoid problems, the number provided to this parameter
should be comfortably larger than the largest anticipated number of
recombination paths that will be needed during a single mating event (i.e.
larger than two times the largest antipicated number of offspring to be born
to the <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> during one timestep).</p>
<p><strong>n_recomb_paths_tot</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#total number of recomb paths to simulate</span>
<span class="s1">&#39;n_recomb_paths&#39;</span><span class="p">:</span>           <span class="nb">int</span><span class="p">(</span><span class="mf">1e5</span><span class="p">),</span>
</pre></div>
</div>
<p>This defines the total number of recombination paths that Geonomics will
generate. Geonomics models recombination by using the interlocus
recombination rates to draw a large number of recombination ‘paths’
along the 2xL genome array (when the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is first built), and
then shuffling and cycling through those recombination paths as
needed during <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> runs. The larger the total number of these paths
that is created, the more closely Geonomics will model truly
free recombination and the more prceisely it will model the exact
interlocus recombination rates defined in a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>.</p>
<p><strong>mut_log</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to save mutation logs</span>
<span class="s1">&#39;mut_log&#39;</span><span class="p">:</span>                  <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the location of the mutation-log file where  Geonomics should
save a record of each mutation that occurs for a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>
<code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, for each iteration. If <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, no mutation log
will be created and written to.</p>
<hr class="docutils" />
<div class="section" id="traits">
<h7>Traits<a class="headerlink" href="#traits" title="Permalink to this headline">¶</a></h7>
<p><strong>trait_&lt;n&gt;</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#trait name (TRAIT NAMES MUST BE UNIQUE!)</span>
<span class="s1">&#39;trait_0&#39;</span> <span class="p">:</span>   <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name"><span class="pre">trait_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>.
(Note that unlike most parameters, this parameter is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> key,
the value for which is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>
must have a unique name (so that a parameterized
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> isn’t overwritten in the <code class="code py python docutils literal"><span class="name"><span class="pre">ParametersDict</span></span></code> by a
second, identically-named <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> represents. Example valid values include: 0, ‘trait0’,
‘tmp_trait’, ‘bill length’. Names default to <code class="code py python docutils literal"><span class="name"><span class="pre">trait_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>,
where n is a series of integers starting from 0 and counting the
number of <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s for this <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>.</p>
<p><strong>layer</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#trait-selection Layer name</span>
<span class="s1">&#39;layer&#39;</span><span class="p">:</span>                <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘layer_0’</span></span></code></p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> that serves as the selective force
acting on this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>. (For example, if this Trait is selected upon by
annual mean temperature, then the name of the <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>
representing annual mean temperature should be provided here.)</p>
<p><strong>phi</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#polygenic selection coefficient</span>
<span class="s1">&#39;phi&#39;</span><span class="p">:</span>                  <span class="mf">0.05</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>s}</p>
<p>default: 0.05</p>
<p>reset? P</p>
<p>This defines the polygenic selection coefficient on this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> (i.e
the selection coefficient acting on the phenotypes, rather than the genotypes,
of this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>). The effect of this value can be thought of as the
reduction (from 1) in an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s survival probability when that
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> is maximally unfit (i.e. when that <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> has a
phenotypic value of 1.0 but is located in a location with an environmental
value of 0.0, or vice versa). When the value is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code> then the
strength of selection will be the same for all locations on the
<code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>. When the value is an <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> of
equal dimensions to the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> then the strength of
selection will vary across space, as indicated by the values in this array
(what Geonomics refers to as a “spatially contingent” selection regime).</p>
<p><strong>n_loci</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#number of loci underlying trait</span>
<span class="s1">&#39;n_loci&#39;</span><span class="p">:</span>               <span class="mi">10</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 10</p>
<p>reset? P</p>
<p>This defines the number of loci that should contribute to the phenotypes
of this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>. These loci will be randomly drawn from across the
genome.</p>
<p><strong>mu</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#mutation rate at loci underlying trait</span>
<span class="s1">&#39;mu&#39;</span><span class="p">:</span>                   <span class="mf">1e-9</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1e-9</p>
<p>reset? P</p>
<p>This defines the mutation rate for this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> (i.e. the rate at which
mutations that affect the phenotypes of this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> will arise). Set to
0 to disable mutation for this <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>.</p>
<p><strong>alpha_distr_mu</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of effect sizes</span>
<span class="s1">&#39;alpha_distr_mu&#39;</span> <span class="p">:</span>      <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0</p>
<p>reset? N</p>
<p>This defines the mean of the normal distribution from which a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>’s
new mutations’ effect sizes are drawn. The user will likely want to keep this
value set at 0 and adjust <strong>alpha_distr_sigma</strong>, because for multigenic
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s the baseline phenotypic value is 0.5 (the central value on a
Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>), so new mutations in a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> would then
be equally likely to decrease or increase <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ phenotypes from
that baseline.</p>
<p><strong>alpha_distr_sigma</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of effect size</span>
<span class="s1">&#39;alpha_distr_sigma&#39;</span><span class="p">:</span>    <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.5</p>
<p>reset? P</p>
<p>This defines the standard deviation of the normal distribution from which
a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>’s new mutations’ effect sizes are drawn. The user will likely
want to adjust this value and keep <strong>alpha_distr_mu</strong> set at 0, because
for multigenic <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s the baseline phenotypic value is 0.5
(the central value on a Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>), so new
mutations in a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> would then be equally likely to decrease
or increase <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ phenotypes from that baseline.</p>
<p><strong>gamma</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#curvature of fitness function</span>
<span class="s1">&#39;gamma&#39;</span><span class="p">:</span>                <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This defines the curvature of the fitness function (i.e.
how fitness decreases as the absolute difference between an
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s optimal and actual phenotypes increases). The user
will probably have no need to change this from the default value of 1
(which causes fitness to decrease linearly around the optimal
phenotypic value). Values &lt; 1 will cause the fitness function to be
concave up; values &gt; 1 will cause it to be concave down.</p>
<p><strong>univ_adv</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether the trait is universally advantageous</span>
<span class="s1">&#39;univ_adv&#39;</span><span class="p">:</span>             <span class="bp">False</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether whether selection on a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> should be
universal (i.e. whether a phenotype of 1 should be optimal everywhere
on the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>). When set to True, selection of the <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>
will be directional on the entire <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>, regardless
of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ spatial contexts.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="species-change">
<h6>Species change<a class="headerlink" href="#species-change" title="Permalink to this headline">¶</a></h6>
<div class="section" id="demographic-change">
<h7>Demographic change<a class="headerlink" href="#demographic-change" title="Permalink to this headline">¶</a></h7>
<p><strong>kind</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#kind of event {&#39;monotonic&#39;, &#39;stochastic&#39;,</span>
<span class="c1">#&#39;cyclical&#39;, &#39;custom&#39;}</span>
<span class="s1">&#39;kind&#39;</span><span class="p">:</span>             <span class="s1">&#39;monotonic&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘custom’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code></p>
<p>reset? P</p>
<p>This indicates what type of demographic change is being parameterized.
Each event has a certain length (in timesteps; defined by the <strong>start</strong> and
<strong>end</strong> parameters). Note that of the other parameters in this section, only
those that are necessary to parameterize the type of change event indicated
here will be used.</p>
<p>In <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code> change events, a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code>’
carrying capacity raster (K) is multiplied by a constant factor
(<strong>rate</strong>) at each timestep during the event.
In <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> change events, K fluctuates
around the baseline value (i.e. the K-raster at the time that the change event
begins) at each required timestep during the event (where the sizes of the
fluctuations are drawn from the distribution indicated by
<strong>distr</strong>, the floor and ceiling on those sizes are set by
<strong>size_range</strong>, and the required timesteps are determined by <strong>interval</strong>).
In <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code> change events, K undergoes a number (indicated
by <strong>n_cycles</strong>) of sinusoidal cycles between some minimum and maximum
values (indicated by <strong>size_range</strong>).
In <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘custom’</span></span></code> change events, the baseline K is multiplied by a series
of particular factors (defined by <strong>sizes</strong>) at a series of particular
timesteps (defined by <strong>timesteps</strong>).</p>
<p><strong>start</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#starting timestep</span>
<span class="s1">&#39;start&#39;</span><span class="p">:</span>            <span class="mi">49</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 49</p>
<p>reset? P</p>
<p>This indicates the timestep at which the demographic change event
should start.</p>
<p><strong>end</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#ending timestep</span>
<span class="s1">&#39;end&#39;</span><span class="p">:</span>            <span class="mi">99</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 99</p>
<p>reset? P</p>
<p>This indicates the timestep at which the demographic change event should end.</p>
<p><strong>rate</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#rate, for monotonic change</span>
<span class="s1">&#39;rate&#39;</span><span class="p">:</span>             <span class="mf">1.02</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1.02</p>
<p>reset? P</p>
<p>This indicates the rate at which a <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code> change event should occur.
At each timestep during the event, a new carrying capacity raster (K)
will be calculated by multiplying the previous step’s K by this factor.
Thus, values should be expressed relative to 1.0 indicating no change.</p>
<p><strong>interval</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#interval of changes, for stochastic change</span>
<span class="s1">&#39;interval&#39;</span><span class="p">:</span>         <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 1</p>
<p>reset? P</p>
<p>This indicates the interval at which fluctutations should occur during a
<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> change event (i.e. the number of timesteps to wait
between fluctuations).</p>
<p><strong>distr</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#distr, for stochastic change {&#39;uniform&#39;, &#39;normal&#39;}</span>
<span class="s1">&#39;distr&#39;</span><span class="p">:</span>            <span class="s1">&#39;uniform&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘uniform’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘normal’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘uniform’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the distribution from which to draw the sizes of
fluctuations in a <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> change event. Valid options are
<cite>‘uniform’</cite> and <cite>‘normal’</cite>.</p>
<p><strong>n_cycles</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#num cycles, for cyclical change</span>
<span class="s1">&#39;n_cycles&#39;</span><span class="p">:</span>         <span class="mi">10</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 10</p>
<p>reset? P</p>
<p>This indicates the number of cyclical fluctuations that should occur during
a <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code> change event.</p>
<p><strong>size_range</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#min &amp; max sizes, for stochastic &amp; cyclical change</span>
<span class="s1">&#39;size_range&#39;</span><span class="p">:</span>       <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">(</span></span><span class="literal number float"><span class="pre">0.5</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal number float"><span class="pre">1.5</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset? P</p>
<p>This defines the minimum and maximum sizes of fluctuations that can occur
during <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> and <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code> change events.</p>
<p><strong>timesteps</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#list of timesteps, for custom change</span>
<span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">90</span><span class="p">,</span> <span class="mi">95</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: [50, 90, 95]</p>
<p>reset? P</p>
<p>This defines the series of particular timesteps at which fluctutations should
occur during a <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘custom’</span></span></code> change event.</p>
<p><strong>sizes</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#list of sizes, for custom change</span>
<span class="s1">&#39;sizes&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: [2, 5, 0.5]</p>
<p>reset? P</p>
<p>This defines the series of particular fluctutations that should occur
during a <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘custom’</span></span></code> change event.</p>
</div>
<hr class="docutils" />
<div class="section" id="life-history-change">
<h7>Life-history change<a class="headerlink" href="#life-history-change" title="Permalink to this headline">¶</a></h7>
<p><strong>&lt;life_hist_param&gt;</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#life-history parameter to change</span>
<span class="s1">&#39;&lt;life_hist_param&gt;&#39;</span><span class="p">:</span> <span class="p">{</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘&lt;life_hist_param&gt;’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the life-history parameter to be changed by this life-history
change event. (Note that unlike most parameters, this parameter is
a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> key, the value for which is a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters controlling how the life-history parameter that is named
will change.)</p>
<p><strong>timesteps</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#list of timesteps</span>
<span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">[]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the timesteps at which the life-history parameter being changed
should change (to the values indicated by <strong>vals</strong>).</p>
<p><strong>vals</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#list of values</span>
<span class="s1">&#39;vals&#39;</span><span class="p">:</span>        <span class="p">[],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">[]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the values to which the life-history parameter being changed
should change (at the timesteps indicated by <strong>timesteps</strong>).</p>
</div>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="model-parameters">
<h4>Model parameters<a class="headerlink" href="#model-parameters" title="Permalink to this headline">¶</a></h4>
<div class="section" id="id3">
<h5>Main<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h5>
<p><strong>T</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#total Model runtime (in timesteps)</span>
<span class="s1">&#39;T&#39;</span><span class="p">:</span>            <span class="mi">100</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 100</p>
<p>reset? Y</p>
<p>This indicates the total number of timesteps for which the main portion of
a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> (i.e. the portion after the burn-in has completed) will be run
during each iteration.</p>
<p><strong>burn_T</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#min burn-in runtime (in timesteps)</span>
<span class="s1">&#39;burn_T&#39;</span><span class="p">:</span>       <span class="mi">30</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 30</p>
<p>reset? P</p>
<p>This indicates the minimum number of timesteps for which a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>’s
burn-in will run. (Note this is only a minimum because the test for
burn-in completion includes a check that at least this many timesteps have
elapsed, but also includes two statistical checks of stationarity of the
size of each <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code> in a <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code>.)</p>
<p><strong>seed</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#seed number</span>
<span class="s1">&#39;num&#39;</span><span class="p">:</span>          <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates whether or not to set the seeds of the random number
generators (by calling <code class="code py python docutils literal"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">random</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">seed</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">random</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">seed</span></span></code>)
before building and running a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>. If value is an integer, the seeds
will be set to that value. If value is <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>, seeds will not be set.</p>
</div>
<hr class="docutils" />
<div class="section" id="iterations">
<h5>Iterations<a class="headerlink" href="#iterations" title="Permalink to this headline">¶</a></h5>
<p><strong>num_iterations</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#num iterations</span>
<span class="s1">&#39;n_its&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 3</p>
<p>reset? Y</p>
<p>This indicates the number of iterations for which the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>
should be run. (Note that for each iteration a separate subdirectory of
data and stats will be written, if your <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> has parameterized data
and stats to be collected.)</p>
<p><strong>rand_landscape</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to randomize Landscape each iteration</span>
<span class="s1">&#39;rand_landscape&#39;</span><span class="p">:</span>    <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> should be randomized for each
iteration. If True, a new <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> will be generated at the start
of each iteration. If False, the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> from iteration 0 will be
saved and reused for each subsequent iteration.</p>
<p><strong>rand_community</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to randomize Community each iteration</span>
<span class="s1">&#39;rand_comm&#39;</span><span class="p">:</span>    <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> should be randomized for each
iteration. If True, a new <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> will be generated at the start
of each iteration. If False, the <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> from iteration 0 will be
saved and reused for each subsequent iteration (and whether that
<code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> is saved before or after being burned in will depend on
the value provided to the <strong>repeat_burn</strong> parameter).</p>
<p><strong>repeat_burn</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to burn in each iteration</span>
<span class="s1">&#39;repeat_burn&#39;</span><span class="p">:</span>  <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether a reused <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> should be burned in
separately for each iteration for which it is reused. If True, the
<code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> from iteration 0 will be saved as soon as its instantiated,
but will have a new burn-in run for each iteration in which it is used. If
False, the <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> from iteration 0 will be saved after its burn-in
is complete, and then will only have the main portion of its <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> run
separately during each iteration. (Note that if <strong>rand_community</strong> is set to True then
the value of this parameter will not be used.)</p>
<hr class="docutils" />
<div class="section" id="data">
<h6>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h6>
<hr class="docutils" />
<div class="section" id="sampling">
<h7>Sampling<a class="headerlink" href="#sampling" title="Permalink to this headline">¶</a></h7>
<p><strong>scheme</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#sampling scheme {&#39;all&#39;, &#39;random&#39;, &#39;point&#39;, &#39;transect&#39;}</span>
<span class="s1">&#39;scheme&#39;</span><span class="p">:</span>               <span class="s1">&#39;random&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘all’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘random’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘random’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the sampling scheme to use when collecting data from a
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>. Currently valid values include <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘all’</span></span></code>,
<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘random’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code>, and <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>.</p>
<p>With <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘all’</span></span></code>, data will be collected for all <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
at each sampling timestep. With <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘random’</span></span></code>, data will be collected from a
random sample of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s (of size indicated by parameter <strong>n</strong>)
from anywhere on the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code>.
With <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code>, data will be collected from random samples of size <strong>n</strong>
within a certain distance (<strong>radius</strong>) of each of a set of particular
points (<strong>points</strong>). With <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>, a linear transect of some
number of points (<strong>n_transect_points</strong>) between some endpoints
(<strong>transect_endpoints</strong>) will be created, and then data will be collected
from random samples of size <strong>n</strong> with a certain distance (<strong>radius</strong>)
of each point along the transect.</p>
<p><strong>n</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#sample size at each point, for point &amp; transect sampling</span>
<span class="s1">&#39;n&#39;</span><span class="p">:</span>                    <span class="mi">250</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 250</p>
<p>reset? P</p>
<p>This indicates the total number of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s to sample each time
data is collected (if <strong>scheme</strong> is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘random’</span></span></code>), or the number of
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s to sample around each one of a set of points (if <strong>scheme</strong>
is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code> or <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>). This parameter will only be used if
<strong>scheme</strong> is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘random’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code>, or <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>; otherwise
it may be set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.</p>
<p><strong>points</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#coords of collection points, for point sampling</span>
<span class="s1">&#39;points&#39;</span><span class="p">:</span>               <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code> of 2-<code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code>s, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the points around which to sample <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s for data
collection. This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code>;
otherwise it may be set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.</p>
<p><strong>transect_endpoints</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#coords of transect endpoints, for transect sampling</span>
<span class="s1">&#39;transect_endpoints&#39;</span><span class="p">:</span>   <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{2-<code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code> of 2-<code class="code py python docutils literal"><span class="name builtin"><span class="pre">tuple</span></span></code>s, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the endpoints between which to create a transect, along which
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s will be sampled for data collection.
This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>;
otherwise it may be set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.</p>
<p><strong>n_transect_points</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#num points along transect, for transect sampling</span>
<span class="s1">&#39;n_transect_points&#39;</span><span class="p">:</span>    <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the number of points to create on the transect along which
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s will be sampled for data collection.
This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>;
otherwise it may be set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.</p>
<p><strong>radius</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#collection radius around points, for point &amp; transect sampling</span>
<span class="s1">&#39;radius&#39;</span><span class="p">:</span>               <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the radius around sampling points within which
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s may be sampled for data collection.
This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘point’</span></span></code> or
<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘transect’</span></span></code>; otherwise it may be set to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.</p>
<p><strong>when</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#when to collect data</span>
<span class="s1">&#39;when&#39;</span><span class="p">:</span>                 <span class="bp">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>s, <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the timesteps during main <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> iterations
at which data should be collected (in addition to after the final timestep
of each iteration, when data is always collected for any <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> for which
data collection is parameterized). If value is a non-zero <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>,
it will be treated as a frequency at which data should be collected (e.g.
a value of 5 will cause data to be collected every 5 timesteps). If value
is a list of <code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code>s, they will be treated as the particular timesteps
at which data should be collected. If value is 0 or <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>,
data will be collected only after the final timestep.</p>
<p><strong>include_landscape</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to save current Layers when data is collected</span>
<span class="s1">&#39;include_landscape&#39;</span><span class="p">:</span>         <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether to include the <code class="code py python docutils literal"><span class="name"><span class="pre">Landscape</span></span></code> <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code>s among the
data that is collected. If True, each <code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> will be written to a raster
or array file (according to the format indicated by
<strong>geo_rast_format</strong>) each time data is collected.</p>
<p><strong>include_fixed_sites</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to include fixed loci in VCF files</span>
<span class="s1">&#39;include_fixed_sites&#39;</span><span class="p">:</span>  <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether fixed sites (i.e. loci which are fixed for either the
0 or 1 allele) should be included in any VCF files that are written. Thus,
this parameter is only relevant if <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘vcf’</span></span></code> is one of the genetic data
formats indicated by <strong>gen_format</strong>.</p>
</div>
<hr class="docutils" />
<div class="section" id="format">
<h7>Format<a class="headerlink" href="#format" title="Permalink to this headline">¶</a></h7>
<p><strong>gen_format</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#format for genetic data {&#39;vcf&#39;, &#39;fasta&#39;}</span>
<span class="s1">&#39;gen_format&#39;</span><span class="p">:</span>           <span class="p">[</span><span class="s1">&#39;vcf&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">],</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘vcf’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘fasta’</span></span></code>, <code class="code py python docutils literal"><span class="punctuation"><span class="pre">[</span></span><span class="literal string single"><span class="pre">‘vcf’</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal string single"><span class="pre">‘fasta’</span></span><span class="punctuation"><span class="pre">]</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="punctuation"><span class="pre">[</span></span><span class="literal string single"><span class="pre">‘vcf’</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal string single"><span class="pre">‘fasta’</span></span><span class="punctuation"><span class="pre">]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format or formats to use for writing genetic data.
data. Currently valid formats include <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘vcf’</span></span></code> and <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘fasta’</span></span></code> formats.
Either or both formats may be specified; all formats that are specified will
be written each time data is collected.</p>
<p><strong>geo_vect_format</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#format for vector geodata {&#39;csv&#39;, &#39;shapefile&#39;, &#39;geojson&#39;}</span>
<span class="s1">&#39;geo_vect_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;csv&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘csv’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘shapefile’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘geojson’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘csv’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format to use for writing geographic vector data (i.e.
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ point locations).
Currently valid formats include <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘csv’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘shapefile’</span></span></code>,
and <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘geojson’</span></span></code>. Any one format may be specified.</p>
<p><strong>geo_rast_format</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#format for raster geodata {&#39;geotiff&#39;, &#39;txt&#39;}</span>
<span class="s1">&#39;geo_rast_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;geotiff&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘geotiff’</span></span></code>, <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘txt’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘geotiff’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format to use for writing geographic raster data (i.e.
<code class="code py python docutils literal"><span class="name"><span class="pre">Layer</span></span></code> arrays). Currently valid formats include <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘geotiff’</span></span></code>
and <code class="code py python docutils literal"><span class="literal string single"><span class="pre">‘txt’</span></span></code>. Either format may be specified. Note that this parameter
will only be used if the <strong>include_landscape</strong> parameter is set to True.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="stats">
<h6>Stats<a class="headerlink" href="#stats" title="Permalink to this headline">¶</a></h6>
<p>The stats parameters section has subsection for each statistc that Geonomics
can calculate. (Currently valid statistics include:</p>
<blockquote>
<div><ul class="simple">
<li><em>‘Nt’</em>: number of individuals at timestep t</li>
<li><em>‘het’</em>: heterozygosity</li>
<li><em>‘maf’</em>: minor allele frequency</li>
<li><em>‘mean_fit’</em>: mean fitness of a <code class="code py python docutils literal"><span class="name"><span class="pre">Species</span></span></code></li>
<li><em>‘ld’</em>: linkage disequilibrium</li>
</ul>
<p>There are only a few parameters, which are shared across all</p>
</div></blockquote>
<p>of those subsections, and each parameter always means the same thing despite
which statistic it is parameterizing. Thus, hereafter we provide a single of
each of those parameters are how it works, regardless of the statistic for
which it used:</p>
<p><strong>calc</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to calculate</span>
<span class="s1">&#39;calc&#39;</span><span class="p">:</span>     <span class="bp">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: (varies by statistic)</p>
<p>reset? P</p>
<p>This indicates whether or not a given statistic should be calculated. Thus,
only those statistics whose <strong>calc</strong> parameters are set to True will be
calculated and saved when their <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is run.</p>
<p><strong>freq</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#calculation frequency (in timesteps)</span>
<span class="s1">&#39;freq&#39;</span><span class="p">:</span>     <span class="mi">5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: (varies by statistic)</p>
<p>reset? P</p>
<p>This indicates the frequency with which a given statistic should be calculated
during each iteration (in timesteps). If set to 0, Geonomics will calculate
and save this statistic for only the first and last timesteps
of each iteration.</p>
<p><strong>mean</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#whether to mean across sampled individs</span>
<span class="s1">&#39;mean&#39;</span><span class="p">:</span> <span class="bp">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: (varies by statistic, and only valid for certain statistics)</p>
<p>reset? P</p>
<p>For some statistics that produce a vector of values each timestep
when they are collected (containing one value per <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>),
such as heterozygosity, this indicates
whether those values should instead be meaned and saved as a
single value for each timestep.</p>
</div>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="class-and-function-docstrings">
<h3>Class and function docstrings<a class="headerlink" href="#class-and-function-docstrings" title="Permalink to this headline">¶</a></h3>
<div class="section" id="gnx-make-parameters-file">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code><a class="headerlink" href="#gnx-make-parameters-file" title="Permalink to this headline">¶</a></h4>
<p>Create a new parameters file.</p>
<p>Write to disk a new, template parameters file. The file will contain the
numbers and types of sections indicated by the parameters fed to this
function. It can often be used ‘out of the box’ to make a new Model
object, but typically it will be edited by the user to stipulate
the scenario being simulated, then used to instantiate a Model.</p>
<div class="section" id="id4">
<h5>Parameters<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>filepath <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd>Where to write the resulting parameters file, in /path/to/filename.py
format. Defaults to None. If None, a file named
“GEONOMICS_params_&lt;datetime&gt;.py” will be written to the working
directory.</dd>
<dt>scapes <span class="classifier-delimiter">:</span> <span class="classifier">{int, list of dicts}, optional</span></dt>
<dd><p class="first">Number (and optionally, types) of Layer-parameter sections to include
in the parameters file that is generated. Defaults to 1. Valid values
and their associated behaviors are:</p>
<dl class="last docutils">
<dt>int:</dt>
<dd><p class="first">Add sections for the stipulated number of Layers, each with default
settings:</p>
<blockquote class="last">
<div><ul class="simple">
<li>parameters for creating Layers of type ‘random’ (i.e.
Layers that will be generated by interpolation from
randomly valued random points)</li>
<li>no LayerChanger parameters</li>
</ul>
</div></blockquote>
</dd>
<dt>[dict, …, dict]:</dt>
<dd><p class="first">Each dict in this list should be of the form:</p>
<p>{‘type’:    ‘random’, ‘defined’, ‘file’, or ‘nlmpy’,</p>
<p>‘change’:   bool</p>
<p>}</p>
<p class="last">This will add one section of Layer parameters, with the
contents indicated, for each dict in this list.</p>
</dd>
</dl>
</dd>
<dt>species <span class="classifier-delimiter">:</span> <span class="classifier">{int, list of dicts}, optional</span></dt>
<dd><p class="first">Number (and optionally, types) of Species-parameter sections to
include in the parameters file that is generated. Defaults to 1. Valid
values and their associated behaviors are:</p>
<dl class="last docutils">
<dt>int:</dt>
<dd><p class="first">Add sections for the stipulated number of Species, each with
default settings:</p>
<blockquote class="last">
<div><ul class="simple">
<li>parameters for movement without a MovementSurface</li>
<li>parameters for a GenomicArchitecture with 0 Traits (i.e. with
only neutral loci)</li>
<li>no SpeciesChanger parameters</li>
</ul>
</div></blockquote>
</dd>
<dt>[dict, …, dict]:</dt>
<dd>Each dict should contain at least one argument from among the
following:
{‘movement’:                       bool,
‘movement_surface’:                bool,
‘genomes’:                         bool,
‘n_traits’:                        int,
‘custom_genomic_architecture’:     bool,
‘demographic_change’:              int,
‘parameter_change’:                bool
}
This will add one section of Species parameters, customized
as indicated, for each dict in the list.</dd>
</dl>
</dd>
<dt>data <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first">Whether to include a Data-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="last docutils">
<dt>None, False:</dt>
<dd>Will not add a section for parameterizing data to be collected.
No DataCollector will be created for the Model object made from
the resulting parameters file, and no data will be collected
during the model runs.</dd>
<dt>True:</dt>
<dd>Will add a section that can be used to parameterize which
data will be collected during the model runs, when, and what
file formats will be used to write it to disk.
(This which will be managed by the model’s DataCollector
object.)</dd>
</dl>
</dd>
<dt>stats <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first">Whether to include a Stats-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="last docutils">
<dt>None, False:</dt>
<dd>Will not add a section for parameterizing the statistics to be
calculated. No StatsCollector will be created for the Model
object made from the resulting parameters file, and no
statistics will be calculated during the model runs.</dd>
<dt>True:</dt>
<dd>Will add a section that can be used to parameterize which
statistics will be calculated during the model runs, and when.
(This will be managed by the model’s StatsCollector object.)</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="returns">
<h5>Returns<a class="headerlink" href="#returns" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>out <span class="classifier-delimiter">:</span> <span class="classifier">None</span></dt>
<dd>Returns no output. Resulting parameters file will be written to the
location and filename indicated (or by default, will be written to a
file named “GEONOMICS_params_&lt;datetime&gt;.py” in the working directory).</dd>
</dl>
</div>
<div class="section" id="see-also">
<h5>See Also<a class="headerlink" href="#see-also" title="Permalink to this headline">¶</a></h5>
<p>sim.params.make_parameters_file</p>
</div>
<div class="section" id="notes">
<h5>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h5>
<p>All parameters of this function are optional. Calling the function without
providing any parameters will always produce the parameters file for the
default model scenario. This file can be instantiated as a Model object and
run without being edited. Those three steps (create default parameters file;
create model from that parameters file; run the model) serve as a base case
to test successful package installation, and are wrapped around by the
convenience function <cite>gnx.run_default_model</cite>.</p>
</div>
<div class="section" id="examples">
<h5>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h5>
<p>In the simplest example, we can create a parameters file for the default
model. Then (assuming it is the only Geonomics parameters file in the
current working directory, so that it can be unambiguously identified) we
can call the gnx.make_model function to create a Model object from that
file, and then call the Model.run method to run the model (setting the
‘verbose’ parameter to True, so that we can observe model output).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">TODO: PUT TYPICAL MODEL OUTPUT HERE, EVEN THOUGH IT&#39;S ONLY PRINTED?</span>
</pre></div>
</div>
<p>We can use some of the function’s arguments, to create a parameters
file for a model with 3 Layers and 1 Species (all with the default
components for their sections of the parameters file) and with a section
for parameterizing data collection.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span><span class="n">scapes</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>As a more complex example that is likely to be similar to most use cases,
we can create a parameters file for a model scenario with:</p>
<blockquote>
<div><ul class="simple">
<li>2 Layers (one being an nlmpy Layer that will not change over model
time, the other being a raster read in from a GIS file and being
subject to change over model time);</li>
<li>2 Species (the first having genomes, 2 Traits, and movement
that is dictated by a MovementSurface; the second not having
genomes but having a MovementSurface as well, and undergoing
demographic change)</li>
<li>data-collection;</li>
<li>stats-collection;</li>
</ul>
</div></blockquote>
<p>We can save this to a file named “2-spp_2-trait_model.py” in our current
working directory.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Layer&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">scapes</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;nlmpy&#39;</span><span class="p">},</span>                              <span class="c1">#scape 1</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;gis&#39;</span><span class="p">,</span>                                 <span class="c1">#scape 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Species&#39;</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">species</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>                               <span class="c1">#spp 1</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;n_traits&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>                              <span class="c1">#spp 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;demographic_change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#arguments to the data and stats parameters</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#destination to which to write the resulting parameter file</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">filepath</span> <span class="o">=</span> <span class="s1">&#39;2-spp_2-trait_model.py&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="read-parameters-file">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">read_parameters_file</span></span></code><a class="headerlink" href="#read-parameters-file" title="Permalink to this headline">¶</a></h4>
<p>Create a new ParametersDict object.</p>
<p>Read the Geonomics parameters file saved at the location indicated by
‘filepath’, check its validity (i.e. that all the Layers and Species
parameterized in that file have been given distinct names), then use the
file to instantiate a ParametersDict object.</p>
<div class="section" id="id5">
<h5>Parameters<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>filepath <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>String indicating the location of the Geonomics parameters file that
should be made into a ParametersDict object.</dd>
</dl>
</div>
<div class="section" id="id6">
<h5>Returns<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h5>
<p>An object of the ParametersDict class (a dict of nested dicts, all
of which have key-value pairs whose values can be accessed using typical
dict notation or using dot notation with the keys).</p>
</div>
<div class="section" id="raises">
<h5>Raises<a class="headerlink" href="#raises" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>AssertionError</dt>
<dd>If either the Layers or the Species parameterized in the parameters
file have not all been given distinct names</dd>
</dl>
</div>
<div class="section" id="id7">
<h5>See Also<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h5>
<p>sim.params.read
sim.params.ParametersDict</p>
</div>
<div class="section" id="id8">
<h5>Examples<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h5>
<p>Read a parameters file called “null_model.py” (located in the current
working directory).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">read_parameters_file</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.params.ParametersDict&#39;&gt;</span>
<span class="go">Model name:                                     GEONOMICS_params_13-10-2018_15:54:03</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="make-model">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">make_model</span></span></code><a class="headerlink" href="#make-model" title="Permalink to this headline">¶</a></h4>
<p>Create a new Model object.</p>
<p>Use either a ParametersDict object or the path to a valid Geonomics
parameters file (whichever is provided to the ‘parameters’ argument) to
create a new Model object.</p>
<div class="section" id="id9">
<h5>Parameters<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>parameters <span class="classifier-delimiter">:</span> <span class="classifier">{ParametersDict, str}, optional</span></dt>
<dd>The parameters to be used to make the Model object.
If <cite>parameters</cite> is a ParametersDict object, the object will be used to
make the Model.
If <cite>parameters</cite> is a string, Geonomics will call
<cite>gnx.read_parameters_file</cite> to make a ParametersDict object, then use
that object to make the Model.
If <cite>parameters</cite> is None, or is not provided, then Geonomics will
attempt to find a single parameters file in the current working
directory with the filename “GEONOMICS_params_&lt;…&gt;.py”, will use that
file to make a ParametersDict object, then will use that object to
make the Model.</dd>
</dl>
</div>
<div class="section" id="id10">
<h5>Returns<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>out <span class="classifier-delimiter">:</span> <span class="classifier">Model</span></dt>
<dd>An object of the Model class</dd>
</dl>
</div>
<div class="section" id="id11">
<h5>Raises<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>ValueError</dt>
<dd>If the <cite>parameters</cite> argument was not provided and a single, valid
Geonomics parameters file could not be identified in the current
working directory</dd>
<dt>ValueError</dt>
<dd>If the <cite>parameters</cite> arugment was given a string that does not point
to a valid parameters file</dd>
<dt>ValueError</dt>
<dd>If the ParametersDict provided to the <cite>parameters</cite> argument, or created
from the parameters file being used, cannot be successfully made into a
Model</dd>
</dl>
</div>
<div class="section" id="id12">
<h5>See Also<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h5>
<p>gnx.read_parameters_file
sim.model.Model</p>
</div>
<div class="section" id="id13">
<h5>Examples<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h5>
<p>Make a Model from a single, valid “GEONOMICS_params_&lt;…&gt;.py” file that can
be found in the current working directory (such as a file that would be
produced by calling gnx.make_parameters_file without any arguments).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     GEONOMICS_params_13-10-2018_15:54:03</span>
<span class="go">Layers:                                         0: &#39;0&#39;</span>
<span class="go">Species:                                        0: &#39;0&#39;</span>
<span class="go">Number of iterations:                           1</span>
<span class="go">Number of burn-in timesteps (minimum):          30</span>
<span class="go">Number of main timesteps:                       100</span>
<span class="go">Geo-data collected:                             {}</span>
<span class="go">Gen-data collected:                             {}</span>
<span class="go">Stats collected:                                {}</span>
</pre></div>
</div>
<p>Make a Model from a file called ‘null_model.py’, in the current working
directory.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
<span class="go">Layers:                                         0: &#39;tmp&#39;</span>
<span class="go">                                                1: &#39;ppt&#39;</span>
<span class="go">Species:                                        0: &#39;C. fasciata&#39;</span>
<span class="go">Number of iterations:                           2500</span>
<span class="go">Number of burn-in timesteps (mininum):          100</span>
<span class="go">Number of main timesteps:                       1000</span>
<span class="go">Geo-data collected:                             {csv, geotiff}</span>
<span class="go">Gen-data collected:                             {vcf, fasta}</span>
<span class="go">Stats collected:                                {maf, ld, mean_fit, het, Nt}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Drew Hart.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.7</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.8</a>
      
      |
      <a href="_sources/doc.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>