<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>geonomics.sim package &mdash; geonomics 1.4.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="geonomics.structs package" href="geonomics.structs.html" />
    <link rel="prev" title="geonomics.ops package" href="geonomics.ops.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> geonomics
            <img src="_static/gnx_mini.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.4
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro_video.html">Intro Video</a></li>
<li class="toctree-l1"><a class="reference internal" href="procedural_diagram.html">Workflow Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="motivation.html">Motivation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="conceptual_diagram.html">Conceptual Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="params.html">Parameters</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="geonomics.html">geonomics package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="geonomics.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="geonomics.help.html">geonomics.help package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.ops.html">geonomics.ops package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">geonomics.sim package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.structs.html">geonomics.structs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.utils.html">geonomics.utils package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="geonomics.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="geonomics.html#module-geonomics.main">geonomics.main module</a></li>
<li class="toctree-l3"><a class="reference internal" href="geonomics.html#module-geonomics">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">geonomics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="modules.html">API</a> &raquo;</li>
          <li><a href="geonomics.html">geonomics package</a> &raquo;</li>
      <li>geonomics.sim package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/geonomics.sim.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="geonomics-sim-package">
<h1>geonomics.sim package<a class="headerlink" href="#geonomics-sim-package" title="Permalink to this headline"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
</section>
<section id="module-geonomics.sim.burnin">
<span id="geonomics-sim-burnin-module"></span><h2>geonomics.sim.burnin module<a class="headerlink" href="#module-geonomics.sim.burnin" title="Permalink to this headline"></a></h2>
<p>Functions for testing burn-in stationarity.</p>
<dl class="py class">
<dt class="sig sig-object py" id="geonomics.sim.burnin.SpatialTester">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.sim.burnin.</span></span><span class="sig-name descname"><span class="pre">SpatialTester</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.burnin.SpatialTester" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.burnin.SpatialTester.plot_diffs">
<span class="sig-name descname"><span class="pre">plot_diffs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.burnin.SpatialTester.plot_diffs" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.burnin.SpatialTester.run_test">
<span class="sig-name descname"><span class="pre">run_test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_timesteps_back</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.burnin.SpatialTester.run_test" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.burnin.SpatialTester.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.burnin.SpatialTester.update" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.burnin.SpatialTester.update_counts_cell">
<span class="sig-name descname"><span class="pre">update_counts_cell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.burnin.SpatialTester.update_counts_cell" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-geonomics.sim.data">
<span id="geonomics-sim-data-module"></span><h2>geonomics.sim.data module<a class="headerlink" href="#module-geonomics.sim.data" title="Permalink to this headline"></a></h2>
<p>Classes and functions for implementing data collection and output</p>
</section>
<section id="module-geonomics.sim.model">
<span id="geonomics-sim-model-module"></span><h2>geonomics.sim.model module<a class="headerlink" href="#module-geonomics.sim.model" title="Permalink to this headline"></a></h2>
<p>Defines the core Model class, as well as its public and private methods</p>
<dl class="py class">
<dt class="sig sig-object py" id="geonomics.sim.model.Model">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.sim.model.</span></span><span class="sig-name descname"><span class="pre">Model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Hierarchical data structure containing a Landscape and a Community as
attributes, and serving as an interface for all of the core simulation
functionality (as methods to run simulations and visualize results).
The principal Geonomics data structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – A character string to be assigned as the Model’s name.</p></li>
<li><p><strong>params</strong> (<a class="reference internal" href="#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object containing all of the parameter values needed
instantiate the Model. (See the online documentation for a detailed
discussion of all parameters.)</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not to display information about the Model’s status
during simulations. Defaults to False.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.add_individuals">
<span class="sig-name descname"><span class="pre">add_individuals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coords</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recip_spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_msprime_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.add_individuals" title="Permalink to this definition"></a></dt>
<dd><p>add individuals to a recipient species object, either using a second
species object as the source population or feeding a dict of parameters
into msprime to simulate the source population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em>) – the number of individuals to be added.</p></li>
<li><p><strong>coords</strong> (<em>{tuple</em><em>, </em><em>list</em><em>, </em><em>numpy.ndarray}</em>) – a tuple, list of numpy.ndarray indicating the x,y coordinates where the
new individuals should be added to the recipient species.
if <cite>coords</cite> contains 2 values then all added individuals will be placed
at the x,y coordinate pair indicated by those 2 values.
otherwise, <cite>coords</cite> may be of size <cite>n</cite> x 2, indicating the x,y
coordinate pair at which each of the <cite>n</cite> individuals will be placed.</p></li>
<li><p><strong>recip_spp</strong> (<em>{gnx.structs.species.species</em><em>, </em><em>int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default:</em>) – None
an int or str referring to the species in this model to which
individuals should be added.</p></li>
<li><p><strong>source_spp</strong> (<em>{gnx.structs.species.species</em><em>, </em><em>int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default:</em>) – None
a geonomics.structs.species.species object to act as the source
population from which individuals will be taken
(or an int or str referring to that species within this model).</p></li>
<li><p><strong>source_msprime_params</strong> (<em>dict</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – a dict of keyword args to be fed to <cite>gnx.sim_msprime_individs()</cite>,
to parameterize the msprime source population model from which
new individuals will be drawn. these include:
- recomb_rate         (required)
- mut_rate            (required)
- demography          (optional; default: None)
- population_size     (optional; default: None)
- ancestry_model      (optional; default: None)
- random_seed         (optional; default: None)
see <cite>gnx.sim_msprime_individs</cite> for details.</p></li>
<li><p><strong>individs</strong> (<em>{tuple</em><em>, </em><em>list</em><em>, </em><em>numpy.ndarray}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – a 1d numpy.ndarray or equivalent array-like containing the int indices
of the individuals to be taken from <cite>source_spp</cite> and added to
<cite>recip_spp</cite>. If None, Indivduals from <cite>source_spp</cite> are randomly
chosen.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>alters <cite>recip_spp</cite> in place by adding the provided
(if <cite>source_spp</cite> is not None) or simulated (if <cite>source_msprime_params</cite>
is not None) individuals.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.burn">
<span class="sig-name descname"><span class="pre">burn</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.burn" title="Permalink to this definition"></a></dt>
<dd><p>Helper function to burn a model in, if it hasn’t been burned in already.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>out</strong> – Returns no output. Alters Model in place by running burn-in phase.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.calc_stats">
<span class="sig-name descname"><span class="pre">calc_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.calc_stats" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_cells">
<span class="sig-name descname"><span class="pre">get_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_cells" title="Permalink to this definition"></a></dt>
<dd><p>returns the cell coordinates of any or all individuals in a species</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose cell coordinates should be returned.
if None, all individuals’ cell coordinates will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>returns an nx2 numpy array, where n is the number of individuals
whose coordinates were requested, and the 2 columns correspond to
the i then j (i.e. lat then lon) array coordinates.</p>
<p>note: ccolumns are in the reverse order of the columns
of the array returned by <cite>model.get_coords()</cite>, to facilitate
array-subsetting of the landscape layer raster arrays.</p>
<p>note: rows are sorted by index number of the individuals,
regardless of input index order.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_coords">
<span class="sig-name descname"><span class="pre">get_coords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_coords" title="Permalink to this definition"></a></dt>
<dd><p>returns the x,y coordinates of any or all individuals in a species</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose coordinates should be returned.
if None, all individuals’ coordinates will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>np.array</em> – returns an nx2 numpy array, where n is the number of individuals
whose coordinates were requested, and the 2 columns correspond to
the x then y (i.e. lon then lat) coordinates.</p></li>
<li><p><strong>note</strong> (<em>rows are sorted by index number of the individuals,</em>) – regardless of input index order.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_e">
<span class="sig-name descname"><span class="pre">get_e</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_e" title="Permalink to this definition"></a></dt>
<dd><p>returns the current environmental values of any or all
individuals in a species, for one or all landscape layers</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose environmental values should be returned.
if None, all individuals’ environmental values  will be returned.</p></li>
<li><p><strong>lyr_num</strong> (<em>int</em><em>, </em><em>default:None</em>) – if provided, indicates the layer for which values should be
returned. if None, values will be returned for all layers.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>returns an nxl numpy array, where n is the number of individuals
whose coordinates were requested, and l is the number of layers
requested, ordered by increasing layer number.</p>
<p>note: rows are sorted by index number of the individuals,
regardless of input index order.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_fitness">
<span class="sig-name descname"><span class="pre">get_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_fitness" title="Permalink to this definition"></a></dt>
<dd><p>returns the current fitness of any or all individuals in a species,
for one or all traits</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose fitness should be returned.
if None, all individuals’ fitness values will be returned.</p></li>
<li><p><strong>trt</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the trait for which the gea should be run.
can be either the trait’s index number (i.e. its integer key in the
genomicarchitecture’s traits dict), or its name (as a character
string).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>returns an nxt numpy array, where n is the number of individuals
whose coordinates were requested, and t is the number of traits
requested.</p>
<p>note: rows are sorted by index number of the individuals,
regardless of input index order.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_genotypes">
<span class="sig-name descname"><span class="pre">get_genotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loci</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biallelic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_genotypes" title="Permalink to this definition"></a></dt>
<dd><p>returns the genotypes of any or all individuals in a species
and for any or all loci</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>loci</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the loci for which genotypes should be
returned. if None, genotypes will be returned for all loci.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the only individuals
for whom genotypes should be returned.
if None, all individuals’ genotypes will be returned.</p></li>
<li><p><strong>biallelic</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – if False, results are returned as nxl numpy arrays, where n is the
number of individuals requested and l is the number of loci
requested.
if True, results are returned as nxlx2 numpy arrays, where the 2
dimensions on the third axis represent the two haploid genotypes of
each diploid individual.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>returns a numpy array containing the requested genotypes.
the shape of the array depends on the number of individuals and
loci and the return format requested.
the first and second axes represent the individuals and loci,
returned in sorted numerical order. the optional third axis
represents ploidy, if biallelic genotypes are requested.
(see explanation of the argument ‘biallelic’, above,
for more detail.)</p>
<p>note: all axes are sorted by index number of the object they
represent (individuals, loci), regardless of input index order.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_random_individs">
<span class="sig-name descname"><span class="pre">get_random_individs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_random_individs" title="Permalink to this definition"></a></dt>
<dd><p>returns a size-n sample of index numbers
for random individuals in a species.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em>) – the number of random individuals to sample</p></li>
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns a length-n 1d numpy array containing the sorted index
numbers of the sampled individuals.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_tree_sequence">
<span class="sig-name descname"><span class="pre">get_tree_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_tree_sequence" title="Permalink to this definition"></a></dt>
<dd><p>returns the tskit.treesequence object for the indicated species</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_x">
<span class="sig-name descname"><span class="pre">get_x</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_x" title="Permalink to this definition"></a></dt>
<dd><p>returns the x coordinates of any or all individuals in a species</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose coordinates should be returned.
if None, all individuals’ coordinates will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>np.array</em> – returns an nx1 numpy array, where n is the number of individuals
whose coordinates were requested.</p></li>
<li><p><strong>note</strong> (<em>rows are sorted by index number of the individuals,</em>) – regardless of input index order.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_y">
<span class="sig-name descname"><span class="pre">get_y</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_y" title="Permalink to this definition"></a></dt>
<dd><p>returns the y coordinates of any or all individuals in a species</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose coordinates should be returned.
if None, all individuals’ coordinates will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>np.array</em> – returns an nx1 numpy array, where n is the number of individuals
whose coordinates were requested.</p></li>
<li><p><strong>note</strong> (<em>rows are sorted by index number of the individuals,</em>) – regardless of input index order.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.get_z">
<span class="sig-name descname"><span class="pre">get_z</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.get_z" title="Permalink to this definition"></a></dt>
<dd><p>returns the phenotypic values of any or all individuals in a species,
for one or all traits</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – if provided, indicates the integer indices of the individuals
whose phenotypes should be returned.
if None, all individuals’ phenotypes will be returned.</p></li>
<li><p><strong>trt</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the trait for which the gea should be run.
can be either the trait’s index number (i.e. its integer key in the
genomicarchitecture’s traits dict), or its name (as a character
string).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>returns an nxt numpy array, where n is the number of individuals
whose coordinates were requested, and t is the number of traits
requested.</p>
<p>note: rows are sorted by index number of the individuals,
regardless of input index order.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_land</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'face'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">land_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">animate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot" title="Permalink to this definition"></a></dt>
<dd><p>Create a basic plot of the Model in its current state.</p>
<p>Plot the current state of the model, with any number of overlain
Landscape Layers, and with or without points representing the
Individuals in a Species.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Species whose Individuals should be scattered
on the plot. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string). If None, will cause only the Landscape to be plotted.</p></li>
<li><p><strong>lyr</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Layer whose raster should be plotted. Can be
either the Layer’s index number (i.e. its integer key in the
Landscape dict), or its name (as a character string). Defaults
to None, which will cause all Layers to be plotted as an overlay
of transparent rasters, each with a different colormap.</p></li>
<li><p><strong>hide_land</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, the Landscape will be ‘hidden’, i.e. the plot will only
show a scatter of the plotted Individuals. Defaults to False.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If provided, indicates the integer indices of the only Individuals
to be plotted. If None, all Individuals will be plotted.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s index number will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'black'</em>) – Face color for the points in the Individual scatter. Passed to the
<em>c</em> argument of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>edge_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'face'</em>) – Edge color for the points in the Individual scatter. If ‘face’,
will always match the face color (i.e. the color provided to
<em>color</em>). Passed to the <em>edgecolor</em> argument of
matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>default 'black'</em>) – Color for the plotted text. (Will only be used if the <em>text</em>
argument is True.) Passed to the <em>color</em> argument of
matplotlib.pyplot.text.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the Landscape’s environmental values.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 25</em>) – Size of the scatter points. Passed to the <em>s</em> argument
of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_size</strong> (<em>{size in points</em><em>, </em><em>valid string}</em><em>, </em><em>optional</em><em>, </em><em>default: 9</em>) – Text size. (Will only be used if <em>text</em> is True.) Can be expressed
as a numeric size in points, or as any string that is a valid size
argument for matplotlib.text.Text.</p></li>
<li><p><strong>land_cmap</strong> (<em>{valid string</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Colormap to use for plotting Layer rasters. If None, will default
to the colormap automatically assigned to the Layer by Geonomics
based on the Layer’s index number in the Landscape dict. Can be
passed any string that references a colormap in
matplotlib.pyplot.cm.</p></li>
<li><p><strong>pt_cmap</strong> (<em>{valid string</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Colormap to use for plotting points. (Will only be used if the
Individual scatter is being plotted, and if the value provided to
<em>color</em> is an ordered collection of values to which to map the
Individuals’ colors.) If None, no colormap will be used.
Can be passed any string that references a colormap in
matplotlib.pyplot.cm, as well as an
matplotlib.colors.LinearSegmentedColormap instance.</p></li>
<li><p><strong>alpha</strong> (<em>{scalar</em><em>, </em><em>bool</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The transparency level of the points. If scalar is passed, it is
passed on to the <em>alpha</em> value fed to matplotlib.pyplot.scatter.
If True, alpha value will be set to 0.6. If False or None, points
will be opaque.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>vmin</strong> (<em>{scalar}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Used to normalize the luminance data that determine how colors
are colormapped onto Individuals’ point colors. Passed to the
<em>vmin</em> and <em>vmax</em> arguments of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>vmax</strong> (<em>{scalar}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Used to normalize the luminance data that determine how colors
are colormapped onto Individuals’ point colors. Passed to the
<em>vmin</em> and <em>vmax</em> arguments of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
<li><p><strong>animate</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, this function will return a list of the Individuals’
points, as a matplotlib.collections.PathCollection. This is used
internally by the Model.walk method, to create animated simulation
plots.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output (unless <em>animate</em> is passed True, in which case
a PathCollection of the Individuals’ points is returned, for
internal use by Model.walk, to create animated simulations)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>{None, <code class="xref py py-class docutils literal notranslate"><span class="pre">matplotlib.collections.PathCollection</span></code>}</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_allele_frequencies">
<span class="sig-name descname"><span class="pre">plot_allele_frequencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_allele_frequencies" title="Permalink to this definition"></a></dt>
<dd><p>Plot a Species’ current allele frequencies for all alleles</p>
<p>For the Species indicated, create a plot showing, across all loci,
the current allele frequencies (in the indicated color) versus the starting
frequencies (in dashed black).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose allele frequencies will be
plotted. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>color</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: 'red'</em>) – The color to use for plotting the current allele frequencies.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_demographic_pyramid">
<span class="sig-name descname"><span class="pre">plot_demographic_pyramid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_demographic_pyramid" title="Permalink to this definition"></a></dt>
<dd><p>plot a demographic pyramid for the chosen species</p>
<p>plot a paired, horizontal bar plot of the chosen species’
count of individuals of each age.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species whose demographic pyramid will be
plotted. can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_density">
<span class="sig-name descname"><span class="pre">plot_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'face'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_density" title="Permalink to this definition"></a></dt>
<dd><p>Plot the population-density raster for a given Species.</p>
<p>Plot the population density of a Species, as estimated by the Species’
_DensityGridStack.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose Individuals should be scattered
on the plot. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>normalize</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, the population-density values of the output raster will be
normalized to 0 &lt;= density &lt;= 1. Otherwise, actual estimates
(individuals per cell) are returned.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If provided, indicates the integer indices of the only Individuals
to be plotted. If None, all Individuals will be plotted.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s index number will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'black'</em>) – Face color for the points in the Individual scatter. Passed to the
<em>c</em> argument of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>edge_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'face'</em>) – Edge color for the points in the Individual scatter. If ‘face’,
will always match the face color (i.e. the color provided to
<em>color</em>). Passed to the <em>edgecolor</em> argument of
matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>default 'black'</em>) – Color for the plotted text. (Will only be used if the <em>text</em>
argument is True.) Passed to the <em>color</em> argument of
matplotlib.pyplot.text.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 25</em>) – Size of the scatter points. Passed to the <em>s</em> argument
of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_size</strong> (<em>{size in points</em><em>, </em><em>valid string}</em><em>, </em><em>optional</em><em>, </em><em>default: 9</em>) – Text size. (Will only be used if <em>text</em> is True.) Can be expressed
as a numeric size in points, or as any string that is a valid size
argument for matplotlib.text.Text.</p></li>
<li><p><strong>alpha</strong> (<em>{scalar</em><em>, </em><em>bool</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The transparency level of the points. If scalar is passed, it is
passed on to the <em>alpha</em> value fed to matplotlib.pyplot.scatter.
If True, alpha value will be set to 0.6. If False or None, points
will be opaque.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_dispersal">
<span class="sig-name descname"><span class="pre">plot_dispersal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_dispersals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_distr_param1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_distr_param2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_distr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lognormal'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">land_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_by_individ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increasing_linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_start_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_dispersal" title="Permalink to this definition"></a></dt>
<dd><blockquote>
<div><p>plot example dispersal events for a number of
individuals of the indicated species and using the indicated parameter
values. note that directions will be drawn using whatever directional
behavior is parameterized for the given species (i.e., from a surface,
or simply from a universally parameterized von mises distribution).</p>
<dl class="simple">
<dt>spp<span class="classifier">{int, str}, optional, default: 0</span></dt><dd><p>a reference to the species for which dispersal should be plotted.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string). If None, the first species will be plotted to be plotted.</p>
</dd>
<dt>lyr<span class="classifier">{int, str}, optional, default: None</span></dt><dd><p>a reference to the layer whose raster should be plotted. can be
either the layer’s index number (i.e. its integer key in the
landscape dict), or its name (as a character string). defaults
to None, in which case the first layer will be plotted.</p>
</dd>
<dt>n_dispersals<span class="classifier">int</span></dt><dd><p>the number of independent dispersals to plot for each individual.</p>
</dd>
<dt>n_individs<span class="classifier">{int, None}, default: None</span></dt><dd><p>the number of individuals whose dispersal should be
plotted. defaults to None, in which case all individuals will be
plotted.</p>
</dd>
</dl>
<p>distance_distr_param1,
distance_distr_param2 : float, optional default: None</p>
<blockquote>
<div><p>these are the parameters defining the distributions from which the
direction and distance values for individual dispersals are drawn. if
passed, their values will be used to determine the dispersal behavior
that is plotted (even if a parametersdict was passed to <em>params</em>,
because these values will override any values therein). for more
details, see the ‘parameters’ section of the online documentation.</p>
</div></blockquote>
<dl class="simple">
<dt>distance_distr: str, optional default: ‘lognormal’</dt><dd><p>this indicates the distribution to use for drawing dispersal
distancs. defaults to ‘lognormal’, but can also be set to ‘wald’ or
‘levy’.</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>color<span class="classifier">valid mpl.plt color value, optional, default: ‘black’</span></dt><dd><blockquote>
<div><p>face color for the dipseral lines.</p>
</div></blockquote>
<dl class="simple">
<dt>alpha<span class="classifier">valid mpl.plt alpha value, optional, default: None</span></dt><dd><p>transparency value for the dispersal lines.</p>
</dd>
<dt>size<span class="classifier">scalar or array_like, optional, default: 10</span></dt><dd><p>size of the scatter points.</p>
</dd>
<dt>land_cmap<span class="classifier">{valid string, None}, optional, default: ‘plasma’</span></dt><dd><p>colormap to use for plotting layer rasters. If None, will default
to the colormap automatically assigned to the layer by geonomics
based on the layer’s index number in the landscape dict. can be
passed any string that references a colormap in matplotlib.pyplot.cm.</p>
</dd>
<dt>ticks<span class="classifier">bool, optional, default: False</span></dt><dd><p>If True, x- and y-axis ticks will be added to the plot.</p>
</dd>
<dt>color_by_individ<span class="classifier">bool, optional, default: True</span></dt><dd><p>If True, each dispersal line for each plotted individual will have its own
distinct color, making it easier to follow the track and distinguish it
from others in the plot.</p>
</dd>
<dt>include_start_points<span class="classifier">bool, optional, default: False</span></dt><dd><p>If True, a scatter of points will be plotted to signify the starting
points from which each individual is dispersing.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_dispersal_surface">
<span class="sig-name descname"><span class="pre">plot_dispersal_surface</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hist'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_fact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_dispersal_surface" title="Permalink to this definition"></a></dt>
<dd><p>Create any of 4 types of plots to summarize a Species’ DispersalSurface</p>
<p>Plot a Species’ DispersalSurface with either circularized histograms,
a circle of directional draws, or a vector on each cell, to visualize
the von Mises distributions of each cell on the MovementSurface.
Or plot a plain histogram of the von Mises distribution at a given
cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose DispersalSurface should be
plotted. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>style</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: 'hist'</em>) – <dl class="simple">
<dt>Determines the style of plot produced. There are four options:</dt><dd><dl class="simple">
<dt>’hist’:</dt><dd><p>Plot a classic histogram approximating the von Mises
distribution at the cell indicated by position x,y.
(Note that <em>x</em> and <em>y</em> must be provided for this style.)</p>
</dd>
<dt>’chist’:</dt><dd><p>Plot a circular histogram approximating the von Mises
distribution at the cell indicated by position x,y;
plot will be drawn inside the chosen cell on the
DispersalSurface raster.</p>
</dd>
<dt>’cdraws’:</dt><dd><p>Plot points on the unit circle, whose locations were
drawn from the von Mises distribution at the cell indicated
by position x,y; plot will be drawn inside the chosen cell
on the DispersalSurface raster.</p>
</dd>
<dt>’vect’:</dt><dd><p>Inside each cell of the DispersalSurface raster,
plot the mean direction vector of directions drawn
from that cell’s von Mises distribution.</p>
</dd>
</dl>
</dd>
</dl>
</p></li>
<li><p><strong>color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'black'</em>) – Color to use for the plot elements. Passed to the
<em>c</em> argument of matplotlib.pyplot.hist, matplotlib.pyplot.plot,
matplotlib.pyplot.scatter, or matplotlib.pyplot.arrow,
depending on <em>style</em>.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the MovementSurface’s environmental values.</p></li>
<li><p><strong>cmap</strong> (<em>{valid string</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: 'plasma'</em>) – Colormap to use for plotting the DispersalSurface. If None,
will default to ‘plasma’, a common colormap used to plot resistance
surfaces.Can be passed any string that references a colormap in
matplotlib.pyplot.cm.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Not valid for Species without a DisersalSurface</p></li>
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_example_recombinant_genome">
<span class="sig-name descname"><span class="pre">plot_example_recombinant_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_example_recombinant_genome" title="Permalink to this definition"></a></dt>
<dd><p>plot an example recombinant genome for the chosen species</p>
<p>plot a vertical image depicting an example recombinant genome for the
chosen species. each gamete (i.e. vertical half) is colored using a
different colormap, and for each one alternating bands of color
indicate breakpoints where recombination occurred between the two
parental chromatids (or haploid chromatid sets, for multi-chromosome
models). the resulting plot can be useful for introspecting or
exploring a model, because each plot will be unique, and will be a
result of the array of interlocus recombination rates defined in the
species’ genomicarchitecture.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which an example recombinant genome
should be plotted. can be either the species’ index number
(i.e. its integer key in the community dict), or its name
(as a character string).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>not valid for species without genomes.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_fitness">
<span class="sig-name descname"><span class="pre">plot_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gray'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_fitness" title="Permalink to this definition"></a></dt>
<dd><p>Plot a Species’ Individuals, colored by fitness</p>
<p>Create a scatter plot of the Individuals in a Species,
on top of any Landscape Layer, and color the points by
their fitnesses. Fitness values can be either overall fitness
(i.e. calculated across all traits), or can be fitnesses for
a certain Trait.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose Individuals should be scattered
on the plot. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string). If None, will cause only the Landscape
to be plotted.</p></li>
<li><p><strong>trt</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Trait for which fitness should be calculated
and used to color Individuals in the plot. Can be either the
Trait’s index number (i.e. its integer key in the
GenomicArchitecture’s traits dict), or its name (as a character
string). If None, Individuals’ overall fitnesses will be
calculated instead.</p></li>
<li><p><strong>lyr</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Layer whose raster should be plotted. Can be
either the Layer’s index number (i.e. its integer key in the
Landscape dict), or its name (as a character string). Defaults
to None, which will either cause only the Layer associated with the
chosen Trait to be plotted, if <em>trt</em> is not None; or else
will cause all Layers to be plotted as an overlay of transparent
rasters, each with a different colormap, if <em>trt</em> is None.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If provided, indicates the integer indices of the only Individuals
to be plotted. If None, all Individuals will be plotted.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s index number will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>phenotype_text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s phenotype will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>fitness_text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s calculated fitness value will be
displayed next to it. Can be useful for model introspection.</p></li>
<li><p><strong>edge_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'face'</em>) – Edge color for the points in the Individual scatter. If ‘face’,
will always match the face color (i.e. the color provided to
<em>color</em>). Passed to the <em>edgecolor</em> argument of
matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>default 'black'</em>) – Color for the plotted text. (Will only be used if the <em>text</em>
argument is True.) Passed to the <em>color</em> argument of
matplotlib.pyplot.text.</p></li>
<li><p><strong>phenotype_text_color</strong> (<em>valid mpl.plt color value</em><em>,</em>) – </p></li>
<li><p><strong>fitness_text_color</strong> (<em>valid mpl.plt color value</em><em>,</em>) – </p></li>
<li><p><strong>'black'</strong> (<em>default</em>) – Color for the plotted phenotype-value or fitness-value text.
(Will only be used if the <em>phenotype_text</em> or <em>fitness_text</em>
arguments are True.) Passed to the <em>color</em> argument
of matplotlib.pyplot.text.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the Landscape’s environmental values.</p></li>
<li><p><strong>fitness_cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the Landscape’s environmental values.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 25</em>) – Size of the scatter points. Passed to the <em>s</em> argument
of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_size</strong> (<em>{size in points</em><em>, </em><em>valid string}</em><em>, </em><em>optional</em><em>, </em><em>default: 9</em>) – Text size. (Will only be used if <em>text</em> is True.) Can be expressed
as a numeric size in points, or as any string that is a valid size
argument for matplotlib.text.Text.</p></li>
<li><p><strong>fit_cmap</strong> (<em>{valid string</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: 'gray'</em>) – Colormap to use for color Individuals by their fitness values.
If no value passed, will defaul to grayscale. Can be passed any
string that references a colormap in matplotlib.pyplot.cm.</p></li>
<li><p><strong>alpha</strong> (<em>{scalar</em><em>, </em><em>bool</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The transparency level of the points. If scalar is passed, it is
passed on to the <em>alpha</em> value fed to matplotlib.pyplot.scatter.
If True, alpha value will be set to 0.6. If False or None, points
will be opaque.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_genetic_PCA">
<span class="sig-name descname"><span class="pre">plot_genetic_PCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_land</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_genetic_PCA" title="Permalink to this definition"></a></dt>
<dd><p>Create a scatter plot of the Individuals in a Species,
on top of any Landscape Layer, and color the points by
the first 3 PCS in a genetic PCA (Principal Components Analysis).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose Individuals should be scattered
on the plot. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string). If None, will cause only the Landscape
to be plotted.</p></li>
<li><p><strong>n_pcs</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default: 3</em>) – The number of top PCs to use for the plot. If 1 is used, all three
red, green, and blue (RGB) channels will be scaled the same,
creating a grayscale mapping of the top PC. If 2 are used, the red
and blue channels only will be used, creating a visualization of
the top 2 PCs that is colorblind-friendly. If 3 are used, the top 3
PCs will be fed into all RGB channels, creating a visualization
with the most information but which is not colorblind-friendly.</p></li>
<li><p><strong>lyr</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Layer whose raster should be plotted. Can be
either the Layer’s index number (i.e. its integer key in the
Landscape dict), or its name (as a character string). Defaults
to None, which will either cause only the Layer associated with the
chosen Trait to be plotted, if <em>trt</em> is not None; or else
will cause all Layers to be plotted as an overlay of transparent
rasters, each with a different colormap, if <em>trt</em> is None.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If provided, indicates the integer indices of the only Individuals
to be plotted. If None, all Individuals will be plotted.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s index number will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>edge_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'face'</em>) – Edge color for the points in the Individual scatter. If ‘face’,
will always match the face color (i.e. the color provided to
<em>color</em>). Passed to the <em>edgecolor</em> argument of
matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>default 'black'</em>) – Color for the plotted text. (Will only be used if the <em>text</em>
argument is True.) Passed to the <em>color</em> argument of
matplotlib.pyplot.text.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the Landscape’s environmental values.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 25</em>) – Size of the scatter points. Passed to the <em>s</em> argument
of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_size</strong> (<em>{size in points</em><em>, </em><em>valid string}</em><em>, </em><em>optional</em><em>, </em><em>default: 9</em>) – Text size. (Will only be used if <em>text</em> is True.) Can be expressed
as a numeric size in points, or as any string that is a valid size
argument for matplotlib.text.Text.</p></li>
<li><p><strong>alpha</strong> (<em>{scalar</em><em>, </em><em>bool</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The transparency level of the points. If scalar is passed, it is
passed on to the <em>alpha</em> value fed to matplotlib.pyplot.scatter.
If True, alpha value will be set to 0.6. If False or None, points
will be opaque.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>hide_land</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, Landscape Layer raster will not be plotted beneath the
scatterplot.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_genotype">
<span class="sig-name descname"><span class="pre">plot_genotype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_dominance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_genotype" title="Permalink to this definition"></a></dt>
<dd><p>Plot a Species’ Individuals, colored by their genotypes for a locus</p>
<p>Create a scatter plot of the Individuals in a Species,
on top of any Landscape Layer, and color the points by
their genotypes for the chosen locus.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose Individuals should be scattered
on the plot. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>locus</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – The index of the locus for which Individuals’ genotypes should be
depicted. Must be 0 &lt;= val &lt;= (genome length - 1)</p></li>
<li><p><strong>lyr</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Layer whose raster should be plotted. Can be
either the Layer’s index number (i.e. its integer key in the
Landscape dict), or its name (as a character string). Defaults
to None, which will cause all Layers to be plotted as an overlay
of transparent rasters, each with a different colormap.</p></li>
<li><p><strong>by_dominance</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If False, the Individuals will be colored by their actual genotypes
(i.e. a 0|0 homozygote will be depicted as 0, a 1|1 homozygote as
1, and a 0|1 heterozygote as 0.5). If True, then the dominance
value for the chosen locus will be used to the determine the
‘effective genotype’, which will instead be displayed as an
Individual’s color. In this case, 0|1 heterozygotes
will be be depicted as 0.5’s for loci with codominance (i.e.
dominance values of 0), but
will be depicted as 1’s for loci with 1-allele dominance (i.e.
dominance values of 1). The dominance values for all loci in the
genome are set in a Species’ GenomicArchitecture object. For more
detail, see the “Genomes, GenomicArchitecture, and Traits”
subsection of the “Data Structures” section of the online
documentation.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If provided, indicates the integer indices of the only Individuals
to be plotted. If None, all Individuals will be plotted.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s index number will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>edge_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'face'</em>) – Edge color for the points in the Individual scatter. If ‘face’,
will always match the face color (i.e. the color provided to
<em>color</em>). Passed to the <em>edgecolor</em> argument of
matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>default 'black'</em>) – Color for the plotted text. (Will only be used if the <em>text</em>
argument is True.) Passed to the <em>color</em> argument of
matplotlib.pyplot.text.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the Landscape’s environmental values.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 25</em>) – Size of the scatter points. Passed to the <em>s</em> argument
of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_size</strong> (<em>{size in points</em><em>, </em><em>valid string}</em><em>, </em><em>optional</em><em>, </em><em>default: 9</em>) – Text size. (Will only be used if <em>text</em> is True.) Can be expressed
as a numeric size in points, or as any string that is a valid size
argument for matplotlib.text.Text.</p></li>
<li><p><strong>alpha</strong> (<em>{scalar</em><em>, </em><em>bool</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The transparency level of the points. If scalar is passed, it is
passed on to the <em>alpha</em> value fed to matplotlib.pyplot.scatter.
If True, alpha value will be set to 0.6. If False or None, points
will be opaque.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_hist_fitness">
<span class="sig-name descname"><span class="pre">plot_hist_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_hist_fitness" title="Permalink to this definition"></a></dt>
<dd><p>Plot a Species</p>
<p>For the Species indicated, create a histogram of Individuals’ current
fitness values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose Individuals’ fitness will be
plotted. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_movement">
<span class="sig-name descname"><span class="pre">plot_movement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_timesteps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_distr_param1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_distr_param2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_distr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lognormal'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">land_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_by_individ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increasing_linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_start_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_movement" title="Permalink to this definition"></a></dt>
<dd><p>plot example movement tracks for a number of
individuals of the indicated species and using the indicated parameter
values. note that directions will be drawn using whatever directional
behavior is parameterized for the given species (i.e., from a surface,
or simply from a universally parameterized von mises distribution).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which movement should be plotted.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string). if None, the first species will be plotted to be plotted.</p></li>
<li><p><strong>lyr</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – a reference to the layer whose raster should be plotted. can be
either the layer’s index number (i.e. its integer key in the
landscape dict), or its name (as a character string). defaults
to None, in which case the first layer will be plotted.</p></li>
<li><p><strong>n_timesteps</strong> (<em>int</em>) – a number of timesteps (i.e. individual movement steps) to be plotted
in each individual’s movement track.</p></li>
<li><p><strong>n_individs</strong> (<em>{int</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – the number of individuals whose movement should be
plotted. defaults to None, in which case all individuals will be
plotted.</p></li>
<li><p><strong>distance_distr_param1</strong> – </p></li>
</ul>
</dd>
</dl>
<p>:param :
:param distance_distr_param2: these are the parameters defining the distributions from which the</p>
<blockquote>
<div><p>direction and distance values for individual movements are drawn. if
passed, their values will be used to determine the movement behavior
that is plotted (even if a parametersdict was passed to <em>params</em>,
because these values will override any values therein). for more
details, see the ‘parameters’ section of the online documentation.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>distance_distr</strong> (<em>str</em><em>, </em><em>optional default: 'lognormal'</em>) – this indicates the distribution to use for drawing movement
distancs. defaults to ‘lognormal’, but can also be set to ‘wald’ or
‘levy’.</p></li>
<li><p><strong>color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'black'</em>) – face color for the lines in the movement-track plot.</p></li>
<li><p><strong>alpha</strong> (<em>valid mpl.plt alpha value</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – transparency value for the lines in the movement-track plot.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 10</em>) – size of the scatter points.</p></li>
<li><p><strong>land_cmap</strong> (<em>{valid string</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: 'plasma'</em>) – colormap to use for plotting layer rasters. If None, will default
to the colormap automatically assigned to the layer by geonomics
based on the layer’s index number in the landscape dict. can be
passed any string that references a colormap in matplotlib.pyplot.cm.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>color_by_individ</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, each plotted individual’s movement track will have its own
distinct color, making it easier to follow the track and distinguish it
from others in the plot.</p></li>
<li><p><strong>increasing_linewidth</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, each individual’s movement track will get wider with each
successive movement step, making it easier to distinguish
directionality in the movement track.</p></li>
<li><p><strong>include_start_points</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, a scatter of points will be plotted to signify the starting
points of individuals’ movement tracks.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_movement_surface">
<span class="sig-name descname"><span class="pre">plot_movement_surface</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hist'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_fact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_movement_surface" title="Permalink to this definition"></a></dt>
<dd><p>Create any of 4 types of plots to summarize a Species’ MovementSurface</p>
<p>Plot a Species’ MovemementSurface with either circularized histograms,
a circle of directional draws, or a vector on each cell, to visualize
the von Mises distributions of each cell on the MovementSurface.
Or plot a plain histogram of the von Mises distribution at a given
cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose MovementSurface should be plotted.
Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>style</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: 'hist'</em>) – <dl class="simple">
<dt>Determines the style of plot produced. There are four options:</dt><dd><dl class="simple">
<dt>’hist’:</dt><dd><p>Plot a classic histogram approximating the von Mises
distribution at the cell indicated by position x,y.
(Note that <em>x</em> and <em>y</em> must be provided for this style.)</p>
</dd>
<dt>’chist’:</dt><dd><p>Plot a circular histogram approximating the von Mises
distribution at the cell indicated by position x,y;
plot will be drawn inside the chosen cell on the
MovementSurface raster.</p>
</dd>
<dt>’cdraws’:</dt><dd><p>Plot points on the unit circle, whose locations were
drawn from the von Mises distribution at the cell indicated
by position x,y; plot will be drawn inside the chosen cell
on the MovementSurface raster.</p>
</dd>
<dt>’vect’:</dt><dd><p>Inside each cell of the MovementSurface raster,
plot the mean direction vector of directions drawn
from that cell’s von Mises distribution.</p>
</dd>
</dl>
</dd>
</dl>
</p></li>
<li><p><strong>color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'black'</em>) – Color to use for the plot elements. Passed to the
<em>c</em> argument of matplotlib.pyplot.hist, matplotlib.pyplot.plot,
matplotlib.pyplot.scatter, or matplotlib.pyplot.arrow,
depending on <em>style</em>.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the MovementSurface’s environmental values.</p></li>
<li><p><strong>cmap</strong> (<em>{valid string</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: 'plasma'</em>) – Colormap to use for plotting the MovementSurface. If None,
will default to ‘plasma’, a common colormap used to plot resistance
surfaces.Can be passed any string that references a colormap in
matplotlib.pyplot.cm.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Not valid for Species without a DisersalSurface</p></li>
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_phenotype">
<span class="sig-name descname"><span class="pre">plot_phenotype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_rast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">animate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_phenotype" title="Permalink to this definition"></a></dt>
<dd><p>Plot a Species’ Individuals, colored by their phenotypes for a Trait</p>
<p>Create a scatter plot of the Individuals in a Species,
on top of any Landscape Layer, and color the points by
their phenotypes for the chosen Trait.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species whose Individuals should be scattered
on the plot. Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>trt</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Trait whose phenotypes should be used to color
Individuals in the plot. Can be either the Trait’s index number
(i.e. its integer key in the GenomicArchitecture’s traits dict),
or its name (as a character string).</p></li>
<li><p><strong>lyr</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – A reference to the Layer whose raster should be plotted. Can be
either the Layer’s index number (i.e. its integer key in the
Landscape dict), or its name (as a character string). Defaults
to None, which will cause only the Layer associated with the chosen
Trait to be plotted.</p></li>
<li><p><strong>individs</strong> (<em>iterable collection of ints</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If provided, indicates the integer indices of the only Individuals
to be plotted. If None, all Individuals will be plotted.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, each Individual’s index number will be displayed next to
it. Can be useful for model introspection.</p></li>
<li><p><strong>edge_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>optional</em><em>, </em><em>default: 'face'</em>) – Edge color for the points in the Individual scatter. If ‘face’,
will always match the face color (i.e. the color provided to
<em>color</em>). Passed to the <em>edgecolor</em> argument of
matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_color</strong> (<em>valid mpl.plt color value</em><em>, </em><em>default 'black'</em>) – Color for the plotted text. (Will only be used if the <em>text</em>
argument is True.) Passed to the <em>color</em> argument of
matplotlib.pyplot.text.</p></li>
<li><p><strong>cbar</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – If True, a colorbar will be included, depicting the mapping of
color onto the Landscape’s environmental values.</p></li>
<li><p><strong>size</strong> (<em>scalar</em><em> or </em><em>array_like</em><em>, </em><em>optional</em><em>, </em><em>default: 25</em>) – Size of the scatter points. Passed to the <em>s</em> argument
of matplotlib.pyplot.scatter.</p></li>
<li><p><strong>text_size</strong> (<em>{size in points</em><em>, </em><em>valid string}</em><em>, </em><em>optional</em><em>, </em><em>default: 9</em>) – Text size. (Will only be used if <em>text</em> is True.) Can be expressed
as a numeric size in points, or as any string that is a valid size
argument for matplotlib.text.Text.</p></li>
<li><p><strong>alpha</strong> (<em>{scalar</em><em>, </em><em>bool</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The transparency level of the points. If scalar is passed, it is
passed on to the <em>alpha</em> value fed to matplotlib.pyplot.scatter.
If True, alpha value will be set to 0.6. If False or None, points
will be opaque.</p></li>
<li><p><strong>zoom_width</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Width, in raster cells, of the window to which to zoom to resulting
plot. If None, plot shows full Landscape.</p></li>
<li><p><strong>x</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>y</strong> (<em>{scalar</em><em>, </em><em>None}</em><em>, </em><em>default: None</em>) – The x and y coordinates of the center of the resulting plot.
Only used if <em>zoom_width</em> is not None.</p></li>
<li><p><strong>ticks</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, x- and y-axis ticks will be added to the plot.</p></li>
<li><p><strong>mask_rast</strong> (<em>array_like</em><em>, </em><em>optional default: None</em>) – An array to use to mask values in the Layer(s) being plotted.
All np.nan values in the array will be plotted as light gray,
instead of the color that would otherwise map to their
environmental values in their raster(s).
The array must be of the same x,y dimensions as the Layer(s)
being plotted.</p></li>
<li><p><strong>animate</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, this function will return a list of the Individuals’
points, as a matplotlib.collections.PathCollection. This is used
internally by the Model.walk method, to create animated simulation
plots.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output (unless <em>animate</em> is passed True, in which case
a PathCollection of the Individuals’ points is returned, for
internal use by Model.walk, to create animated simulations)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>{None, <code class="xref py py-class docutils literal notranslate"><span class="pre">matplotlib.collections.PathCollection</span></code>}</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For more detail on plotting parameters, see the documentation for
matplotlib, matplotlib.pyplot.pcolormesh, matplotlib.pyplot.scatter,
and matplotlib.pyplot.text.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.plot_pop_growth">
<span class="sig-name descname"><span class="pre">plot_pop_growth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">actual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expected_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">actual_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blue'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.plot_pop_growth" title="Permalink to this definition"></a></dt>
<dd><p>plot the chosen species’ population over time</p>
<p>plot a line plot of the species’ population over model time (in blue),
as well as the expected population-size trend (based on summing the
species’ carrying-capacity raster and using that total expected
population size to solve the the logistic growth equation; in red).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species whose population growth will be plotted.
can be either the species’ index number (i.e. its integer key
in the community dict), or its name (as a character string).</p></li>
<li><p><strong>expected</strong> (<em>bool</em><em>, </em><em>default: True</em>) – whether or not to plot the line showing the expected (under
the logistic growth model) trajectory of populaion growth, in red.</p></li>
<li><p><strong>actual</strong> (<em>bool</em><em>, </em><em>default: True</em>) – whether or not to plot the line showing the actual
trajectory of population growth, in blue.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns no output</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.remove_individuals">
<span class="sig-name descname"><span class="pre">remove_individuals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.remove_individuals" title="Permalink to this definition"></a></dt>
<dd><p>remove individuals from a species.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which values should be returned.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>n</strong> (<em>{int}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – the number of random individuals to be removed.
If None then <cite>n_out</cite> or <cite>individs</cite> must be provided.</p></li>
<li><p><strong>n_left</strong> (<em>{int}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – the number of individuals left in the species
after a random selection of individuals have been removed.
if None then <cite>n</cite> or <cite>individs</cite> must be provided.</p></li>
<li><p><strong>individs</strong> (<em>{list</em><em>, </em><em>tuple</em><em>, </em><em>numpy.ndarray}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – an iterable containing the integer ids (i.e., the species dict’s keys)
of the individuals to be removed. if None then <cite>n</cite> or <cite>n_out</cite> must be provided.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>alters the specified species in place by removing all specified
(if <cite>individs</cite> is not None) or randomly drawn (if <cite>n</cite> is not None)
individuals.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.run" title="Permalink to this definition"></a></dt>
<dd><p>Run a Model.</p>
<p>Run the Model for the number of iterations that was stipulated in the
parameters file used to create it (which is saved as the <cite>n_its</cite>
attribute). During each iteration, the model will burn in for at least
the stipulated number of burn-in timesteps (the <cite>burn_T</cite> attribute),
then will run for the stipulated number of main timesteps (the ‘T’
attribute). If stats and/or data are going to be collected (also
stipulated by the parameters file used to create the Model), then the
output files for each iteration will be saved in a separate
subdirectory, containing further subdirectories for each Species.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not to run the Model should provide written output.
If True, formatted messages will be printed to STDOUT at each
timestep (which could be piped to a log file, if desired).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>out</strong> – Returns no output. Writes information to STDOUT if verbose is
True. Writes data and statistics to file if so parameterized.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>We can create and run the default model (as long as there is no
“GNX_params_&lt;…&gt;.py” file in the current working directory
before <cite>gnx.make_parameters_file</cite> is called).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">NOTE: Using the following file, in the current working directory to create the Model object:</span>
<span class="go">        GNX_params_21-01-2020_17:22:08.py</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">###############################################################################################</span>
<span class="go">Running model &quot;GNX_params_21-01-2020_17:22:08&quot;...</span>
<span class="go">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span class="go">Setting up iteration 0...</span>
<span class="go">Creating the burn-in function queue...</span>
<span class="go">Creating the main function queue...</span>
<span class="go">Running burn-in, iteration 0...</span>
<span class="go">burn:   it=0:   t=0</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">burn:   it=0:   t=1</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">burn:   it=0:   t=2</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">.</span>
<span class="go">.</span>
<span class="go">.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.run_gea">
<span class="sig-name descname"><span class="pre">run_gea</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_sd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.run_gea" title="Permalink to this definition"></a></dt>
<dd><p>runs one of a number of available gea models, returning the results and
optionally plotting them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: &quot;cca&quot;</em>) – a string indicating which type of gea model to run.
note: currently, only cca (default) is available.</p></li>
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the species for which the gea should be run.
can be either the species’ index number (i.e. its
integer key in the community dict), or its name (as a character
string).</p></li>
<li><p><strong>trt</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – a reference to the trait for which the gea should be run.
can be either the trait’s index number (i.e. its integer key in the
genomicarchitecture’s traits dict), or its name (as a character
string).</p></li>
<li><p><strong>plot</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – whether or not the gea model should be plotted.</p></li>
<li><p><strong>plot_sd</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – whether or not a standard-deviation ellipse should be added to the
gea’s plot. (only used if plot==True and method==”cca”.)</p></li>
<li><p><strong>scale</strong> (<em>{int</em><em>, </em><em>float}</em><em>, </em><em>optional</em><em>, </em><em>default: 3</em>) – the scaling factor to use when plotting the gea results. (only used
if plot==True.)</p></li>
<li><p><strong>sd</strong> (<em>{int</em><em>, </em><em>float}</em><em>, </em><em>optional</em><em>, </em><em>default: 3</em>) – the number of standard deviations to use when plotting the
standard-deviation ellipse. (only used if plot==True and
method==”cca”.)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns a dict containing the labeled (as keys) output data
structures (as values) of the chosen method of gea.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.walk">
<span class="sig-name descname"><span class="pre">walk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'main'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">animate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.walk" title="Permalink to this definition"></a></dt>
<dd><p>Walk through a Model (i.e. run it for a certain number of timesteps).</p>
<p>This function will run a Model for a specific number of timesteps,
either in ‘burn’ or ‘main’ mode. It is designed to help users to try
out, explore, and introspect their Models in an interactive Python
session. (For best results, we strongly recommend this be done with
the <cite>iPython &lt;https://ipython.org&gt;</cite> shell.)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>T</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of timesteps for which to run the Model. (If in ‘main’
mode, the Model will run for precisely this many timesteps. If in
‘burn’ mode, it will run either until stationarity statistics
indicate that it has burned in, or for this many timesteps,
whichever occurs first.) Defaults to a single timestep.
single step.</p></li>
<li><p><strong>mode</strong> (<em>{'burn'</em><em>, </em><em>'main'}</em><em>, </em><em>optional</em>) – The mode in which to run the Model.
If ‘main’, it will run in main mode (i.e. all functionalities will
be used. (Note that for a Model to run in main mode it must first
have been burned in.)
If ‘burn’, it will run in burn-in mode. This entails the following:
1. Genomes will not yet be assigned, so genomic phenomena such
as crossing-over and natural selection will not occur;
2. Any scheduled landscape and demographic changes will not occur;
3. Stationarity statistics will be checked at the end of each
timestep, to determine if the Model has burned in (in which
case the model will stop running regardless of timestep,
genomes will be randomly assigned to all individuals, and
the Model will be deemed ready to run in main mode).</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not to run the Model should provide written output.
If True, formatted messages will be printed to STDOUT at each
timestep. Defaults to True for <cite>Model.walk</cite>.</p></li>
<li><p><strong>animate</strong> (<em>{tuple of ints</em><em>, </em><em>bool}</em><em>, </em><em>optional</em>) – If a length-2 tuple of integers is provided, the Species indicated
by the first number will be plotted on the Landscape Layer
indicated by the second number at each timestep, in a dynamically
updating plot.
If a length-3 tuple of integers in provided, the Species indicated
by the first number will be plotted on the Landscape Layer
indicated by the second number at each timestep, colored by the
phenotypes of the trait indicated by the third number.
If just True, the first Species (index 0) will be plotted
on top of a transparent stack of all Landscape Layers.
(Note that this will slow down the execution of a model a bit,
because the plot will pause for 0.1 seconds after each timestep.)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>out</strong> – Returns no output. Writes information to STDOUT if verbose is
True. Writes data and statistics to file if so parameterized.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the user attempts to run the Model in ‘main’ mode before it has
    been burned in.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>We can create and walk the default model (as long as there is no
“GNX_params_&lt;…&gt;.py” file in the current working directory
before <cite>gnx.make_parameters_file</cite> is called).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">NOTE: Using the following file, in the current working directory to create the Model object:</span>
<span class="go">        GNX_params_21-01-2020_17:22:08.py</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#run the burn-in until it is complete</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">T</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;burn&#39;</span><span class="p">)</span>
<span class="go">###############################################################################################</span>
<span class="go">Running model &quot;GNX_params_21-01-2020_17:22:08&quot;...</span>
<span class="go">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span class="go">Setting up iteration 0...</span>
<span class="go">Creating the burn-in function queue...</span>
<span class="go">Creating the main function queue...</span>
<span class="go">Running burn-in, iteration 0...</span>
<span class="go">burn:   it=0:   t=0</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">burn:   it=0:   t=1</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">burn:   it=0:   t=2</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">.</span>
<span class="go">.</span>
<span class="go">.</span>
<span class="go">Assigning genomes for species &quot;spp_0&quot;...</span>
<span class="go">Burn-in complete.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#now run in main mode for 50 timesteps</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;main&#39;</span><span class="p">)</span>
<span class="go"> main:   it=1:  t=0</span>
<span class="go">         species: spp_0                         N=131    (births=35      deaths=214)</span>
<span class="go"> .......................................................................................</span>
<span class="go"> main:   it=1:  t=1</span>
<span class="go">         species: spp_0                         N=129    (births=35      deaths=214)</span>
<span class="go"> .......................................................................................</span>
<span class="go"> main:   it=1:  t=2</span>
<span class="go">         species: spp_0                         N=133    (births=35      deaths=214)</span>
<span class="go"> .......................................................................................</span>
<span class="go"> .</span>
<span class="go"> .</span>
<span class="go"> .</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.write_gendata">
<span class="sig-name descname"><span class="pre">write_gendata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_fixed_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.write_gendata" title="Permalink to this definition"></a></dt>
<dd><p>A convenience function, designed for interactive use, that writes
genetic data for a sample of n Individuals to the given filepath</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>str</em>) – The path and filename indicating the destination of the output file.
Output format (‘VCF’, ‘FASTA’) will be determined by the
filepath extension (‘.vcf’, .’fasta’).</p></li>
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species for which values should be returned.
Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>n</strong> (<em>{int</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The number of individuals of randomly sample.
If None, output file will contain the entire current population.</p></li>
<li><p><strong>include_fixed_sites</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – A flag indicating whether or not to include fixed sites. Only valid
for VCF files. Defaults to True.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function is only designed for use with small models
or in interactive mode. For simulations being run to generate
datasets for research use, it is preferable to generate a
parameters file containing a section for data section
(<cite>gnx.make_parameters_file(…, data=True)</cite>), then parameterize
data collection there.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.write_geodata">
<span class="sig-name descname"><span class="pre">write_geodata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.write_geodata" title="Permalink to this definition"></a></dt>
<dd><p>A convenience function, designed for interactive use, that writes
tabular geodata for a sample of n Individuals to the given filepath</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>str</em>) – The path and filename indicating the destination of the output file.
Output format (‘CSV’, ‘Shapefile’, ‘GeoJSON’) will be determined
by the filepath extension (‘.csv’, ‘.shp’, ‘.json’).</p></li>
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species for which values should be returned.
Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>n</strong> (<em>{int</em><em>, </em><em>None}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The number of individuals of randomly sample.
If None, output file will contain the entire current population.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function is only designed for use with small models
or in interactive mode. For simulations being run to generate
datasets for research use, it is preferable to generate a
parameters file containing a section for data section
(<cite>gnx.make_parameters_file(…, data=True)</cite>), then parameterize
data collection there.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="geonomics.sim.model.Model.write_tskit_table_collection">
<span class="sig-name descname"><span class="pre">write_tskit_table_collection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_basename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">','</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.model.Model.write_tskit_table_collection" title="Permalink to this definition"></a></dt>
<dd><p>A convenience function, designed for interactive use, that writes the
tskit.TableCollection containing the spatial pedigree into CSV tables</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_basename</strong> (<em>str</em>) – The path and file basename to use for the resulting table files.
Each file will be appended with a capitalized suffix indicating
the table it contains (NODES, EDGES, SITES, MUTATIONS, INDIVIDUALS).</p></li>
<li><p><strong>spp</strong> (<em>{int</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: 0</em>) – A reference to the Species for which values should be returned.
Can be either the Species’ index number (i.e. its
integer key in the Community dict), or its name (as a character
string).</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: '</em><em>,</em><em>'</em>) – Character separator to use in the output CSV</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function is only designed for use with small models
or in interactive mode. For simulations being run to generate
datasets for research use, it is preferable to generate a
parameters file containing a section for data section
(<cite>gnx.make_parameters_file(…, data=True)</cite>), then parameterize
data collection there.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-geonomics.sim.params">
<span id="geonomics-sim-params-module"></span><h2>geonomics.sim.params module<a class="headerlink" href="#module-geonomics.sim.params" title="Permalink to this headline"></a></h2>
<p>Classes and functions for creating and reading parameters files</p>
<dl class="py class">
<dt class="sig sig-object py" id="geonomics.sim.params.ParametersDict">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.sim.params.</span></span><span class="sig-name descname"><span class="pre">ParametersDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.sim.params.ParametersDict" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.sim.params._DynAttrDict</span></code></p>
</dd></dl>

</section>
<section id="module-geonomics.sim.stats">
<span id="geonomics-sim-stats-module"></span><h2>geonomics.sim.stats module<a class="headerlink" href="#module-geonomics.sim.stats" title="Permalink to this headline"></a></h2>
<p>Classes and functons to implement calculation and output of statistics</p>
</section>
<section id="module-geonomics.sim">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-geonomics.sim" title="Permalink to this headline"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="geonomics.ops.html" class="btn btn-neutral float-left" title="geonomics.ops package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="geonomics.structs.html" class="btn btn-neutral float-right" title="geonomics.structs package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Drew Hart.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>