<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>geonomics package &mdash; geonomics 1.4.4 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="geonomics.help package" href="geonomics.help.html" />
    <link rel="prev" title="API" href="modules.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> geonomics
            <img src="_static/gnx_mini.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.4
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro_video.html">Intro Video</a></li>
<li class="toctree-l1"><a class="reference internal" href="procedural_diagram.html">Workflow Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="motivation.html">Motivation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="conceptual_diagram.html">Conceptual Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="params.html">Parameters</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">geonomics package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subpackages">Subpackages</a><ul>
<li class="toctree-l4"><a class="reference internal" href="geonomics.help.html">geonomics.help package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.ops.html">geonomics.ops package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.sim.html">geonomics.sim package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.structs.html">geonomics.structs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.utils.html">geonomics.utils package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-geonomics.main">geonomics.main module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#geonomics">Geonomics</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-geonomics">Module contents</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#geonomics-a-python-package-for-running-complex-landscape-genomics-simulations">Geonomics - A Python package for running complex landscape genomics simulations</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">geonomics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="modules.html">API</a> &raquo;</li>
      <li>geonomics package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/geonomics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="geonomics-package">
<h1>geonomics package<a class="headerlink" href="#geonomics-package" title="Permalink to this headline"></a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="geonomics.help.html">geonomics.help package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.help.html#module-geonomics.help">Module contents</a><ul>
<li class="toctree-l3"><a class="reference internal" href="geonomics.help.html#geonomics-help-geonomics-help">Geonomics Help (<code class="xref py py-mod docutils literal notranslate"><span class="pre">geonomics.help</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="geonomics.help.html#memory-help">Memory Help</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.help.html#parameters-help">Parameters Help</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.help.html#background-information">Background information</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.ops.html">geonomics.ops package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.change">geonomics.ops.change module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.demography">geonomics.ops.demography module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.mating">geonomics.ops.mating module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.movement">geonomics.ops.movement module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.mutation">geonomics.ops.mutation module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.selection">geonomics.ops.selection module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.sim.html">geonomics.sim package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.burnin">geonomics.sim.burnin module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.data">geonomics.sim.data module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.model">geonomics.sim.model module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.params">geonomics.sim.params module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.stats">geonomics.sim.stats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.structs.html">geonomics.structs package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.community">geonomics.structs.community module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.genome">geonomics.structs.genome module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.individual">geonomics.structs.individual module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.landscape">geonomics.structs.landscape module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.species">geonomics.structs.species module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.utils.html">geonomics.utils package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils.io">geonomics.utils.io module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils.spatial">geonomics.utils.spatial module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils.viz">geonomics.utils.viz module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
</section>
<section id="module-geonomics.main">
<span id="geonomics-main-module"></span><h2>geonomics.main module<a class="headerlink" href="#module-geonomics.main" title="Permalink to this headline"></a></h2>
<section id="geonomics">
<h3>Geonomics<a class="headerlink" href="#geonomics" title="Permalink to this headline"></a></h3>
<p>Provides:
1. A function for creating custom-structured, editable parameters files.
2. A function for creating model objects from parameters files.
3. Functions for running models.
4. Functions for model visualization.
5. A variety of parameters, for construction of arbitrarily complex simulation scenarios.</p>
<section id="how-to-use-the-documentation">
<h4>How to use the documentation<a class="headerlink" href="#how-to-use-the-documentation" title="Permalink to this headline"></a></h4>
<p>There are two levels of documentation available:
- Basic, procedural documentation is available as docstrings in the code,
which are returned when requesting Python’s help on a class or function.
- Detailed documentation is available at the
<a class="reference external" href="https://geonomics.readthedocs.io/en/latest/">Geonomics homepage</a>.</p>
<p>All documentation examples assume that <cite>geonomics</cite> has been imported as <cite>gnx</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">geonomics</span> <span class="k">as</span> <span class="nn">gnx</span>
</pre></div>
</div>
<p>To get help, call Python’s <cite>help</cite> function on a component of Geonomics that you
are trying to use, such as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">)</span>
</pre></div>
</div>
<p>Or use the question mark, if you are working in the
<a class="reference external" href="https://ipython.org">IPython</a> shell or in the
<a class="reference external" href="https://jupyter.org/">Jupyter</a> Notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; geonomics.make_parameters_file?
</pre></div>
</div>
<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_community">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_community</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">landscape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_community" title="Permalink to this definition"></a></dt>
<dd><p>Make a Community object.</p>
<p>Use the parameter values stored in the input params argument,
and the Landscape object input as the landscape argument,
to instantiate a Community object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>landscape</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><em>Landscape</em></a>) – A Landscape object (such as that returned by the
make_landscape function).</p></li>
<li><p><strong>spp_params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object (such as that output by the
make_params_dict function).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Community instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.community.Community" title="geonomics.structs.community.Community"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.community.Community</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_genomic_architecture">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_genomic_architecture</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">landscape</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_genomic_architecture" title="Permalink to this definition"></a></dt>
<dd><p>Make a GenomicArchitecture object.</p>
<p>Use the parameter values stored in the input params argument,
and the Landscape object input as the landscape argument,
to instantiate a GenomicArchitecture object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object (such as that output by the
make_params_dict function).</p></li>
<li><p><strong>landscape</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><em>Landscape</em></a>) – A Landscape object (such as that returned by the
make_landscape function).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A GenomicArchitecture instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.genome.GenomicArchitecture" title="geonomics.structs.genome.GenomicArchitecture"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.genome.GenomicArchitecture</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_individual">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_individual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomic_architecture</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parental_midpoint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">age</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_individual" title="Permalink to this definition"></a></dt>
<dd><p>Make an Individual object.</p>
<p>Make a new Individual, whose index number, genome,
location, sex, and age are determined by the input arguments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx</strong> (<em>int</em>) – The index number to be assigned to the new individual.</p></li>
<li><p><strong>genomic_architecture</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.genome.GenomicArchitecture" title="geonomics.structs.genome.GenomicArchitecture"><em>GenomicArchitecture</em></a><em>, </em><em>optional</em>) – A GenomicArchitecture object, describing the genomic
characteristics of the Species to which the Individual belongs.
If the new_genome argument is not provided then the relevant parameter
values in the GenomicArchitecture will be used when randomly
drawing the individual’s new genome.</p></li>
<li><p><strong>new_genome</strong> (<em>nx2 np.ndarray</em><em>, </em><em>optional</em>) – The Individual’s genome, as an nx2 Numpy ndarray of 0’s and 1’s.</p></li>
<li><p><strong>dim</strong> (<em>(</em><em>int</em><em>, </em><em>int</em><em>)</em><em>, </em><em>optional</em>) – A tuple of the landscape’s (x, y) dimensionality, to be used if the
individual’s location is to be randomly randomly drawn from anywhere on
the landscape (rather than being drawn by dispersal from the parental
midpoint, which will occur if the parental_midpoint argument is
provided).</p></li>
<li><p><strong>parental_midpoint</strong> (<em>(</em><em>float</em><em>, </em><em>float</em><em>)</em><em>, </em><em>optional</em>) – A tuple of the (x,y) continuous coordinates, representing the point
from which dispersal should occur to determine the individual’s
location. This is normally thought of as the midpoint between the individual’s
two parents, but if you are creating an individual without reference to
parents then this can also simply be thought of as representing the center of
the neighborhood into which the individual will disperse.</p></li>
<li><p><strong>sex</strong> (<em>int</em><em>, </em><em>optional</em>) – An integer representing the individual’s sex (0 for female, 1 for
male). Can be left as None for asexual or non-sexed Species.</p></li>
<li><p><strong>age</strong> (<em>int</em><em>, </em><em>optional</em>) – An integer representing the number of timesteps for which the
individual has already survived - hence its age, or stage.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An Individual instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.individual.Individual" title="geonomics.structs.individual.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.individual.Individual</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_landscape">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_landscape</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_landscape" title="Permalink to this definition"></a></dt>
<dd><p>Make a Landscape object.</p>
<p>Use the parameter values stored in the input params argument
to instantiate a Landscape object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object, such as that output by the
make_params_dict function.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Landscape instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.landscape.Landscape</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_model">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_model" title="Permalink to this definition"></a></dt>
<dd><p>Create a new Model object.</p>
<p>Use either a ParametersDict object or the path to a valid Geonomics
parameters file (whichever is provided to the ‘parameters’ argument) to
create a new Model object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parameters</strong> (<em>{ParametersDict</em><em>, </em><em>dict</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – <p>The parameters to be used to make the Model object.</p>
<p>If <cite>parameters</cite> is a ParametersDict object, the object will be used to
make the Model.</p>
<p>If <cite>parameters</cite> is a dict object, Geonomics will attempt to convert it
to a ParametersDict object, then use that object to make the Model.</p>
<p>If <cite>parameters</cite> is a string, Geonomics will call
<cite>gnx.read_parameters_file</cite> to make a ParametersDict object, then use
that object to make the Model.</p>
<p>If <cite>parameters</cite> is None, or is not provided, then Geonomics will
attempt to find a single parameters file in the current working
directory with the filename “GNX_params_&lt;…&gt;.py”, will use that
file to make a ParametersDict object, then will use that object to
make the Model.</p>
</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The name to assign to the model being created. If None, the name will
either be taken from parameters[‘model’][‘name’], if available;
or else set as the parameters filename, if <cite>parameters</cite> is a string;
or else set as ‘unnamed_model’.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – Whether or not to run the model in verbose mode, printing model-run
information to STDOUT at each time step.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An object of the Model class</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.Model</span></code></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If the <cite>parameters</cite> argument was not provided and a single, valid
    Geonomics parameters file could not be identified in the current
    working directory</p></li>
<li><p><strong>ValueError</strong> – If the <cite>parameters</cite> arugment was given a string that does not point
    to a valid parameters file</p></li>
<li><p><strong>ValueError</strong> – If the ParametersDict provided to the <cite>parameters</cite> argument, or created
    from the parameters file being used, cannot be successfully made into a
    Model</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnx.read_parameters_file</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">sim.model.Model</span></code></p>
</div>
<p class="rubric">Examples</p>
<p>Make a Model from a single, valid “GNX_params_&lt;…&gt;.py” file that can
be found in the current working directory (such as a file that would be
produced by calling gnx.make_parameters_file without any arguments).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     GNX_params_13-10-2018_15:54:03</span>
<span class="go">Layers:                                         0: &#39;0&#39;</span>
<span class="go">Species:                                        0: &#39;0&#39;</span>
<span class="go">Number of iterations:                           1</span>
<span class="go">Number of burn-in timesteps (minimum):          30</span>
<span class="go">Number of main timesteps:                       100</span>
<span class="go">Geo-data collected:                             {}</span>
<span class="go">Gen-data collected:                             {}</span>
<span class="go">Stats collected:                                {}</span>
</pre></div>
</div>
<p>Make a Model from a file called ‘null_model.py’, in the current working
directory.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
<span class="go">Layers:                                         0: &#39;tmp&#39;</span>
<span class="go">                                                1: &#39;ppt&#39;</span>
<span class="go">Species:                                        0: &#39;C. fasciata&#39;</span>
<span class="go">Number of iterations:                           2500</span>
<span class="go">Number of burn-in timesteps (mininum):          100</span>
<span class="go">Number of main timesteps:                       1000</span>
<span class="go">Geo-data collected:                             {csv, geotiff}</span>
<span class="go">Gen-data collected:                             {vcf, fasta}</span>
<span class="go">Stats collected:                                {maf, ld, mean_fit, het, Nt}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_parameters_file">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_parameters_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_parameters_file" title="Permalink to this definition"></a></dt>
<dd><p>Create a new parameters file.</p>
<p>Write to disk a new template parameters file. The file will contain the
numbers and types of sections indicated by the parameters fed to this
function. The new file can then be used ‘out of the box’
to make a new Model object, but typically it will be
edited by the user to stipulate the scenario being simulated,
then used to instantiate a Model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Where to write the resulting parameters file, in /path/to/filename.py
format. Defaults to None. If None, a file named
“GNX_params_&lt;datetime&gt;.py” will be written to the working
directory.</p></li>
<li><p><strong>layers</strong> (<em>{int</em><em>, </em><em>list of dicts}</em><em>, </em><em>optional</em><em>, </em><em>default: 1</em>) – Number (and optionally, types) of Layer-parameter sections to include
in the parameters file that is generated. Defaults to 1. Valid values
and their associated behaviors are:
1.) <em>If an int is passed</em>, add sections for the stipulated number of
Layers, each with default settings, including parameters for creating
Layers of type ‘random’ (i.e. Layers that will be generated by
interpolation from randomly valued random points),
and no LandscapeChanger parameters.
2.) <em>If [dict, …, dict] is passed</em>, each dict in the list
should have the following key-value pairs:
‘type’: {‘random’, ‘defined’, ‘file’, ‘nlmpy’}, ‘change’: bool.
This will add one section of Layer parameters, with the
contents indicated, for each dict in this list.</p></li>
<li><p><strong>species</strong> (<em>{int</em><em>, </em><em>list of dicts}</em><em>, </em><em>optional</em><em>, </em><em>default: 1</em>) – Number (and optionally, types) of Species-parameter sections to
include in the parameters file that is generated. Defaults to 1. Valid
values and their associated behaviors are:
1.) <em>If an int is passed</em>, add sections for the stipulated number
of Species, each with default settings, including parameters for
movement and dispersal without _ConductanceSurfaces, parameters for
a GenomicArchitecture with 0 Traits (i.e. with only neutral loci),
and no _SpeciesChanger parameters.
2.) <em>If [dict, …, dict] is passed</em>, each dict in the list
can contain any of the following key-value pairs:
‘msprime_source_pops’: int, ‘movement’: bool, ‘movement_surface’: bool,
‘dispersal_surface’: bool, ‘genomes’: {bool, ‘custom’}, ‘n_traits’: int,
‘demographic_change’: int, ‘parameter_change’: bool.
This will add one section of Species parameters,
customized as indicated, for each dict in the list.
(Note that if the ‘genomes’ argument is True or ‘custom’,
a section for parameterization of the genomic architecture will be added,
but if it is ‘custom’ then a template custom genomic architecture
CSV file will also be created, which can be filled in to stipulate
the locus-wise values for starting allele frequency,
recombination rate, dominance, associated traits, and effect sizes.)</p></li>
<li><p><strong>data</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – Whether to include a Data-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:
1.) None, False: Will not add a section for parameterizing
data to be collected. No _DataCollector will be created
for the Model object made from the resulting parameters file,
and no data will be collected during the model runs.
2.) True: Will add a section that can be used to parameterize which
data will be collected during the model runs, when, and what
file formats will be used to write it to disk.
(This which will be managed by the model’s _DataCollector
object.)</p></li>
<li><p><strong>stats</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – Whether to include a Stats-parameter section in the
parameters file that is generated.  Defaults to None.
Valid values and their associated behaviors are:
1.) None, False: Will not add a section for parameterizing
the statistics to be calculated. No _StatsCollector
will be created for the Model object made from
the resulting parameters file, and no statistics
will be calculated during the model runs.
2.) True: Will add a section that can be used
to parameterize which statistics will be calculated
during the model runs, and when. (This will be
managed by the model’s _StatsCollector object.)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output. Resulting parameters file will be written to the
location and filename indicated (or by default, will be written to a
file named “GNX_params_&lt;datetime&gt;.py” in the working directory).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sim.params.make_parameters_file</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>All parameters of this function are optional. Calling the function without
providing any parameters will always produce the parameters file for the
default model scenario. This file can be instantiated as a Model object and
run without being edited. Those three steps (create default parameters file;
create model from that parameters file; run the model) serve as a base case
to test successful package installation, and are wrapped around by the
convenience function <cite>gnx.run_default_model</cite>.</p>
<p class="rubric">Examples</p>
<p>In the simplest example, we can create a parameters file for the default
model. Then (assuming it is the only Geonomics parameters file in the
current working directory, so that it can be unambiguously identified) we
can call the gnx.make_model function to create a Model object from that
file, and then call the Model.run method to run the model (setting the
‘verbose’ parameter to True, so that we can observe model output).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">NOTE: Using the following file, in the current working directory to create the Model object:</span>
<span class="go">        GNX_params_21-01-2020_17:22:08.py</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">###############################################################################################</span>
<span class="go">Running model &quot;GNX_params_21-01-2020_17:22:08&quot;...</span>
<span class="go">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span class="go">Setting up iteration 0...</span>
<span class="go">Creating the burn-in function queue...</span>
<span class="go">Creating the main function queue...</span>
<span class="go">Running burn-in, iteration 0...</span>
<span class="go">burn:   it=0:   t=0</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">burn:   it=0:  t=1</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">burn:   it=0:  t=2</span>
<span class="go">        species: spp_0                         N=250    (births=35      deaths=214)</span>
<span class="go">.......................................................................................</span>
<span class="go">.</span>
<span class="go">.</span>
<span class="go">.</span>
</pre></div>
</div>
<p>We could also use some of this function’s arguments, to create a parameters
file for a model with 3 Layers and 1 Species (all with the default
components for their sections of the parameters file) and with a section
for parameterizing data collection.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span><span class="n">layers</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>As a more complex example that is likely to be similar to most use cases,
we can create a parameters file for a model scenario with:
- 2 Layers (one being an nlmpy Layer that will not change over model
time, the other being a raster read in from a GIS file and being
subject to change over model time);
- 2 Species (the first having genomes, 2 Traits, and movement
that is dictated by a _ConductanceSurface; the second not having
genomes but having dispersal determined by a
_ConductanceSurface, and undergoing
demographic change)
- data-collection;
- stats-collection;
We can save this to a file named “2-spp_2-trait_model.py” in our current
working directory.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Layer&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">layers</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;nlmpy&#39;</span><span class="p">},</span>                              <span class="c1">#layer 1</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;gis&#39;</span><span class="p">,</span>                                 <span class="c1">#layer 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Species&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">species</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>                               <span class="c1">#spp 1</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;n_traits&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>                              <span class="c1">#spp 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;dispersal_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;demographic_change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#arguments to the data and stats parameters</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#destination to which to write the resulting parameter file</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">filepath</span> <span class="o">=</span> <span class="s1">&#39;2-spp_2-trait_model.py&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_params_dict">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_params_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_params_dict" title="Permalink to this definition"></a></dt>
<dd><p>Create a ParametersDict object from a dict object.</p>
<p>Use a plain Python dict object, and an optional model name, to create
a new ParametersDict object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> (<em>{dict}</em>) – The plain Python dict that is to be turned into a Geonomics
ParametersDict object.</p></li>
<li><p><strong>model_name</strong> (<em>{str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The name to be assigned to the model. (If not provided, the model
will be called ‘unnamed_model’.)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dict of nested dicts, all of which have key-value pairs whose values
can be accessed using typical dict notation or using dot notation. If
formatted correctly, can be fed into the geonomics.make_model function
to create a Geonomics model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.ParametersDict</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.make_species">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">make_species</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">landscape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spp_params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_species" title="Permalink to this definition"></a></dt>
<dd><p>Make a Species object.</p>
<p>Use the parameter values stored in the input spp_params argument,
and the Landscape object input as the landscape argument,
to instantiate a Species object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>landscape</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><em>Landscape</em></a>) – A Landscape object (such as that returned by the
make_landscape function).</p></li>
<li><p><strong>spp_params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – The subsection of a full ParametersDict object (such as that output by the
make_params_dict function) that pertains to the particular species to
be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Species instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.species.Species" title="geonomics.structs.species.Species"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.species.Species</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.read_parameters_file">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">read_parameters_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.read_parameters_file" title="Permalink to this definition"></a></dt>
<dd><p>Create a new ParametersDict object.</p>
<p>Read the Geonomics parameters file saved at the location indicated by
‘filepath’, check its validity (i.e. that all the Layers and Species
parameterized in that file have been given distinct names), then use the
file to instantiate a ParametersDict object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filepath</strong> (<em>str</em>) – String indicating the location of the Geonomics parameters file that
should be made into a ParametersDict object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dict of nested dicts, all of which have key-value pairs whose values
can be accessed using typical dict notation or using dot notation.
Can be fed into the geonomics.make_model function to create a
Geonomics model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.ParametersDict</span></code></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>AssertionError</strong> – If either the Layers or the Species parameterized in the parameters
    file have not all been given distinct names</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sim.params.read</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">sim.params.ParametersDict</span></code></p>
</div>
<p class="rubric">Examples</p>
<p>Read a parameters file called “null_model.py” (located in the current
working directory).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">read_parameters_file</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.params.ParametersDict&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.run_default_model">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">run_default_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delete_params_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">animate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.run_default_model" title="Permalink to this definition"></a></dt>
<dd><p>Run the default Geonomics model.</p>
<p>This will create a parameters file for the default Geonomics model,
read that in as a model, and run a single iteration of that model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selection</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If False (default), will run a neutral model without traits or selection.
If True, will run a two-layer model in which the first layer serves as
the carrying-capacity and movement layer and the second layer is the
basis for selection on a 4-locus trait.</p></li>
<li><p><strong>delete_params_file</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – Whether or not to delete the parameters file used for this model
after the model has finished running</p></li>
<li><p><strong>animate</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – Whether or not plot the model’s species as an animated Matplotlib
plot while the model is running.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Model object that was created, after it has run to completion.
(If saved to a variable, it can be used for any number of additional
runs using the Model.walk method. It can also be used to try out
visualization methods, or to introspect its structure.)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.Model</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.run_demo">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">run_demo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_figs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_it</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.run_demo" title="Permalink to this definition"></a></dt>
<dd><p>Run a Geonomics demo.</p>
<p>This will run whichever of the Geonomics demos is stipulated by the <em>name</em>
argument.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – <p>Can be one of the following values:</p>
<blockquote>
<div><ul>
<li><p><em>’IBD IBE’</em>: Runs the IBD, IBE demo (example 1 from the methods
paper).</p></li>
<li><p><em>’simult select’</em>: Runs the simultaneous selection demo (example
2 from the methods paper).</p></li>
<li><p><em>’Yosemite’</em>: Runs the Yosemite demo (example 3 from the methods
paper).</p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>save_figs</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, all figures that are produced will be saved to their default
filenames in the current working directory.</p></li>
<li><p><strong>time_it</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – If True, the run time for the main portion of the model (excluding
calculation of data for figures and creation of figures, as much as
possible) will be calculated, then displayed after the model has run
to completion.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Some of the available demos come from the original Geonomics methods paper,
Hart and Wang 2020, “Geonomics: forward-time, agent-based, spatially
explicit, and arbitrarily complex landscape genomic simulations”.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The Model object that was created, after it has run to completion.
(If saved to a variable, it can be used for any number of additional
runs using the Model.walk method. It can also be used to try out
visualization methods, or to introspect its structure.)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.Model</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.main.sim_msprime_individuals">
<span class="sig-prename descclassname"><span class="pre">geonomics.main.</span></span><span class="sig-name descname"><span class="pre">sim_msprime_individuals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">L</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recomb_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">demography</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestry_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_tskit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.sim_msprime_individuals" title="Permalink to this definition"></a></dt>
<dd><p>Use an msprime simulation to generate new Geonomics Individuals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em>) – The number of Geonomics Individuals to be simulated</p></li>
<li><p><strong>L</strong> (<em>int</em>) – The length of the simulated genome.</p></li>
<li><p><strong>recomb_rate</strong> (<em>{float</em><em>, </em><em>list</em><em>, </em><em>numpy.ndarray}</em>) – Either a scalar float, indicating the genome-wide recombination rate;
or a 1d numpy.ndarray or 1d-array-like of length L-1, indicating the
inter-locus recombination rates along the length-L simulated genome.
(Note: This parameter is expressed according to the convention
used by Geonomics, as explained above. This differs from the convention
used by msprime, the translation to which will be handled internally by
Geonomics.)</p></li>
<li><p><strong>mut_rate</strong> (<em>{float</em><em>, </em><em>list</em><em>, </em><em>numpy.ndarray}</em>) – Either a scalar float, indicating the genome-wide neutral mutation rate;
or a 1d numpy.ndarray or 1d-array-like of length L, indicating the
mutation rates of all loci in the the length-L simulated genome.
(Note: This parameter is expressed according to the convention
used by Geonomics, as explained above. This differs from the convention
used by msprime, the translation to which will be handled internally by
Geonomics.)</p></li>
<li><p><strong>demography</strong> (<em>msprime.Demography</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – An <cite>msprime.Demography</cite> object to be used by <cite>msprime.sim_ancestry</cite>.
<cite>msprime</cite> allows specification of complex demographic histories,
including multiple extant and ancestral populations with fixed
or variable sizes, growth rates, or migration rates, as well as
timed demographic events. Defaults to None, in which case
<cite>msprime.sim_ancestry</cite> will simulate a single population
whose size will be determined by the <cite>population_size</cite> parameter
(or will default to 1, if <cite>population_size</cite> is also None).
Cannot be specified along with <cite>population_size</cite>.
(See the <cite>msprime.Demography</cite> documentation and references therein
for more details.)</p></li>
<li><p><strong>population_size</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The population size of the default single population to be simulated
by <cite>msprime.sim_ancestry</cite>. That default single population will only be
simulated if a more complex demographic scenario is not specified using
the <cite>demography</cite> parameter (i.e., if <cite>demography</cite> is None).
Thus, <cite>population_size</cite> cannot be specified along with <cite>demography</cite>.
If both are None then <cite>msprime</cite> simulates a single population of size 1.
(See the <cite>msprime.sim_ancestry</cite> documentation for more details.)</p></li>
<li><p><strong>ancestry_model</strong> ({<cite>msprime.AncestryModel</cite>, list}, optional, default: None) – The <cite>msprime.AncestryModel</cite> instance (or list of them) to be used by
<cite>msprime.sim_ancestry</cite>. See <cite>msprime.sim_ancestry</cite> documentation and
references therein for more details.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The random seed value to be provided to <cite>msprime.sim_ancestry</cite> and
<cite>msprime.sim_mutations</cite>. Use if reproducibility is desirable.
See <cite>msprime.sim_ancestry</cite> and <cite>msprime.sim_mutations</cite> documentation
for detail.</p></li>
<li><p><strong>use_tskit</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – Whether or not the Geonomics <cite>Species</cite> for which the <a href="#id1"><span class="problematic" id="id2">`</span></a>Individual`s are
intended uses tskit (i.e., has Species.gen_arch.use_tskit as True).
If not, <a href="#id3"><span class="problematic" id="id4">`</span></a>Individual`s will each store a copy of their Lx2 genotype
array in the ‘g’ attribute; otherwise, this attribute will be None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of <cite>geonomics.structs.individual.Individual</cite> objects.
Individuals are assigned sequential index integers starting from 0,
placed at x,y coordinates (0, 0), and otherwise given default attribute
values. Attributes should be updated as necessary in downstream code.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>
<section id="module-geonomics">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-geonomics" title="Permalink to this headline"></a></h2>
<section id="geonomics-a-python-package-for-running-complex-landscape-genomics-simulations">
<h3>Geonomics - A Python package for running complex landscape genomics simulations<a class="headerlink" href="#geonomics-a-python-package-for-running-complex-landscape-genomics-simulations" title="Permalink to this headline"></a></h3>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="modules.html" class="btn btn-neutral float-left" title="API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="geonomics.help.html" class="btn btn-neutral float-right" title="geonomics.help package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Drew Hart.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>