<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>geonomics.structs package &mdash; geonomics 1.3.9 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="geonomics.utils package" href="geonomics.utils.html" />
    <link rel="prev" title="geonomics.sim package" href="geonomics.sim.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> geonomics
            <img src="_static/gnx_mini.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.3
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro_video.html">Intro Video</a></li>
<li class="toctree-l1"><a class="reference internal" href="procedural_diagram.html">Workflow Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="motivation.html">Motivation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="conceptual_diagram.html">Conceptual Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="params.html">Parameters</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="geonomics.html">geonomics package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="geonomics.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="geonomics.help.html">geonomics.help package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.ops.html">geonomics.ops package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.sim.html">geonomics.sim package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">geonomics.structs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="geonomics.utils.html">geonomics.utils package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="geonomics.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="geonomics.html#geonomics-main-module">geonomics.main module</a></li>
<li class="toctree-l3"><a class="reference internal" href="geonomics.html#module-contents">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">geonomics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="modules.html">API</a> &raquo;</li>
          <li><a href="geonomics.html">geonomics package</a> &raquo;</li>
      <li>geonomics.structs package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/geonomics.structs.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="geonomics-structs-package">
<h1>geonomics.structs package<a class="headerlink" href="#geonomics-structs-package" title="Permalink to this headline"></a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
</div>
<div class="section" id="geonomics-structs-community-module">
<h2>geonomics.structs.community module<a class="headerlink" href="#geonomics-structs-community-module" title="Permalink to this headline"></a></h2>
</div>
<div class="section" id="module-geonomics.structs.genome">
<span id="geonomics-structs-genome-module"></span><h2>geonomics.structs.genome module<a class="headerlink" href="#module-geonomics.structs.genome" title="Permalink to this headline"></a></h2>
<p>Classes, associated methods, and supporting functions for all genomic
components</p>
<dl class="py class">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.genome.</span></span><span class="sig-name descname"><span class="pre">GenomicArchitecture</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g_params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">land</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recomb_rates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recomb_positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of the genomic architecture of a Species.</p>
<p>Inheres to its Species as the Species.gen_arch attribute.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.NOTE">
<span class="sig-name descname"><span class="pre">NOTE</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.NOTE" title="Permalink to this definition"></a></dt>
<dd><p>that correspond to many of the following attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>For more detail, see the documentation for the parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.delet_alpha_distr_scale">
<span class="sig-name descname"><span class="pre">delet_alpha_distr_scale</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.delet_alpha_distr_scale" title="Permalink to this definition"></a></dt>
<dd><p>The scale parameter of the Gamma distribution from which
the effect sizes (i.e. alphas) of deleterious loci are drawn</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.delet_alpha_distr_shape">
<span class="sig-name descname"><span class="pre">delet_alpha_distr_shape</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.delet_alpha_distr_shape" title="Permalink to this definition"></a></dt>
<dd><p>The shape parameter of the Gamma distribution from which
the effect sizes (i.e. alphas) of deleterious loci are drawn</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.delet_loci">
<span class="sig-name descname"><span class="pre">delet_loci</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.delet_loci" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array that tracks all deleterious loci</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.delet_loci_idxs">
<span class="sig-name descname"><span class="pre">delet_loci_idxs</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.delet_loci_idxs" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array containing the index positions within an
Individual’s genotypes array (attribute ‘g’ of an Individual)
that correspond to the deleterious loci.
(This is designed such that an Individual’s genotype for
the deleterious locus number stored at
GenomicArchitecture.delet_loci[i] can be accessed by calling
Individual.g[GenomicArchitecture.delet_loci_idxs[i], :].)
This method of tracking allows Individuals’ full genotypes to be
stored in the tskit tables (if tskit is being used)
while they carry their non-neutral genotypes with them
(a computational optimization).
Set to None if tskit is not being used.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.delet_loci_s">
<span class="sig-name descname"><span class="pre">delet_loci_s</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.delet_loci_s" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array containing the selection strengths
of each of the deleterious loci. (Selection strengths are
listed in an order corresponding to the order of
GenomicArchitecture.delet_loci, such that the selection strength of
the deleterious locus referenced at
GenomicArchitecture.delet_loci[i] is stored at
GenomicArchitetcture.delet_loci_s[i].)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.dom">
<span class="sig-name descname"><span class="pre">dom</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.dom" title="Permalink to this definition"></a></dt>
<dd><p>Dominance values for all loci (stored as an L-length,
1d numpy array in which 0 indicates a codominant locus
and 1 indicates a locus for which the ‘1’ allele is dominant)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.L">
<span class="sig-name descname"><span class="pre">L</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.L" title="Permalink to this definition"></a></dt>
<dd><p>Length of the simulated genome</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.mu_delet">
<span class="sig-name descname"><span class="pre">mu_delet</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.mu_delet" title="Permalink to this definition"></a></dt>
<dd><p>Genome-wide deleterious mutation rate (expressed in mutations
per locus per timestep). Deleterious mutations do not influence
trait phenotypes, but instead are univerally
deleterious (to an extent determined by their effect sizes).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.mu_neut">
<span class="sig-name descname"><span class="pre">mu_neut</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.mu_neut" title="Permalink to this definition"></a></dt>
<dd><p>Genome-wide neutral mutation rate (expressed in mutations
per locus per timestep).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.neut_loci">
<span class="sig-name descname"><span class="pre">neut_loci</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.neut_loci" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array that tracks all loci that do not influence the
phenotypes of any traits.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.nonneut_loci">
<span class="sig-name descname"><span class="pre">nonneut_loci</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.nonneut_loci" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array that tracks all loci that either influence the
phenotype of at least one trait or are deleterious.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.p">
<span class="sig-name descname"><span class="pre">p</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.p" title="Permalink to this definition"></a></dt>
<dd><p>Starting allele frequencies for all loci (stored as an L-length,
1d numpy array)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.pleiotropy">
<span class="sig-name descname"><span class="pre">pleiotropy</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.pleiotropy" title="Permalink to this definition"></a></dt>
<dd><p>A bool indicating whether or not to allow pleiotropy (i.e. whether
or not to allow the same locus to subtend multiple traits)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.recombinations">
<span class="sig-name descname"><span class="pre">recombinations</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.recombinations" title="Permalink to this definition"></a></dt>
<dd><p>The Recombinations object, which contains the parameters and data
necessary for simulation of recombination. Not intended for public
use at this time.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.sex">
<span class="sig-name descname"><span class="pre">sex</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.sex" title="Permalink to this definition"></a></dt>
<dd><p>A bool indicating whether or not Individuals of this Species
will be assigned a sex (0 for female, 1 for male)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.traits">
<span class="sig-name descname"><span class="pre">traits</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.traits" title="Permalink to this definition"></a></dt>
<dd><p>A dict containing all of a Species’ Trait objects, keyed by serial
integers.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.GenomicArchitecture.x">
<span class="sig-name descname"><span class="pre">x</span></span><a class="headerlink" href="#geonomics.structs.genome.GenomicArchitecture.x" title="Permalink to this definition"></a></dt>
<dd><p>Ploidy
NOTE: only diploidy is currently implemented, but this attribute</p>
<blockquote>
<div><p>serves as a placeholder for use in possible
future generalization to arbitrary ploidy.)</p>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="geonomics.structs.genome.MutationRateError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.genome.</span></span><span class="sig-name descname"><span class="pre">MutationRateError</span></span><a class="headerlink" href="#geonomics.structs.genome.MutationRateError" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="geonomics.structs.genome.Recombinations">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.genome.</span></span><span class="sig-name descname"><span class="pre">Recombinations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">L</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_distr_alpha</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_distr_beta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recomb_rates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter_breakpoints</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.genome.Recombinations" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Container for the parameters and data necessary to simulate recombination.</p>
<p>Not intended for public use.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.genome.</span></span><span class="sig-name descname"><span class="pre">Trait</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_loci</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_distr_mu</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_distr_sigma</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_alpha_mag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gamma</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">univ_adv</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.genome.Trait" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of a single trait and its parameterization.</p>
<p>Trait objects are collected in a serial integer-keyed dict,
one per trait that is parameterized for the corresponding Species,
which inheres to the Species’ GenomicArchitecture object as the attribute
<cite>GenomicArchitecture.traits</cite>, from which individual Traits can be indexed
using their index-number keys (e.g. <cite>GenomicArchitecture.traits[&lt;idx&gt;]</cite>).</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.NOTE">
<span class="sig-name descname"><span class="pre">NOTE</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.NOTE" title="Permalink to this definition"></a></dt>
<dd><p>that correspond to many of the following attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>For more detail, see the documentation for the parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.alpha">
<span class="sig-name descname"><span class="pre">alpha</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.alpha" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array containing the effect sizes of all loci subtending
this trait. The value at index i represents the effect size of the
locus at index i in the ‘loci’ attribute of this Trait.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.alpha_distr_mu">
<span class="sig-name descname"><span class="pre">alpha_distr_mu</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.alpha_distr_mu" title="Permalink to this definition"></a></dt>
<dd><p>Mean of the normal (Gaussian) distribution
of effect sizes for this trait’s loci.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.alpha_distr_sigma">
<span class="sig-name descname"><span class="pre">alpha_distr_sigma</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.alpha_distr_sigma" title="Permalink to this definition"></a></dt>
<dd><p>Standard deviation of the normal (Gaussian) distribution
of effect sizes for this trait’s loci.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.gamma">
<span class="sig-name descname"><span class="pre">gamma</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.gamma" title="Permalink to this definition"></a></dt>
<dd><p>Curvature of the fitness function.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.idx">
<span class="sig-name descname"><span class="pre">idx</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.idx" title="Permalink to this definition"></a></dt>
<dd><p>Index number of the trait (i.e. its key within the traits dict)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.loci_idxs">
<span class="sig-name descname"><span class="pre">loci_idxs</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.loci_idxs" title="Permalink to this definition"></a></dt>
<dd><p>A 1d numpy array containing the index positions within an
Individual’s genotypes array
(attribute ‘g’ of an Individual) that
correspond to the loci subtending this trait.
(This is designed such that an Individual’s genotype
for the locus number stored at Trait.loci[i]
can be accessed by calling Individual.g[Trait.loci_idxs[i], :].)
This method of tracking allows Individuals’ full genotypes to be
stored in tskit tables (if tskit is being used)
while they carry their non-neutral genotypes with them
(a computational optimization).
Set to None if tskit is not being used.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.loci">
<span class="sig-name descname"><span class="pre">loci</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.loci" title="Permalink to this definition"></a></dt>
<dd><p>A numerically sorted 1d numpy array containing
the locus numbers of all loci subtending this trait.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.lyr_num">
<span class="sig-name descname"><span class="pre">lyr_num</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.lyr_num" title="Permalink to this definition"></a></dt>
<dd><p>Index number of the Landscape layer to which this trait corresponds
(i.e. of the environmental variable that serves as the selective
force on this trait)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.max_alpha_mag">
<span class="sig-name descname"><span class="pre">max_alpha_mag</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.max_alpha_mag" title="Permalink to this definition"></a></dt>
<dd><p>Maximum absolute value that can be drawn for a locus’
effect size (i.e. alpha). Effect sizes are clipped
to the closed interval [-max_alpha_mag, max_alpha_mag]
when the model is parameterized.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.mu">
<span class="sig-name descname"><span class="pre">mu</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.mu" title="Permalink to this definition"></a></dt>
<dd><p>Mutation rate for this trait (expressed in mutations per locus per
timestep)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.name" title="Permalink to this definition"></a></dt>
<dd><p>The string name of the trait</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.n_loci">
<span class="sig-name descname"><span class="pre">n_loci</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.n_loci" title="Permalink to this definition"></a></dt>
<dd><p>Number of loci subtending this trait</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.phi">
<span class="sig-name descname"><span class="pre">phi</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.phi" title="Permalink to this definition"></a></dt>
<dd><p>Polygenic selection coefficient (i.e the selection coefficient
acting on the phenotypes, rather than the genotypes, of this Trait)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.genome.Trait.univ_adv">
<span class="sig-name descname"><span class="pre">univ_adv</span></span><a class="headerlink" href="#geonomics.structs.genome.Trait.univ_adv" title="Permalink to this definition"></a></dt>
<dd><p>Whether or not this trait will be treated as universally
advantageous. When False (the default behavior), individuals will
be fitter when their phenotypic values for this trait more closely
align with the local environmental value of the landscape layer
corresponding to this trait (i.e. selection will be spatially
varying). When False, phenotypes closer to 1 will be more
advantageous globally on the Landscape.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.structs.genome.read_pickled_genomic_architecture">
<span class="sig-prename descclassname"><span class="pre">geonomics.structs.genome.</span></span><span class="sig-name descname"><span class="pre">read_pickled_genomic_architecture</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.genome.read_pickled_genomic_architecture" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-geonomics.structs.individual">
<span id="geonomics-structs-individual-module"></span><h2>geonomics.structs.individual module<a class="headerlink" href="#module-geonomics.structs.individual" title="Permalink to this headline"></a></h2>
<p>Defines the Individual class, with its associated methods and supporting
functions</p>
<dl class="py class">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.individual.</span></span><span class="sig-name descname"><span class="pre">Individual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">age</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.individual.Individual" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of an individual of a given species.</p>
<p>Multiple Individuals are collected as serial integer-keyed values within a
Species dict. Those serial indices continually increment through model time
as new Individuals (i.e. offspring) are produced, such that no two
Individuals within the full history of a simulation will ever have the same
index number.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.NOTE">
<span class="sig-name descname"><span class="pre">NOTE</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.NOTE" title="Permalink to this definition"></a></dt>
<dd><p>that correspond to many of the following attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>For more detail, see the documentation for the parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.age">
<span class="sig-name descname"><span class="pre">age</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.age" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s age (expressed in timesteps since its birth
timestep, when the individual was instantiated at age 0)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.e">
<span class="sig-name descname"><span class="pre">e</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.e" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s current environmental values, organized as a
1d numpy array of length == len(Landscape), where the value stored
at Individual.e[i] gives the Individual’s current environmental
value for Layer number i (i.e. Landscape[i])</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.fit" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s current fitness (calculated as a combination of
the Individual’s phenotypes for all traits, its current
environmental values for all corresponding Landscape Layers, and
any deleterious loci for which it has ‘1’ alleles; see
documentation for further details)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.g">
<span class="sig-name descname"><span class="pre">g</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.g" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s genotype array,
stored as an L_g x 2 numpy array, where L_g is the number
of loci being tracked by each Individual in the Species.</p>
<p>If tskit is being used then Individuals only carry
copies of their non-neutral genotypes
(stored in this attribute); this is a computational optimization,
as it allows fitness-based operations to be calculated
quickly on the fly (using this attribute’s data), while minimizing
the memory required to stored the full (neutral and non-netural)
genotypes of all current Individuals and their ancestors (i.e. the
‘spatial pedigree’ stored in the succinct set of tskit tables).
The successive rows in this array store the genotypes corresponding
to the loci indicated by the successive locus numbers in
Species.GenomicArchitecture.nonneut_loci.</p>
<p>If tskit is not being used then this array will store genotypes
for all loci in the simulated genome (i.e. it will be of shape
L x 2, where L is genome-length parameter in the this Individual’s
Species’ section of the paramters file).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.sex">
<span class="sig-name descname"><span class="pre">sex</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.sex" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s sex (0=female, 1=male; None, if the Species is
unsexed)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.x">
<span class="sig-name descname"><span class="pre">x</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.x" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s current x (i.e. longitudinal) position
(in continuous space, bounded within [0, Landscape.dims[0]])</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.y">
<span class="sig-name descname"><span class="pre">y</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.y" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s current y (i.e. latitudinal) position
(in continuous space, bounded within [0, Landscape.dims[1]])</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.individual.Individual.z">
<span class="sig-name descname"><span class="pre">z</span></span><a class="headerlink" href="#geonomics.structs.individual.Individual.z" title="Permalink to this definition"></a></dt>
<dd><p>The Individual’s phenotypes for each, organized as a
1d numpy array of length == len(Species.traits), where the value
stored at Individual.z[i] gives the Individual’s fitness for
Trait number i (i.e. Species.traits[i]).</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-geonomics.structs.landscape">
<span id="geonomics-structs-landscape-module"></span><h2>geonomics.structs.landscape module<a class="headerlink" href="#module-geonomics.structs.landscape" title="Permalink to this headline"></a></h2>
<p>Defines the Layer and Landscape classes, with their associated methods and
supporting functions</p>
<dl class="py class">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.landscape.</span></span><span class="sig-name descname"><span class="pre">Landscape</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lyrs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">res</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1,</span> <span class="pre">1)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ulc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.landscape.Landscape" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<p>Representation of a multi-layer (i.e. multivariate) landscape.</p>
<p>Organized as a dict of multiple, serial integer-keyed Layer objects,</p>
<p>Because the Landscape class inherits from <cite>dict</cite>, Layers can
be indexed out using their index-number keys (e.g. <cite>mod.land[&lt;idx&gt;]</cite>).</p>
<p>The Landscape is stored as the ‘mod.land’ attribute of its corresponding
Model object.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape.NOTE">
<span class="sig-name descname"><span class="pre">NOTE</span></span><a class="headerlink" href="#geonomics.structs.landscape.Landscape.NOTE" title="Permalink to this definition"></a></dt>
<dd><p>that correspond to many of the following attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>For more detail, see the documentation for the parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape.dim">
<span class="sig-name descname"><span class="pre">dim</span></span><a class="headerlink" href="#geonomics.structs.landscape.Landscape.dim" title="Permalink to this definition"></a></dt>
<dd><p>The x,y (i.e. lon,lat; or i,j in array terms) dimensions of the
Layer. (Must be the same as the Landscape to which the Layer
belongs.)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape.n_lyrs">
<span class="sig-name descname"><span class="pre">n_lyrs</span></span><a class="headerlink" href="#geonomics.structs.landscape.Landscape.n_lyrs" title="Permalink to this definition"></a></dt>
<dd><p>The number of Layers in the Landscape</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape.prj">
<span class="sig-name descname"><span class="pre">prj</span></span><a class="headerlink" href="#geonomics.structs.landscape.Landscape.prj" title="Permalink to this definition"></a></dt>
<dd><p>The projection of the Layer (formatted as a PROJ4 string).
(Must be the same as the Landscape to which the Layer belongs.)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape.res">
<span class="sig-name descname"><span class="pre">res</span></span><a class="headerlink" href="#geonomics.structs.landscape.Landscape.res" title="Permalink to this definition"></a></dt>
<dd><p>The x,y (i.e. lon,lat; or i,j in array terms) spatial resolution
(i.e. cell sizes) of the layer.  (Must be the same as the
Landscape to which the Layer belongs.)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Landscape.ulc">
<span class="sig-name descname"><span class="pre">ulc</span></span><a class="headerlink" href="#geonomics.structs.landscape.Landscape.ulc" title="Permalink to this definition"></a></dt>
<dd><p>The x,y (i.e. lon,lat; or i,j in array terms) coordinates of the
upper left corner of the layer. (Must be the same as the Landscape
to which the Layer belongs.)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">geonomics.structs.landscape.</span></span><span class="sig-name descname"><span class="pre">Layer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rast</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lyr_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">res</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1,</span> <span class="pre">1)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ulc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_prec</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">units</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.landscape.Layer" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of a single environmental layer (i.e. variable).</p>
<p>Multiple Layers are collected as serial integer-keyed values within
a Landscape dict.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.NOTE">
<span class="sig-name descname"><span class="pre">NOTE</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.NOTE" title="Permalink to this definition"></a></dt>
<dd><p>that correspond to many of the following attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>For more detail, see the documentation for the parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.coord_prec">
<span class="sig-name descname"><span class="pre">coord_prec</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.coord_prec" title="Permalink to this definition"></a></dt>
<dd><p>The precision (i.e. number of significant digits) to which
coordinate values should be round when the Layer is plotted.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.dim">
<span class="sig-name descname"><span class="pre">dim</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.dim" title="Permalink to this definition"></a></dt>
<dd><p>The x,y (i.e. lon,lat; or i,j in array terms) dimensions of the
Layer. (Must be the same as the Landscape to which the Layer
belongs.)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.idx">
<span class="sig-name descname"><span class="pre">idx</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.idx" title="Permalink to this definition"></a></dt>
<dd><p>Index number of the Layer (i.e. its key within the Landscape dict)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.name" title="Permalink to this definition"></a></dt>
<dd><p>The string name of the Layer</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.prj">
<span class="sig-name descname"><span class="pre">prj</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.prj" title="Permalink to this definition"></a></dt>
<dd><p>The projection of the Layer (formatted as a PROJ4 string).
(Must be the same as the Landscape to which the Layer belongs.)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.rast">
<span class="sig-name descname"><span class="pre">rast</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.rast" title="Permalink to this definition"></a></dt>
<dd><p>The 2d numpy array, of shape <cite>Landscape.dim</cite>, containing the
environmental values for this Layer.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="geonomics.structs.landscape.Layer.res">
<span class="sig-name descname"><span class="pre">res</span></span><a class="headerlink" href="#geonomics.structs.landscape.Layer.res" title="Permalink to this definition"></a></dt>
<dd><p>The x,y (i.e. lon,lat; or i,j in array terms) spatial resolution
(i.e. cell sizes) of the layer.  (Must be the same as the
Landscape to which the Layer belongs.)</p>
<blockquote>
<div><dl class="simple">
<dt>type:</dt><dd><p>A string indicating the type of the Layer (‘random’, ‘defined’,
‘file’, or ‘nlmpy’)</p>
</dd>
<dt>ulc:</dt><dd><p>The x,y (i.e. lon,lat; or i,j in array terms) coordinates of the
upper left corner of the layer. (Must be the same as the Landscape
to which the Layer belongs.)</p>
</dd>
<dt>units:</dt><dd><p>A string representation of the units of the Layer’s variable
(to be used for plotting).</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="geonomics.structs.landscape.read_pickled_land">
<span class="sig-prename descclassname"><span class="pre">geonomics.structs.landscape.</span></span><span class="sig-name descname"><span class="pre">read_pickled_land</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.structs.landscape.read_pickled_land" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</div>
<div class="section" id="geonomics-structs-species-module">
<h2>geonomics.structs.species module<a class="headerlink" href="#geonomics-structs-species-module" title="Permalink to this headline"></a></h2>
</div>
<div class="section" id="module-contents">
<h2>Module contents<a class="headerlink" href="#module-contents" title="Permalink to this headline"></a></h2>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="geonomics.sim.html" class="btn btn-neutral float-left" title="geonomics.sim package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="geonomics.utils.html" class="btn btn-neutral float-right" title="geonomics.utils package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Drew Hart.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>