<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Parameters &mdash; geonomics 1.4.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API" href="modules.html" />
    <link rel="prev" title="Visualization" href="viz.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> geonomics
            <img src="_static/gnx_mini.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.4
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro_video.html">Intro Video</a></li>
<li class="toctree-l1"><a class="reference internal" href="procedural_diagram.html">Workflow Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="motivation.html">Motivation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="conceptual_diagram.html">Conceptual Diagram</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference Guide</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Parameters</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#landscape-parameters">Landscape parameters</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#main">Main</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layers">Layers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#init">Init</a></li>
<li class="toctree-l4"><a class="reference internal" href="#change">Change</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#community-and-species-parameters">Community and Species parameters</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#species">Species</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">Init</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mating">Mating</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mortality">Mortality</a></li>
<li class="toctree-l4"><a class="reference internal" href="#movement">Movement</a></li>
<li class="toctree-l4"><a class="reference internal" href="#genomicarchitecture">_GenomicArchitecture</a></li>
<li class="toctree-l4"><a class="reference internal" href="#species-change">Species change</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#other-parameters">Other parameters</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Main</a></li>
<li class="toctree-l3"><a class="reference internal" href="#iterations">Iterations</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#data">Data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#stats">Stats</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">geonomics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Parameters</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/params.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="parameters">
<h1>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline"></a></h1>
<p>How to Use This Section:</p>
<p>In order to create and run a Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>, you will need a valid
Geonomics parameters file. No worry though – this is very easy to create!
To generate a new, template parameters file, you will simply call the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> function, feeding it the appropriate
arguments (to indicate how many <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s you
want to include in your <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>; which parameters sections you want
included in the file, both for those
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> and for
other components of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>; and the path and filename for your new
parameters file). Geonomics will then automatically create the file for you,
arranged as you requested and saved where you requested.</p>
<p>When you then open that file, you will see the following:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#&lt;your_filename&gt;.py</span>

<span class="c1">#This is a default parameters file generated by Geonomics</span>
<span class="c1">#(by the gnx.params.make_parameters_file() function).</span>


                 <span class="c1">#   ::::::          :::    :: :::::::::::#</span>
           <span class="c1">#::::::    ::::   :::      ::    :: :: ::::::::::: ::#</span>
        <span class="c1">#:::::::::     ::            ::   :::::::::::::::::::::::::#</span>
      <span class="c1">#::::::::::                      :::::::::: :::::: ::::::::  ::#</span>
    <span class="c1">#  : ::::  ::                    ::::  : ::    :::::::: : ::  :    #</span>
   <span class="c1"># GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #</span>
  <span class="c1"># GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS     #</span>
  <span class="c1"># GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS #</span>
  <span class="c1"># GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS #</span>
  <span class="c1"># GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS #</span>
   <span class="c1"># GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #</span>
    <span class="c1">#    : ::::::::               :::::::::: ::              ::  :   : #</span>
      <span class="c1">#:    :::::                    :::::: :::             :::::::  #</span>
        <span class="c1">#    :::                      :::::  ::              ::::: #</span>
           <span class="c1">#  ::                      ::::                      #</span>
                 <span class="c1"># ::                                     #</span>


<span class="n">params</span> <span class="o">=</span> <span class="p">{</span>

<span class="c1">##############</span>
<span class="c1">#### LAND ####</span>
<span class="c1">##############</span>
    <span class="s1">&#39;land&#39;</span><span class="p">:</span> <span class="p">{</span>

    <span class="c1">##############</span>
    <span class="c1">#### main ####</span>
    <span class="c1">##############</span>
        <span class="s1">&#39;main&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># y,x (a.k.a. i,j) dimensions of the Landscape</span>
            <span class="s1">&#39;dim&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>

   <span class="c1">#.</span>
   <span class="c1">#.</span>
   <span class="c1">#.</span>
</pre></div>
</div>
<p>This is the beginning of a file that is really just a long but simple Python
script (hence the ‘.py’ extension); this whole file just defines a single,
long, nested <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> (i.e. a Python ‘dictionary’) containing all of your
parameter values. It may look like a lot, but don’t be concerned! For two
reasons:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>All the hard work is already done for you. You’ll just need to change
the default values where and how you want to, to set up your particular
simulation scenario.</p></li>
<li><p>You will probably leave a good number of the parameters defined in this
file untouched. Geonomics does its best to set sensible default values
for all its parameters. Though of course, you’ll want to think clearly
nonetheless about whether the default value for each parameter
is satisfactory for your purposes.</p></li>
</ol>
</div></blockquote>
<p>Each parameter in the parameters value is preceded by a terse comment, to
remind you what the parameter does. But for detailed information about each
parameter, you’ll want to refer to the following information.
What follows is a list of all of the Geonomics parameters (in the sections and
the top-to-bottom order in which they’ll appear in your parameters files).
For each parameter, you will see a section with the following information:</p>
<blockquote>
<div><ul class="simple">
<li><p>a snippet of the context (i.e. lines of
Python code) in which it appears in a parameters file;</p></li>
<li><p>the valid Python data type(s) the parameter can take</p></li>
<li><p>the default value of the parameter</p></li>
<li><p>a ranking score, indicating how likely it is that you will want to reset
this parameter (i.e. change it from its default value), and
encoded as follows:</p>
<ul>
<li><p>‘Y’: almost certainly, <em>or</em> must be reset for your <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> to run</p></li>
<li><p>‘P’: it is quite possible that you will want to reset this
parameter, but this will depend on your use and scenario</p></li>
<li><p>‘N’: almost certainly not, <em>or</em> no need to reset because it should be
set intelligently anyhow (Note: this does <em>not</em> mean that you cannot
reset the parameter! if that is the case for any value then it does not
appear in the parameters file)</p></li>
</ul>
</li>
<li><p>other relevant, detailed information about the parameter, including
an explanation of what it defines, how its value is used, where to look
for additioanl information about parameters related to other Python
packages, etcetera</p></li>
</ul>
</div></blockquote>
<p>These section will be formatted as follows:</p>
<p><strong>&lt;param_name&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#brief comment about the parameter</span>
<span class="s1">&#39;&lt;param_name&gt;&#39;</span><span class="p">:</span>               <span class="o">&lt;</span><span class="n">default_param_value</span><span class="o">&gt;</span><span class="p">,</span>
</pre></div>
</div>
<p>&lt;valid Python data type(s)&gt;</p>
<p>default: &lt;default value&gt;</p>
<p>reset? &lt;ranking&gt;</p>
<blockquote>
<div><p>&lt;Explanation of what the parameter defines, how its value is used,
and any other relevant information.&gt;</p>
</div></blockquote>
<p>This section should serve as your primary point of reference
if you confront any uncertainty while creating your own parameters files.
We’ll start with the section of parameters that
pertains to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> object.</p>
<hr class="docutils" />
<section id="landscape-parameters">
<h2>Landscape parameters<a class="headerlink" href="#landscape-parameters" title="Permalink to this headline"></a></h2>
<section id="main">
<h3>Main<a class="headerlink" href="#main" title="Permalink to this headline"></a></h3>
<p><strong>dim</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># x,y (a.k.a. j,i) dimensions of the Landscape</span>
<span class="s1">&#39;dim&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="literal number integer"><span class="pre">20</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">20</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset: P</p>
<blockquote>
<div><p>This defines the y,x dimensions of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>,
in units of cells. As you might imagine, these values are used
for a wide variety of basic operations throughout Geonomics. Change the
default value to the dimensions of the landscape you wish to simulate on.</p>
</div></blockquote>
<p><strong>res</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># x,y resolution of the Landscape</span>
<span class="s1">&#39;res&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="literal number integer"><span class="pre">1</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">1</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset: N</p>
<blockquote>
<div><p>This defines the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> resolution (or cell-size) in the y,x
dimensions (matching the convention of the <strong>dim</strong> parameter).
This information is only used if GIS rasters of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>
layers are to be written out as GIS raster files (as parameterized in the
‘Data’ parameters). Defaults to the meaningless value (1,1), and this value
generally needn’t be changed in your parameters file, because it will
be automatically updated to the resolution of any GIS rasters that
are read in for use as <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layers</span></span></code> (assuming they all share the same
resolution; otherwise, an Error is thrown).</p>
</div></blockquote>
<p><strong>ulc</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># x,y coords of upper-left corner of the Landscape</span>
<span class="s1">&#39;ulc&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="literal number integer"><span class="pre">0</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">0</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset: N</p>
<blockquote>
<div><p>This defines the x,y upper-left corner (ULC) of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> (in the units of
some real-world coordinate reference system, e.g. decimal degrees, or
meters). This information is only used if GIS rasters of
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> layers are to be written out as GIS raster files.
Defaults to the meaningless value
(0,0), and this value usually needn’t be changed in your parameters file,
because it will be automatically updated to match the ULC value
of any GIS rasters that are read in for use as <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layers</span></span></code> (assuming
they all share the same ULC; otherwise, an Error is thrown).</p>
</div></blockquote>
<p><strong>prj</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#projection of the Landscape</span>
<span class="s1">&#39;prj&#39;</span><span class="p">:</span>                      <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>; (WKT projection string)</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset: N</p>
<blockquote>
<div><p>This defines the projection of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, as a
string of Well Known Text (WKT).
This information is only used if GIS rasters of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> layers are
to be written out as GIS raster files. Defaults to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, which is fine,
because this value will be automatically updated to match the projection
of any GIS rasters that are read in for us as <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layers</span></span></code> (assuming they
all share the same projection; otherwise, an Error is thrown)</p>
</div></blockquote>
</section>
<hr class="docutils" />
<section id="layers">
<h3>Layers<a class="headerlink" href="#layers" title="Permalink to this headline"></a></h3>
<p><strong>layer_&lt;n&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">################</span>
<span class="c1">#### layers ####</span>
<span class="c1">################</span>
    <span class="s1">&#39;layers&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="c1">#layer name (LAYER NAMES MUST BE UNIQUE!)</span>
        <span class="s1">&#39;layer_0&#39;</span><span class="p">:</span> <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">layer_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>. (Note that unlike most
parameters, this parameter is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key,
the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> must have a unique name
(so that a parameterized <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> isn’t overwritten in the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">ParametersDict</span></span></code> by a second, identically-named <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> represents. Example valid values include: 0, 0.1, ‘layer_0’, 1994,
‘1994’, ‘mean_ann_tmp’. Names default to <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">layer_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>,
where n is a series of integers starting from 0 and counting the number
of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s.</p>
<hr class="docutils" />
<section id="init">
<h4>Init<a class="headerlink" href="#init" title="Permalink to this headline"></a></h4>
<p>There are four different types of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layers</span></span></code> that can be created. The
parameters for each are explained in the next four subsections.</p>
<section id="random">
<h5>random<a class="headerlink" href="#random" title="Permalink to this headline"></a></h5>
<p><strong>n_pts</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;random&#39;-type Layer</span>
<span class="s1">&#39;rand&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#number of random points</span>
    <span class="s1">&#39;n_pts&#39;</span><span class="p">:</span>                        <span class="mi">500</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 500</p>
<p>reset? P</p>
<p>This defines the number of randomly located, randomly valued points
from which the random <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> will be interpolated. (Locations drawn
from uniform distributions between 0 and the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimensions on
each axis. Values drawn from a uniform distribution between 0 and 1.)</p>
<p><strong>interp_method</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#interpolation method (&#39;linear&#39;, &#39;cubic&#39;, or &#39;nearest&#39;)</span>
<span class="s1">&#39;interp_method&#39;</span><span class="p">:</span>                <span class="s1">&#39;linear&#39;</span><span class="p">,</span>
<span class="p">},</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'linear'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cubic'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'nearest'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'linear'</span></span></code></p>
<p>reset? N</p>
<p>This defines the method to use to interpolate random points to the array that
will serve as the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>’s raster. Whichever of the three valid values
is chosen (<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'linear'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cubic'</span></span></code>, or <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'nearest'</span></span></code>) will be passed
on as an argument to <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">interpolate</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">griddata</span></span></code>. Note that the
<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'nearest'</span></span></code> method will generate a random categorical array, such as
might be used for modeling habitat types.</p>
</section>
<hr class="docutils" />
<section id="defined">
<h5>defined<a class="headerlink" href="#defined" title="Permalink to this headline"></a></h5>
<p><strong>rast</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;defined&#39;-type Layer</span>
<span class="s1">&#39;defined&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#raster to use for the Layer</span>
    <span class="s1">&#39;rast&#39;</span><span class="p">:</span>                    <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">100</span><span class="p">,</span><span class="mi">100</span><span class="p">)),</span>
</pre></div>
</div>
<p>nx2 <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ones</span></span><span class="punctuation"><span class="pre">((</span></span><span class="literal number integer"><span class="pre">100</span></span><span class="punctuation"><span class="pre">,</span></span><span class="literal number integer"><span class="pre">100</span></span><span class="punctuation"><span class="pre">))</span></span></code></p>
<p>reset? Y</p>
<p>This defines the raster that will be used for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>. Can be set to
<code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> if an array for the raster should instead be interpolated from a
set of valued points using the <strong>pts</strong>, <strong>vals</strong>, and <strong>interp_method</strong>
parameters. Dimensions of this array must match the dimensions of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>.</p>
<p><strong>pts</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;defined&#39;-type Layer</span>
<span class="s1">&#39;defined&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#point coordinates</span>
    <span class="s1">&#39;pts&#39;</span><span class="p">:</span>                    <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>nx2 <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? Y</p>
<p>This defines the coordinates of the points to use to
interpolate this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>. Can be left as <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> if the <strong>rast</strong>
parameter is given a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">numpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>.</p>
<p><strong>vals</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#point values</span>
<span class="s1">&#39;vals&#39;</span><span class="p">:</span>                  <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code>, 1xn <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? Y</p>
<p>This defines the values of the points to use to
interpolate this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>. Can be left as <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> if the <strong>rast</strong>
parameter is given a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">numpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>.</p>
<p><strong>interp_method</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#interpolation method {None, &#39;linear&#39;, &#39;cubic&#39;,</span>
<span class="c1">#&#39;nearest&#39;}</span>
<span class="s1">&#39;interp_method&#39;</span><span class="p">:</span>                <span class="kc">None</span><span class="p">,</span>
<span class="p">},</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'linear'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cubic'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'nearest'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? N</p>
<p>This defines the method to use to interpolate random points to the array that
will serve as the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>’s raster. Whichever of the valid string values
is chosen (<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'linear'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cubic'</span></span></code>, or <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'nearest'</span></span></code>) will be passed
on as an argument to <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">interpolate</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">griddata</span></span></code>. Note that the
<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'nearest'</span></span></code> method will generate a random categorical array, such as
might be used for modeling habitat types. Can be left as <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> if
the <strong>rast</strong> parameter is given a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">numpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>.</p>
</section>
<hr class="docutils" />
<section id="file">
<h5>file<a class="headerlink" href="#file" title="Permalink to this headline"></a></h5>
<p><strong>filepath</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#parameters for a &#39;file&#39;-type Layer</span>
<span class="s1">&#39;file&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#&lt;/path/to/file&gt;.&lt;ext&gt;</span>
    <span class="s1">&#39;filepath&#39;</span><span class="p">:</span>                     <span class="s1">&#39;/PATH/TO/FILE.EXT&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'/PATH/TO/FILE.EXT'</span></span></code></p>
<p>reset? Y</p>
<p>This defines the location and name of the file that should be read in as the
raster-array for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>. Valid file types include a ‘.txt’ file
containing a 2d <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>, or any GIS raster file that can be read
by <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">rasterio</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">open</span></span></code>. In all cases, the raster-array read in from the
file must have dimensions equal to the stipulated dimensions of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> (as defined in the <strong>dims</strong> parameter, above); otherwise,
Geonomics will throw an Error. Defaults to a dummy filename that must be
changed.</p>
<p><strong>scale_min_val</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#minimum value to use to rescale the Layer to [0,1]</span>
<span class="s1">&#39;scale_min_val&#39;</span><span class="p">:</span>                <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This defines the minimum value (in the units of the variable represented by
the file you are reading in) to use when rescaling the file’s array to
values between 0 and 1. (This is done to satisfy the requirement that all
Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s have arrays in that interval). Defaults to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>
(in which case Geonomics will set it to the minimum value observed in this
file’s array). But note that you should put good thought into
this parameter, because it <em>won’t</em> necessarily be the minimum value
observed in the file; for example, if this file is being used
to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that will undergo environmental change
in your <cite>Model</cite>, causing its real-world values to drop
below this file’s minimum value, then you will probably want to set
this value to the minimum real-world value that will occur for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
during your <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> scenario, so that low values
that later arise on this <cite>Layer</cite> don’t get truncated at 0.</p>
<p><strong>scale_max_val</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#maximum value to use to rescale the Layer to [0,1]</span>
<span class="s1">&#39;scale_max_val&#39;</span><span class="p">:</span>                <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This defines the maximum value (in the units of the variable represented by
the file you are reading in) to use when rescaling the file’s array to
values between 0 and 1. (This is done to satisfy the requirement that all
Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s have arrays in that interval). Defaults to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>
(in which case Geonomics will set it to the maximum value observed in this
file’s array). But note that you should put good thought into
this parameter, because it <em>won’t</em> necessarily be the maximum value
observed in the file; for example, if this file is being used
to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that will undergo environmental change
in your <cite>Model</cite>, causing its real-world values to increase
above this file’s maximum value, then you will probably want to set
this value to the maximum real-world value that will occur for this
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> during your <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> scenario, so that high values that
later arise on this <cite>Layer</cite> don’t get truncated at 1.</p>
<p><strong>coord_prec</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#decimal-precision to use for coord-units (ulc &amp; res)</span>
<span class="s1">&#39;coord_prec&#39;</span><span class="p">:</span>                <span class="mi">5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 5</p>
<p>reset? P</p>
<p>This defines number of decimals to which to round upper-left corner
coordinates and resolution values read in from a raster file.
Because Geonomics requires equality of these values amongst all
input raster files, this allows the user to stipulate
the level of precision of their coordinate system, avoiding
false coordinate-system mismatch errors because of
arbitrary float imprecision.
(Note that for <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s for which change rasters will be read in,
the same coordinate precision value will be used for all input rasters.)</p>
<p><strong>units</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#units of this file&#39;s variable</span>
<span class="s1">&#39;units&#39;</span><span class="p">:</span>                       <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None</p>
<p>reset? P</p>
<p>This is an optional parameter providing a string-representation
of the units in which a raster file’s variable is expressed.
If provided, it will be used to label the colorbar on plots
of the raster’s <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>.</p>
</section>
<hr class="docutils" />
<section id="nlmpy">
<h5>nlmpy<a class="headerlink" href="#nlmpy" title="Permalink to this headline"></a></h5>
<p><strong>function</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#parameters for an &#39;nlmpy&#39;-type Layer</span>
<span class="s1">&#39;nlmpy&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#nlmpy function to use the create this Layer</span>
    <span class="s1">&#39;function&#39;</span><span class="p">:</span>                 <span class="s1">&#39;mpd&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code> that is the name of an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> function</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'mpd'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> function that should be used to generate
this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>’s array. (<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> is a Python package for
generating neutral landscape models; NLMs.) Defaults to <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'mpd'</span></span></code> (the
function for creating a midpoint-displacement NLM). Can be set to any other
<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code> that identifies a valid <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> function, but then the
remaining parameters in this section must be changed to the parameters
that that function needs, and <em>only</em> those parameters
(because they will be unpacked into this function,
i.e. passed on to it, at the time it is called.
(Visit the <a class="reference external" href="https://pypi.org/project/nlmpy/">Cheese Shop</a> for more
information about the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> package and available functions).</p>
<p><strong>nRow</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of rows (MUST EQUAL LAND DIMENSION y!)</span>
<span class="s1">&#39;nRow&#39;</span><span class="p">:</span>                     <span class="mi">20</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 20</p>
<p>reset? P</p>
<p>This defines the number of rows in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> array that is created.
As the capitalized reminder in the parameters file mentions, this must be
equal to the y-dimension of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>; otherwise, an error
will be thrown. Note that this parameter (as for the remaining parameters in
this section, other than the <strong>function</strong> parameter) is valid for the
default <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">mpd</span></span></code> function that is set by the
<strong>function</strong> parameter); if you are using a different <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code>
function to create this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> then this and the remaining parameters
must be changed to the parameters that that function needs,
and <em>only</em> those parameters (because they will be unpacked into that function,
i.e. passed on to it, at the time it is called).</p>
<p><strong>nCol</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of cols (MUST EQUAL LAND DIMENSION x!)</span>
<span class="s1">&#39;nCol&#39;</span><span class="p">:</span>                     <span class="mi">20</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 20</p>
<p>reset? P</p>
<p>This defines the number of columns in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> array that is created.
As the capitalized reminder in the parameters file mentions, this must be
equal to the x-dimension of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>; otherwise, an error
will be thrown. Note that this parameter (as for the remaining parameters in
this section, other than the <strong>function</strong> parameter) is valid for the
default <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">mpd</span></span></code> function that is set by the
<strong>function</strong> parameter); if you are using a different <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code>
function to create this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> then this and the remaining parameters
must be changed to the parameters that that function needs,
and <em>only</em> those parameters (because they will be unpacked into that function,
i.e. passed on to it, at the time it is called).</p>
<p><strong>h</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#level of spatial autocorrelation in element values</span>
<span class="s1">&#39;h&#39;</span><span class="p">:</span>                     <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the level of spatial autocorrelation in the element values
of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> array that is created.
Note that this parameter (and the remaining parameters in
this section, other than the <strong>function</strong> parameter) is valid for the
default <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code> function (<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">mpd</span></span></code>, which is set by the
<strong>function</strong> parameter); but if you are using a different <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">nlmpy</span></span></code>
function to create this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> then this and the remaining parameters
must be changed to the parameters that that function needs,
and <em>only</em> those parameters (because they will be unpacked into that function,
i.e. passed on to it, at the time it is called).</p>
</section>
</section>
<hr class="docutils" />
<section id="change">
<h4>Change<a class="headerlink" href="#change" title="Permalink to this headline"></a></h4>
<p><strong>change_rast</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#land-change event for this Layer</span>
<span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#array of file for final raster of event, or directory</span>
    <span class="c1">#of files for each stepwise change in event</span>
    <span class="s1">&#39;change_rast&#39;</span><span class="p">:</span>         <span class="s1">&#39;/PATH/TO/FILE.EXT&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{2d <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'/PATH/TO/FILE.EXT'</span></span></code></p>
<p>reset? Y</p>
<p>This defines either the final raster of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> change event
(with valid values being a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">numpy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> or a string pointing
to a valid raster file, i.e. a file that can be read by <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">rasterio</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">open</span></span></code>);
or the stepwise series of changes to be made over the course of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> change event (with the valid value being a string
pointing to a directory full of valid raster files).
Note that whether an array, a raster, or multiple rasters
are input, their dimensions must be equal to the dimensions of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
that is being changed (and hence to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> to which it belongs).
Also note that if a directory of stepwise-change rasters is provided, the
rasters’ filenames must begin with the integer timesteps at which they
should be used during the change event, followed by underscores. (For example,
files with the filenames ‘50_mat_2001.tif’, ‘60_mat_2011.tif’,
‘65_mat_2011.tif’ would be used at timesteps 50, 60, and 65 during a model.)
Defaults to a dummy file name that must be changed.</p>
<p><strong>start_t</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#starting timestep of event</span>
<span class="s1">&#39;start_t&#39;</span><span class="p">:</span>          <span class="mi">50</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 50</p>
<p>reset? P</p>
<p>This indicates the first timestep of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>-change event.
Defaults to 50, but should be set to suit your specific scenario.
If a directory of files is provided for the <strong>change_rast</strong> parameter,
then this must match the earliest timestep in that series of files
(as indicated by the integers at the beginning of the file names).</p>
<p><strong>end_t</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#ending timestep of event</span>
<span class="s1">&#39;end_t&#39;</span><span class="p">:</span>          <span class="mi">100</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 100</p>
<p>reset? P</p>
<p>This indicates the last timestep of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>-change event.
Defaults to 100, but should be set to suit your specific scenario.
If a directory of files is provided for the <strong>change_rast</strong> parameter,
then this must match the final timestep in that series of files
(as indicated by the integers at the beginning of the file names).</p>
<p><strong>n_steps</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of stepwise changes in event</span>
<span class="s1">&#39;n_steps&#39;</span><span class="p">:</span>          <span class="mi">5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 5</p>
<p>reset? P</p>
<p>This indicates the number of stepwise changes to use to model a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>-change event.
If the the <strong>change_rast</strong> parameter is a directory of files,
then the value of this parameter must be the number of files in that directory.
If the <strong>change_rast</strong> parameter is either an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> or a file name,
then the changes during the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>-change event
are linearly interpolated (cellwise for the whole <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>) to this
number of discrete, instantaneous <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> changes between
the starting and ending rasters. Thus, the fewer the number of
steps, the larger, magnitudinally, each change will be. So more
steps may be ‘better’, as it will better approximate change that is continuous
in time. However, there is a potenitally significant memory trade-off here:
The whole series of stepwise-changed arrays is computed when the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is created, then saved and used at the appropriate timestep
during each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> run (and if the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that is changing is used
by any <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> as a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> then each
intermediate <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is also calculated
when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is first built, which can be much more memory-intensive
because these are 3-dimensional arrays).
These objects take up memory, which may be limiting for larger
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>s and/or <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> objects. This often will not be a
major issue, but depending on your use case it could pose a problem, so
is worth considering.</p>
</section>
</section>
</section>
<hr class="docutils" />
<section id="community-and-species-parameters">
<h2>Community and Species parameters<a class="headerlink" href="#community-and-species-parameters" title="Permalink to this headline"></a></h2>
<section id="species">
<h3>Species<a class="headerlink" href="#species" title="Permalink to this headline"></a></h3>
<p><strong>spp_&lt;n&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#spp name (SPECIES NAMES MUST BE UNIQUE!)</span>
<span class="s1">&#39;spp_0&#39;</span> <span class="p">:</span>   <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">spp_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.
(Note that unlike most parameters, this parameter is
a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key, the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
must have a unique name (so that a parameterized
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> isn’t overwritten in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">ParametersDict</span></span></code> by a
second, identically-named <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> represents. Example valid values include: 0, ‘spp0’,
‘high-dispersal’, ‘C. fasciata’. Names default to
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">spp_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>, where n is a series of
integers starting from 0 and counting the number of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.</p>
<hr class="docutils" />
<section id="id1">
<h4>Init<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h4>
<p><strong>N</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#starting number of individs</span>
    <span class="s1">&#39;N&#39;</span><span class="p">:</span>                <span class="mi">250</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 250</p>
<p>reset? P</p>
<p>This defines the starting size of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Importantly, this
may or may not be near the stationary size of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> after
the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> has burned in, because that size will depend on the
carrying-capacity raster (set by the <strong>K</strong> parameter), and on
the dynamics of specific a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> (because of the interaction of
its various parameters).</p>
<p><strong>K_layer</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#name of the carrying-capacity Layer</span>
<span class="s1">&#39;K_layer&#39;</span><span class="p">:</span>         <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: ‘layer_0’</p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> to be used as the
carrying-capacity raster for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. The values of this
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>, multiplied by <strong>K_factor</strong>, should express
the carrying capacity at each cell, in number
of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s. Note that the sum of the values of the product of
this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> and <strong>K_factor</strong>
can serve as a rough estimate of the expected stationary
number of individuals of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>;
however, observed stationary size could vary
substantially depending on various other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> parameters (e.g. birth
and death rates and mean number of offspring per mating event) as well
as on stochastic events (e.g. failure to colonize, or survive in, all
habitable portions of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>).</p>
<p><strong>K_factor</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#multiplicative factor for carrying-capacity layer</span>
<span class="s1">&#39;K_factor&#39;</span><span class="p">:</span>         <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the factor by which the raster of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> indicated
by <strong>K_layer</strong> will be multiplied to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ carrying-
capacity raster. Because <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s’ rasters are constrained to [0,1],
this allows the user to stipulate that cells have carrying capacities in
excess of 1.</p>
<p><strong>msprime</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># params for 1+ msprime pops for sampling starting individs</span>
<span class="s1">&#39;msprime&#39;</span><span class="p">:</span>         <span class="p">{</span>
    <span class="c1"># index of msprime source pop</span>
    <span class="mi">0</span><span class="p">:</span> <span class="p">{</span>
       <span class="c1"># number of individs to sample from pop</span>
       <span class="mi">100</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># 1x2 coord pair or nx2 coord pairs of individs</span>
            <span class="s1">&#39;coords&#39;</span><span class="p">:</span>             <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
            <span class="c1"># valid kwargs dict for Model.add_individuals</span>
            <span class="c1"># &#39;source_msprime_params&#39; argument...</span>
            <span class="s1">&#39;recomb_rate&#39;</span><span class="p">:</span>         <span class="mf">0.5</span><span class="p">,</span>
            <span class="s1">&#39;mut_rate&#39;</span><span class="p">:</span>            <span class="mf">0.0001</span><span class="p">,</span>
            <span class="s1">&#39;demography&#39;</span><span class="p">:</span>          <span class="kc">None</span><span class="p">,</span>
            <span class="s1">&#39;population_size&#39;</span><span class="p">:</span>     <span class="kc">None</span><span class="p">,</span>
            <span class="s1">&#39;ancestry_model&#39;</span><span class="p">:</span>      <span class="kc">None</span><span class="p">,</span>
            <span class="s1">&#39;random_seed&#39;</span><span class="p">:</span>         <span class="kc">None</span><span class="p">,</span>
            <span class="p">}</span>
       <span class="p">},</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: values displayed above</p>
<p>reset? Y</p>
<p>This block of parameters can be used to set up one or more <cite>msprime</cite>-simulated
populations from which all starting <cite>Individuals</cite>
for a Geonomics <cite>Model</cite> are sourced (replacing the <cite>Individuals</cite> simulated
during a Geonomics burn-in).
Each <cite>dict</cite> keyed to a serial integer defines a separate population.
The only key in that <cite>dict</cite> is the number of <cite>Individuals</cite> to draw
from that population, and the value is a dict of parameters that
includes: 1.) the starting coordinates to be assigned to the <cite>Individuals</cite>
(either a single coordinate pair at which all <cite>Individuals</cite> should be
placed or an nx2 <cite>numpy.ndarray</cite> containing a coordinate pair for each
<cite>Individual</cite>; and note, these should be cell coordinates on the <cite>Landscape</cite>
array, not projected coordinates in any input rasters that may have been used!);
and 2.) a series of parameters that can be fed
as kwargs into the ‘source_msprime_params’ parameter of the
<cite>Model.add_individuals</cite> method. The <cite>msprime</cite> functionality
provided through the <cite>Model.add_individuals</cite> method is a wrapper
around <cite>msprime.sim_ancestry</cite>, but it only allows a certain subset
of parameters to be fed through to that <cite>msprime</cite> function (with the remaining
parameters, e.g., <cite>sequence_length</cite>, being drawn directly from the Geonomics
parameters file, to ensure compatibility);
see <cite>help(Model.add_individuals)</cite> for
details on which kwargs can be specified and used.
Also, please note that <cite>msprime</cite> is highly
flexible, allowing for many different combined choices of demography and ancestry models,
so it is the user’s responsibility to ensure that they do not stipulate
<cite>msprime</cite> parameters that lead to nonsensical scenarios
(e.g., generating a number of <cite>msprime</cite>-derived Geonomics Individuals
that is larger than the census size defined for <cite>msprime</cite> population).</p>
</section>
<hr class="docutils" />
<section id="mating">
<h4>Mating<a class="headerlink" href="#mating" title="Permalink to this headline"></a></h4>
<p><strong>repro_age</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;mating&#39;</span>    <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#age(s) at sexual maturity (if tuple, female first)</span>
    <span class="s1">&#39;repro_age&#39;</span><span class="p">:</span>            <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">)</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the age at which <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
can begin to reproduce. If the value provided is a 2-tuple of different
numbers (and the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> uses separate sexes), then the first
number will be used as females’ reproductive age, the second as males’.
If the value is 0, or <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s are capable
of reproduction from time of time.</p>
<p><strong>sex</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to assign sexes</span>
<span class="s1">&#39;sex&#39;</span><span class="p">:</span>                  <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s will be assigned separate sexes
that are used to ensure only male-female mating events.</p>
<p><strong>sex_ratio</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#ratio of males to females</span>
<span class="s1">&#39;sex_ratio&#39;</span><span class="p">:</span>            <span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1/1</p>
<p>reset? P</p>
<p>This defines the ratio of males to females (i.e. it will be converted to
a probability that an offspring is a male, which is used as the probability
of a Bernoulli draw of that offspring’s sex).</p>
<p><strong>R</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#intrinsic growth rate</span>
<span class="s1">&#39;R&#39;</span><span class="p">:</span>                    <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.5</p>
<p>reset? P</p>
<p>This defines a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ intrinsic growth rate, which is used
as the ‘R’ value in the spatialized logistic growth equation that
regulates population density (<span class="math notranslate nohighlight">\(\frac{\mathrm{d}
N_{x,y}}{\mathrm{d}t}=rN_{x,y}(1-\frac{N_{x,y}}{K_{x,y}})\)</span>).</p>
<p><strong>b</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#intrinsic birth rate (MUST BE 0&lt;=b&lt;=1)</span>
<span class="s1">&#39;b&#39;</span><span class="p">:</span>                    <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ intrinsic birth rate, which is
implemented as the probability that an identified potential mating
pair successfully produces offspring. Because this is a probability, as
the capitalized reminder in the parameters file mentions, this value must
be in the inclusive interval [0, 1].</p>
<p>NOTE: this may later need to be re-implemented to allow for spatial
variation in intrinsic rate (i.e.. expression of a birth-rate raster),
and/or for density-dependent birth as well as mortality</p>
<p><strong>n_births_dist_lambda</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#expectation of distr of n offspring per mating pair</span>
<span class="s1">&#39;n_births_distr_lambda&#39;</span><span class="p">:</span>      <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the lambda parameter for the Poisson distribution from
which a mating pair’s number of offspring is drawn (unless <strong>n_births_fixed</strong>
is set to True, in which case it defines the number of offspring
produced by each successful mating event). Hence, this is either the
expected  or exact value for the number of offspring born in a
successful mating event (depending on how <strong>n_births_fixed</strong> is set).</p>
<p><strong>n_births_fixed</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether n births should be fixed at n_births_dist_lambda</span>
<span class="s1">&#39;n_births_fixed&#39;</span><span class="p">:</span>           <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? P</p>
<p>This determines whether or not the number of births for each mating event will
be fixed. If set to true, each successful mating event will produce
<strong>n_births_distr_lambda</strong> new offspring.</p>
<p><strong>mating_radius</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#radius of mate-search area (None, for panmixia)</span>
<span class="s1">&#39;mating_radius&#39;</span><span class="p">:</span>        <span class="mi">1</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: 1</p>
<p>reset? Y</p>
<p>This defines the radius within which an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Indvidual</span></span></code> can find a mate.
This radius is provided to queries run on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_KDTree</span></span></code> object.
(If set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> then true panmixia will be used, i.e. each
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>, with probability equal to its <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ birth
rate, will choose any other individual in the population as its mate,
after which the chosen pair will then go through age- and sex-eligibility
checks as needed given the parameterization.)</p>
<p><strong>choose_nearest_mate</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether individs should choose nearest neighs as mates</span>
<span class="s1">&#39;choose_nearest_mate&#39;</span><span class="p">:</span>     <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">py</dt>
<dd class="field-odd"><p><cite>bool</cite></p>
</dd>
</dl>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether or each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>  will always
choose its nearest neighbor as a mate.
Defaults to False, allowing each focal <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> to
randomly choose from among all other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individauls</span></span></code> occurring
within its mating radius.
(In that case, if <strong>inverse_dist_mating</strong> is False then all other nearby
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individuals</span></span></code> will have equal probability of being chosen;
if <strong>inverse_dist_**mating</strong> is True, then other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individuals</span></span></code> will
have probabilities linearly related to their inverse distance from the
focal <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>.)
(Note that this parameter will only
be used if <strong>mating_radius</strong> is not <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.)</p>
<p><strong>inverse_dist_mating</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether mate-choice should be inverse distance-weighted</span>
<span class="s1">&#39;inverse_dist_mating&#39;</span><span class="p">:</span>     <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">py</dt>
<dd class="field-odd"><p><cite>bool</cite></p>
</dd>
</dl>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether or each focal <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>
will use the inverse of the distance between itself and all
other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individauls</span></span></code> occurring within its mating radius
to weight the mutually exclusive probabilites of choosing
each of those other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individuals</span></span></code> as a mate.
If False, then each other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> within the
focal <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s mating radius has a uniform probability
of being chosen as a mate.
(Note that this parameter will only
be used if <strong>choose_nearest_mate</strong> is False and <strong>mating_radius</strong>
is not <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.)</p>
</section>
<hr class="docutils" />
<section id="mortality">
<h4>Mortality<a class="headerlink" href="#mortality" title="Permalink to this headline"></a></h4>
<p><strong>max_age</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#maximum age</span>
<span class="s1">&#39;max_age&#39;</span><span class="p">:</span>              <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the maximum age an individual can achieve before being
forcibly culled from the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Defaults to 1 (which will create
a Wright-Fisher-like simulation, with discrete generations). Can be set
to any other age, or can be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> (in which case no maxmimum
age is enforced).</p>
<p><strong>d_min</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#min P(death) (MUST BE 0&lt;=d_min&lt;=1)</span>
<span class="s1">&#39;d_min&#39;</span><span class="p">:</span>                     <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0</p>
<p>reset? N</p>
<p>This defines the minimum probabilty of death that an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>
can face each time its Bernoulli death-decision is drawn. Because this
is a probability, as the capitalized reminder in
the parameters file mentions, this value must be in the
inclusive interval [0, 1].</p>
<p><strong>d_max</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#max P(death) (MUST BE 0&lt;=d_max&lt;=1)</span>
<span class="s1">&#39;d_max&#39;</span><span class="p">:</span>                    <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This defines the minimum probabilty of death that an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>
can face each time its Bernoulli death-decision is drawn. Because this
is a probability, as the capitalized reminder in
the parameters file mentions, this value must be in the
inclusive interval [0, 1].</p>
<p><strong>density_grid_window_width</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;mortality&#39;</span>     <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#width of window used to estimate local pop density</span>
    <span class="s1">&#39;dens_grid_window_width&#39;</span><span class="p">:</span>   <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None</p>
<p>reset? N</p>
<p>This defines the width of the window used by the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_DensityGridStack</span></span></code>
to estimate a raster of local <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> densities. The user should
feel free to set different values for this parameter (which could be
especially helpful when calling <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_density</span></span></code> to inspect the
resulting surfaces calculated at different window widths, if trying
to heuristically choose a reasonable value to set for a
particular simulation scenario). But be aware that choosing particularly
small window widths (in our experience, windows smaller than ~1/20th of
the larger <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimension) can cause dramatic increases in the
run-time of the density calculation (which runs twice per timestep).
Defaults to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, which will internally be set to the integer
nearest to 1/10th of the larger <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimension;
for many purposes this will work, but in some cases
the user may wish to control this.</p>
</section>
<hr class="docutils" />
<section id="movement">
<h4>Movement<a class="headerlink" href="#movement" title="Permalink to this headline"></a></h4>
<p><strong>move</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether or not the species is mobile</span>
<span class="s1">&#39;move&#39;</span><span class="p">:</span>                    <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">py</dt>
<dd class="field-odd"><p><cite>bool</cite></p>
</dd>
</dl>
<p>default: True</p>
<p>reset? P</p>
<p>This determines whether the :py: <cite>Species</cite> being parameterized is mobile
(i.e. whether its individuals should move). A <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> without movement
will still undergo dispersal of offspring, but after dispersing
those offspring will remain fixed in location until death.</p>
<p><strong>direction_distr_mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="p">{</span>
     <span class="c1">#mode of distr of movement direction</span>
     <span class="s1">&#39;direction_distr_mu&#39;</span><span class="p">:</span>      <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, :py;`float`}</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This is the <span class="math notranslate nohighlight">\(\mu\)</span> parameter of the VonMises distribution
(a circularized normal distribution) from which
movement directions are chosen when movement is random and isotropic
(rather than
being determined by a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>;
if a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>
is being usen this parameter is ignored). The <span class="math notranslate nohighlight">\(\kappa\)</span> value
that is fed into this same distribution (<strong>direction_distr_kappa</strong>)
causes it to be very dispersed,
such that the distribution is effectively a uniform distribution on
the unit circle (i.e. all directions are effectively equally probable).
For this reason, changing this parameter without changing the
<strong>direction_distr_kappa</strong> value also, will make no change in the directions
drawn for movement.  If random, isotropic
movement is what you aim to model then there is probably little reason
to change these parameters.</p>
<p><strong>direction_distr_kappa</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#concentration of distr of movement direction</span>
<span class="s1">&#39;direction_distr_kappa&#39;</span><span class="p">:</span>  <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0</p>
<p>reset? N</p>
<p>This is the <span class="math notranslate nohighlight">\(\kappa\)</span> parameter of the VonMises distribution
(a circularized normal distribution) from which
movement directions are chosen when movement is random and isotropic
(rather than
being determined by a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>;
if a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>
is being usen this parameter is ignored). The default value of 0 will
cause this distribution to be very dispersed, approximating a uniform
distribution on the unit circle and rendering the <span class="math notranslate nohighlight">\(\mu\)</span>
value (<strong>direction_distr_mu</strong>) effectively meaningless. However, as this
parameter’s value increases the resulting circular distributions will become
more concentrated around <span class="math notranslate nohighlight">\(\mu\)</span>, making the value fed to
<strong>direction_distr_mu</strong> influential. If random, isotropic
movement is what you aim to model then there is probably little reason
to change these parameters.</p>
<p><strong>movement_distance_distr_param1</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#1st param of distr of movement distance</span>
<span class="s1">&#39;movement_distance_distr_param1&#39;</span><span class="p">:</span>      <span class="mf">0.1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.1</p>
<p>reset? Y</p>
<p>This is the first parameter of the distribution used to draw
movement distances.
The values generated by the movement distribution will
be expressed in units of raster-cell widths.
This paramter and <strong>movement_distance_distr_param2</strong> should be
set to reflect a distribution of movement distances that is appropriate
for your scenario.
The distribution to which this parameter applies depends on the value
of the <strong>movement_distance_distr</strong> parameter.</p>
<p><strong>movement_distance_distr_param2</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#2nd param of distr of movement distance</span>
<span class="s1">&#39;movement_distance_distr_param2&#39;</span><span class="p">:</span>   <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the second parameter of the distribution used to draw
movement distances.
The values generated by the movement distribution will
be expressed in units of raster-cell widths.
This paramter and <strong>movement_distance_distr_param1</strong> should be
set to reflect a distribution of movement distances that is appropriate
for your scenario.
The distribution to which this parameter applies depends on the value
of the <strong>movement_distance_distr</strong> parameter.</p>
<p><strong>movement_distance_distr</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#movement distance distr to use (&#39;lognormal&#39;,&#39;levy&#39;,&#39;wald&#39;)</span>
<span class="s1">&#39;movement_distance_distr&#39;</span><span class="p">:</span>    <span class="s1">&#39;lognormal&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: ‘lognormal’</p>
<p>reset? Y</p>
<p>This determines whether movement is modeled using a lognormal distribution
(‘lognormal’; default), a Lévy distribution (‘levy’),
or a Wald distribution (‘wald’).</p>
<p><strong>dispersal_distance_distr_param1</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#1st param of distr of dispersal distance</span>
<span class="s1">&#39;dispersal_distance_distr_param1&#39;</span><span class="p">:</span>     <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: -1</p>
<p>reset? Y</p>
<p>This is the first parameter of the distribution used to draw
dispersal distances.
The values generated by the dispersal distribution will
be expressed in units of raster-cell widths.
This paramter and <strong>dispersal_distance_distr_param2</strong> should be
set to reflect a distribution of dispersal distances that is appropriate
for your scenario.
The distribution to which this parameter applies depends on the value
of the <strong>dispersal_distance_distr</strong> parameter.</p>
<p><strong>dispersal_distance_distr_param2</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#2nd param of distr of dispersal distance</span>
<span class="s1">&#39;dispersal_distance_distr_param2&#39;</span><span class="p">:</span>  <span class="mf">0.05</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.05</p>
<p>reset? Y</p>
<p>This is the second parameter of the distribution used to draw
dispersal distances.
The values generated by the dispersal distribution will
be expressed in units of raster-cell widths.
This paramter and <strong>dispersal_distance_distr_param1</strong> should be
set to reflect a distribution of dispersal distances that is appropriate
for your scenario.
The distribution to which this parameter applies depends on the value
of the <strong>dispersal_distance_distr</strong> parameter.</p>
<p><strong>dispersal_distance_distr</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#dispersal distance distr to use (&#39;lognormal&#39;,&#39;levy&#39;,&#39;wald&#39;)</span>
<span class="s1">&#39;dispersal_distance_distr&#39;</span><span class="p">:</span>    <span class="s1">&#39;lognormal&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: ‘lognormal’</p>
<p>reset? Y</p>
<p>This determines whether dispersal is modeled using a lognormal distribution
(‘lognormal’; default), a Lévy distribution (‘levy’),
or a Wald distribution (‘wald’).</p>
<hr class="docutils" />
<section id="movement-and-dispersal-conductancesurfaces">
<h5>Movement and Dispersal _ConductanceSurfaces<a class="headerlink" href="#movement-and-dispersal-conductancesurfaces" title="Permalink to this headline"></a></h5>
<p><strong>layer</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;move_surf&#39;</span>     <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#move-surf Layer name</span>
    <span class="s1">&#39;layer&#39;</span><span class="p">:</span>                <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'layer_0'</span></span></code></p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> to be used as to construct the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Note that this can also
be thought of as the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that should serve as a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ permeability raster (because <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s moving
on this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> toward the higher
(if mixture distributions are used) or highest
(if unimodl distributions are used) values in their neighborhoods).</p>
<p><strong>mixture</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to use mixture distrs</span>
<span class="s1">&#39;mixture&#39;</span><span class="p">:</span>              <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> should be built using
VonMises mixture distributions or unimodal VonMises distributions.
If True, each cell in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> will have an approximate
circular distribution that is a
weighted sum of 8 unimodal VonMises distributions (one per cell in the 8-cell
neighborhood); each of those summed unimodal distributions will have as its
mode the direction of the neighboring cell on which it is based and as its
weight the relative permeability of the cell on which it is based
(relative to the full neighborhood). If False, each cell in the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> will have an approximated circular distribution
that is a single
VonMises distribution with its mode being the direction of the maximum-valued
cell in the 8-cell neighborhood and its concentration determined by
<strong>vm_distr_kappa</strong>.</p>
<p><strong>vm_distr_kappa</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#concentration of distrs</span>
<span class="s1">&#39;vm_distr_kappa&#39;</span><span class="p">:</span>       <span class="mi">12</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 12</p>
<p>reset? N</p>
<p>This sets the concentration of the VonMises distributions used to build
the approximated circular distributions in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>.
The default value was chosen heuristically as one that provides a reasonable
concentration in the direction of a unimodal VonMises distribution’s mode
without causing VonMises mixture distributions built from an
evenly weighted sum of distributions pointing toward the
8-cell-neighborhood directions to have 8 pronounced modes.
There will probably be little need to change the default value, but if
interested then the user could create <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>s with various values
of this parameter and then use the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_movement_surface</span></span></code>
method to explore the influence of the parameter on the resulting
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>s.</p>
<p><strong>approx_len</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#length of approximation vectors for distrs</span>
<span class="s1">&#39;approx_len&#39;</span><span class="p">:</span>       <span class="mi">5000</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 5000</p>
<p>reset? P</p>
<p>This determines the length of the vector of values used to approximate each
distribution on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> (i.e. the size of the z-axis
of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> used to hold all the distribution-approximations, where
the y and x axes have the same dimensions as the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>). The default
value of 5000 is fine for many cases, but may need to be
reduced depending on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimensions (because for a larger
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, say 1000x1000 cells, it would create a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> that is roughly 4Gb,
and if the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> on which the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is based will be
undergoing landscape changes then numerous versions of an object of this size
would need to be generated when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is built and held in memory).
The value to use for this parameter will depend on the size of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, the exact scenario being simulated, and the memory of the
machine on which the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is to be run.</p>
</section>
</section>
<hr class="docutils" />
<section id="genomicarchitecture">
<h4>_GenomicArchitecture<a class="headerlink" href="#genomicarchitecture" title="Permalink to this headline"></a></h4>
<p><strong>gen_arch_file</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;gen_arch&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#file defining custom genomic arch</span>
    <span class="s1">&#39;gen_arch_file&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: {<code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'&lt;your_model_name&gt;_spp-&lt;n&gt;_gen_arch.csv'</span></span></code></p>
<p>reset? P</p>
<p>This argument indicates whether a custom genomic architecture file should
be used to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and if so,
where that file is located. If the value is <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, no file will be
used and the values of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ other genomic
architecture parameters in the parameters file will be used to create
the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>. If the value is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code> pointing to a
custom genomic-architecture file
(i.e. a CSV file with loci as rows and ‘locus’,
‘p’, ‘dom’, ‘r’, ‘trait’, and ‘alpha’ as columns stipulating the
locus numbers, starting
allele frequencies, dominance values, inter-locus recombination rates,
trait names, and effect sizes of all loci; values can be left blank if not applicable).
Geonomics will create an empty
file of this format for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> for which the
‘genomes’ argument is given the value ‘custom’ when
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> is called (which will be saved as
‘&lt;your_model_name&gt;_spp-&lt;n&gt;_gen_arch.csv’).</p>
<p>Note that when Geonomics reads in a custom genomic architecture file
to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>, it will check
that the length (i.e. number of rows) in this file is equal to the length
stipulated by the <strong>L</strong> parameter, and will also check that the first value
at the top of the ‘r’ column is 0 (which is used to implement independent
assortment during gametogenesis). If either of these checks fails,
Geonomics throws an Error.</p>
<p><strong>L</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num of loci</span>
<span class="s1">&#39;L&#39;</span><span class="p">:</span>                        <span class="mi">1000</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 1000</p>
<p>reset? P</p>
<p>This defines the total number of loci in the genomes in a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.</p>
<p><strong>l_c</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num of chromosomes</span>
<span class="s1">&#39;l_c&#39;</span><span class="p">:</span>                      <span class="p">[</span><span class="mi">100</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[</span></span><span class="literal number integer"><span class="pre">100</span></span><span class="punctuation"><span class="pre">]</span></span></code></p>
<p>reset? P</p>
<p>This defines the lengths (in number of loci) of each of the chromosomes
in the genomes in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.  Note that the sum of this <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code>
must equal <strong>L</strong>, otherwise Geonomics will throw an Error.
Also note that Geonomics models genomes as single <strong>L</strong> x 2
arrays, where separate chromosomes are delineated by points along
the genome where the recombination rate is 0.5;
thus, for a model where recombination rates are often at or near 0.5, this
parameter will have little meaning.</p>
<p><strong>start_p_fixed</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#fixed starting allele freq; None/False -&gt; rand; True -&gt; 0.5</span>
<span class="s1">&#39;start_p_fixed&#39;</span><span class="p">:</span>                      <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? P</p>
<p>If a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> on the interval [0,1] is provided,
that value will be used as the starting allele
frequency at which all loci (except neutral loci,
if <strong>start_neut_zero</strong> is True) will be fixed.
If <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, the starting allele
frequency of each locus will be drawn as a uniform random variable
between 0 and 1, inclusive.
If a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code> is provided, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">True</span></span></code> will fix all
loci at the default starting allele frequency of 0.5,
whereas <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">False</span></span></code> will not fix starting
allele frequencies, effectively having the same effect
as <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.
Defaults to 0.5.</p>
<p><strong>start_neut_zero</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to start neutral locus freqs at 0</span>
<span class="s1">&#39;start_neut_zero&#39;</span><span class="p">:</span>           <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>If True, all neutral loci will start with the ‘1’ allele at a frequency of 0
(i.e. all individuals will be homozygous ‘0’|’0’ at those loci).</p>
<p><strong>mu_neut</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#genome-wide per-base neutral mut rate (0 to disable)</span>
<span class="s1">&#39;mu_neut&#39;</span><span class="p">:</span>                  <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1e-9</p>
<p>reset? P</p>
<p>This defines the genome-wide per-base neutral mutation rate.
This value can be set to 0 to disable neutral mutation.</p>
<p><strong>mu_delet</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#genome-wide per-base deleterious mut rate (0 to disable)</span>
<span class="s1">&#39;mu_delet&#39;</span><span class="p">:</span>                 <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the genome-wide per-base deleterious mutation rate.
This value can be set to 0 to disable deleterious mutation. Note that all
deleterious mutation will fall outside the loci that affect any <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>s
a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> may have, and will behave simply as globally
deleterious mutations (i.e. mutations that reduce the mutated
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s fitness regardless of that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s
spatial location).</p>
<p><strong>delet_alpha_distr_shape</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#shape of distr of deleterious effect sizes</span>
<span class="s1">&#39;delet_alpha_distr_shape&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines the shape parameter of the gamma distribution from which
the effect sizes of deleterious loci are drawn. (Values drawn will be
truncated to the interval [0,1].)</p>
<p><strong>delet_alpha_distr_scale</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#scale of distr of deleterious effect sizes</span>
<span class="s1">&#39;delet_alpha_distr_scale&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines the scale parameter of the gamma distribution from which
the effect sizes of deleterious loci are drawn. (Values drawn will be
truncated to the interval [0,1].)</p>
<p><strong>r_distr_alpha</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#alpha of distr of recomb rates</span>
<span class="s1">&#39;r_distr_alpha&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None</p>
<p>reset? P</p>
<p>This defines the alpha parameter of the beta distribution from which
interlocus recombination rates are drawn. (Values drawn will be truncated to
the interval [0, 0.5].)
If <strong>r_distr_beta</strong> is None, recombination rates will be fixed at this value.
Defaults to None, and <strong>r_distr_beta</strong> defaults to None,
such that all recombination rates will be fixed at a value (1/<strong>L</strong>)
that yields approximately 1 expected recombination event
per gamete per generation.</p>
<p><strong>r_distr_beta</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#beta of distr of recomb rates</span>
<span class="s1">&#39;r_distr_beta&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None,</p>
<p>reset? P</p>
<p>This defines the beta parameter of the beta distribution from which
interlocus recombination rates are drawn. (Values drawn will be truncated to
the interval [0, 0.5].) Defaults to None, which will fix recombination rates
at the value of <strong>r_distr_alpha</strong> (which defaults to 0.5, i.e. independence),
or else will fix all rates at a value (1/<strong>L</strong>) that yields
approximately 1 expected recombination event per gamete per generation
(if <strong>r_distr_alpha</strong> is None).</p>
<p><strong>dom</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether loci should be dominant (for allele &#39;1&#39;)</span>
<span class="s1">&#39;dom&#39;</span><span class="p">:</span>                      <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether loci should be treated as dominant (if True)
for the ‘1’ allele  or as codominant (if False). Codominance is the default
behavior, because it is assumed that Geonomics will often be used
to model quantitative traits, for which this is a reasonable assumption.</p>
<p><strong>pleiotropy</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to allow pleiotropy</span>
<span class="s1">&#39;pleiotropy&#39;</span><span class="p">:</span>               <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether pleiotropy should be allowed. If True, loci will be
permitted to contribute to more than one <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>.</p>
<p><strong>recomb_rate_custom_fn</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#custom fn for drawing recomb rates</span>
<span class="s1">&#39;recomb_rate_custom_fn&#39;</span><span class="p">:</span>    <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">function</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This parameter allows the user to provide a custom function according to which
interlocus recombination rates will be assigned. If set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, the
default behavior (i.e. recombination rates chosen from a beta distribution
using <strong>r_distr_alpha</strong> and <strong>r_distr_beta</strong>) will be used.</p>
<p><strong>n_recomb_paths_mem</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of recomb paths to hold in memory</span>
<span class="s1">&#39;n_recomb_paths_mem&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mf">1e4</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">(</span></span><span class="literal number float"><span class="pre">1e4</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset? P</p>
<p>This defines the maximum number of recombination paths for Genomics to hold in
memory at one time. Geonomics models recombination by using the interlocus
recombination rates to draw a large number of recombination ‘paths’
along the Lx2 genome array (when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is first built), and
then shuffling and cycling through those recombination paths as
needed during <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> runs. Of the total number of paths created, some
subset will be held in memory (the number of these is defined by
this parameter), while the remainder will live in a temporary
file (which is occasionally read in whenever the paths in memory are close to
being used up). Thus, to avoid problems, the number provided to this parameter
should be comfortably larger than the largest anticipated number of
recombination paths that will be needed during a single mating event (i.e.
larger than two times the largest antipicated number of offspring to be born
to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> during one timestep).</p>
<p><strong>n_recomb_paths_tot</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#total number of recomb paths to simulate</span>
<span class="s1">&#39;n_recomb_paths&#39;</span><span class="p">:</span>           <span class="nb">int</span><span class="p">(</span><span class="mf">1e5</span><span class="p">),</span>
</pre></div>
</div>
<p>This defines the total number of recombination paths that Geonomics will
generate. Geonomics models recombination by using the interlocus
recombination rates to draw a large number of recombination ‘paths’
along the Lx2 genome array (when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is first built), and
then shuffling and cycling through those recombination paths as
needed during <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> runs. The larger the total number of these paths
that is created, the more closely Geonomics will model truly
free recombination and the more prceisely it will model the exact
interlocus recombination rates defined in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>.</p>
<p><strong>allow_ad_hoc_recomb</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to generate recombination paths at each timestep</span>
<span class="s1">&#39;allow_ad_hoc_recomb&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether or not recombinants should be drawn each timestep
(rather than recombination paths being drawn and stored when a model is first
built, then used randomly throught the model run).
This is advantageous because it models recombination exactly (rather than
approximating recombination by drawing some number of fixed recombination paths
that get repeatedly used), and for combinations of larger genome sizes (L) and
larger mean population sizes (N) it avoids the memory used by storing so many
recombination paths drawn at model creation, thus making these
parameterizations feasible on computers with memory limitations).
It is disadvantageous, however, because it runs somewhat slower than the
default approach (recombinants drawn at model creation) for a range of
L and N values, and also because it is only available for parameterizations
with homogeneous recombination across the genome.</p>
<p><strong>jitter_breakpoints</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to jitter recomb bps, to correctly track num_trees</span>
<span class="s1">&#39;jitter_breakpoints&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether or not the recombination breakpoints
stored by the tskit TableCollection should be slightly jitted off of their
x.5 default positions. Enabling this will render each recombination effectively
unique, and thus will allow the tskit.TreeSequence to correctly report the
number of trees (TreeSequence.num_trees). However, it will do this
at the expense of additional memory usage, which could potentially be limiting.
Thus, this should be set to <cite>True</cite> if and only if the TreeSequence
information will be used in a way that requires
accurate representation of the number of trees.</p>
<p><strong>mut_log</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to save mutation logs</span>
<span class="s1">&#39;mut_log&#39;</span><span class="p">:</span>                  <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the location of the mutation-log file where  Geonomics should
save a record of each mutation that occurs for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>, for each iteration. If <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, no mutation log
will be created and written to.</p>
<p><strong>use_tskit</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to use tskit (to record full spatial pedigree)</span>
<span class="s1">&#39;use_tskit&#39;</span><span class="p">:</span>                  <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">True</span></span></code></p>
<p>reset? P</p>
<p>This indicates whether Geonomics should use the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span></code> API
to store individuals’ genomes in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">TableCollection</span></span></code>.
If <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span></code> is used the the full spatial pedigree of the current population
will be available at any time, and each Individual will carry a numpy.ndarray
containing just its genotypes for any non-neutral loci.
If <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span></code> is not used, each Individual will carry a numpy.ndarray
containing its genotypes for all simulated loci (neutral and non-neutral).
Defaults to using <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span></code> (True),
but may need to be switched to False for models
using large numbers of independent loci
and not needing access to the spatial pedigree
(because larger numbers of independent loci cause a geometric
expansion of the number of trees stored in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">TreeSequence</span></span></code>,
requiring significantly more memory overall,
and significantly greater runtime to store them at each time step).</p>
<p><strong>tskit_simp_interval</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#time step interval for simplification of tskit tables</span>
<span class="s1">&#39;tskit_simp_interval&#39;</span><span class="p">:</span>          <span class="mi">100</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 100</p>
<p>reset? N</p>
<p>This sets the interval, in timesteps,
between subsequent <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">tskit</span></span></code> simplifications.
Defaults to simplifying every 100 timesteps, as suggested by the
tskit package authors (see
<a class="reference external" href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006581">here</a>
). This most likely need not be changed, but for simulations
with especially large population and/or genome sizes the user may
wish to experiment with reducing this interval so as to improve performance.</p>
<hr class="docutils" />
<section id="traits">
<h5>Traits<a class="headerlink" href="#traits" title="Permalink to this headline"></a></h5>
<p><strong>trait_&lt;n&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#trait name (TRAIT NAMES MUST BE UNIQUE!)</span>
<span class="s1">&#39;trait_0&#39;</span> <span class="p">:</span>   <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">trait_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>.
(Note that unlike most parameters, this parameter is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key,
the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
must have a unique name (so that a parameterized
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> isn’t overwritten in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">ParametersDict</span></span></code> by a
second, identically-named <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> represents. Example valid values include: 0, ‘trait0’,
‘tmp_trait’, ‘bill length’. Names default to <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">trait_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>,
where n is a series of integers starting from 0 and counting the
number of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>s for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.</p>
<p><strong>layer</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#trait-selection Layer name</span>
<span class="s1">&#39;layer&#39;</span><span class="p">:</span>                <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'layer_0'</span></span></code></p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that serves as the selective force
acting on this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>. (For example, if this Trait is selected upon by
annual mean temperature, then the name of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
representing annual mean temperature should be provided here.)</p>
<p><strong>phi</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#phenotypic selection coefficient</span>
<span class="s1">&#39;phi&#39;</span><span class="p">:</span>                  <span class="mf">0.05</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s}</p>
<p>default: 0.05</p>
<p>reset? P</p>
<p>This defines the phenotypic selection coefficient on this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> (i.e
the selection coefficient acting on the phenotypes, rather than the genotypes,
of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>). The effect of this value can be thought of as the
reduction (from 1) in an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s survival probability when that
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> is maximally unfit (i.e. when that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> has a
phenotypic value of 1.0 but is located in a location with an environmental
value of 0.0, or vice versa). When the value is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> then the
strength of selection will be the same for all locations on the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>. When the value is an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> of
equal dimensions to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> then the strength of
selection will vary across space, as indicated by the values in this array
(what Geonomics refers to as a “spatially contingent” selection regime).
Importantly, most individuals will experience selection on a given trait that
is only a fraction of the strength dictated by this parameter,
because an individual’s fitness for a given trait
is determined by the product of the trait’s selection coefficient
and the individual’s degree of mismatch to its local environment,
and it is often unlikely that individuals would occur in environments
that are completely opposed to those individuals’ phenotypic values
(e.g. a 0-valued individual in a 1-valued environmental cell).
Because of this, selection coefficients that would be considered ‘strong’
in classical, aspatial population genetics models will tend to behave
less strongly in Geonomics models.
(For mathematical detail, see the <a class="reference internal" href="overview.html#selection"><span class="std std-ref">Selection</span></a> section.)</p>
<p><strong>n_loci</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of loci underlying trait</span>
<span class="s1">&#39;n_loci&#39;</span><span class="p">:</span>               <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 10</p>
<p>reset? P</p>
<p>This defines the number of loci that should contribute to the phenotypes
of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>. These loci will be randomly drawn from across the
genome.</p>
<p><strong>mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#mutation rate at loci underlying trait</span>
<span class="s1">&#39;mu&#39;</span><span class="p">:</span>                   <span class="mf">1e-9</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1e-9</p>
<p>reset? P</p>
<p>This defines the mutation rate for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> (i.e. the rate at which
mutations that affect the phenotypes of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> will arise). Set to
0 to disable mutation for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>.</p>
<p><strong>alpha_distr_mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of effect sizes</span>
<span class="s1">&#39;alpha_distr_mu&#39;</span> <span class="p">:</span>      <span class="mf">0.1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.1</p>
<p>reset? N</p>
<p>This defines the mean of the normal distribution from which a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>’s
initially parameterized loci and new mutations’ effect sizes are drawn (with
the exception of monogenic traits, whose starting locus always has an alpha
value of 0.5, but whose later mutations are influenced by this parameter).
For effect sizes drawn from a distribution, it is recommended
to set this value set to 0 and adjust <strong>alpha_distr_sigma</strong>.
For fixed effect sizes, set this value to the fixed
effect size and set <strong>alpha_distr_sigma</strong> to 0; effects will alternate
between positive and negative when they are assigned to loci.
In either case, new mutations in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
will then be equally likely to decrease or increase <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’
phenotypes from the multigenic baseline phenotype of 0.5 (which is also
the central value on a Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>).
It is also recmmended that the user consider the number of loci for a trait
when setting the fixed or distributed effect sizes; for example, for a trait
with 10 underlying loci, an average or fixed absolute effect size of 0.1
will enable phenotypes that cover the range of values on a
Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> (i.e. phenotypes 0 &lt;= z &lt;= 1), whereas
0.05 will likely not enable that full range of phenotypes, and 0.5 will
generate many phenotypes that fall outside that range and will be selected
against at all locations on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>.</p>
<p><strong>alpha_distr_sigma</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of effect size</span>
<span class="s1">&#39;alpha_distr_sigma&#39;</span><span class="p">:</span>    <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the standard deviation of the normal distribution from which
a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>’s new mutations’ effect sizes are drawn.
For effect sizes drawn from a distribution, it is recommended
to set this value set to some nonzero number
and set <strong>alpha_distr_mu</strong> to 0. For fixed effect sizes,
set this value to 0 and set <strong>alpha_distr_mu</strong> to the fixed effect size;
effects will alternate between positive and negative when they are
assigned to loci. In either case, new mutations in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
will then be equally likely to decrease or increase <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’
phenotypes from the multigenic baseline phenotype of 0.5 (which is also
the central value on a Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>).</p>
<p><strong>max_alpha_mag</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#max allowed magnitude for an alpha value</span>
<span class="s1">&#39;max_alpha&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: None</p>
<p>reset? P</p>
<p>This defines the maximum value that can be drawn for a locus’ effect size
(i.e. alpha). Defaults to None, but the user may want to set this to some
reasonable value, to prevent chance creation of loci with extreme effects.</p>
<p><strong>gamma</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#curvature of fitness function</span>
<span class="s1">&#39;gamma&#39;</span><span class="p">:</span>                <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This defines the curvature of the fitness function (i.e.
how fitness decreases as the absolute difference between an
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s optimal and actual phenotypes increases). The user
will probably have no need to change this from the default value of 1
(which causes fitness to decrease linearly around the optimal
phenotypic value). Values &lt; 1 will cause the fitness function to be
concave up; values &gt; 1 will cause it to be concave down.</p>
<p><strong>univ_adv</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether the trait is universally advantageous</span>
<span class="s1">&#39;univ_adv&#39;</span><span class="p">:</span>             <span class="kc">False</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether whether selection on a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> should be
universal (i.e. whether a phenotype of 1 should be optimal everywhere
on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>). When set to True, selection of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
will be directional on the entire <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>, regardless
of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’ spatial contexts.</p>
</section>
</section>
<hr class="docutils" />
<section id="species-change">
<h4>Species change<a class="headerlink" href="#species-change" title="Permalink to this headline"></a></h4>
<section id="demographic-change">
<h5>Demographic change<a class="headerlink" href="#demographic-change" title="Permalink to this headline"></a></h5>
<p><strong>kind</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#kind of event {&#39;monotonic&#39;, &#39;stochastic&#39;,</span>
<span class="c1">#&#39;cyclical&#39;, &#39;custom&#39;}</span>
<span class="s1">&#39;kind&#39;</span><span class="p">:</span>             <span class="s1">&#39;monotonic&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'monotonic'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'stochastic'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cyclical'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'custom'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'monotonic'</span></span></code></p>
<p>reset? P</p>
<p>This indicates what type of demographic change is being parameterized.
Each event has a certain length (in timesteps; defined by the <strong>start</strong> and
<strong>end</strong> parameters). Note that of the other parameters in this section, only
those that are necessary to parameterize the type of change event indicated
here will be used.</p>
<p>In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'monotonic'</span></span></code> change events, a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’
carrying capacity raster (K) is multiplied by a constant factor
(<strong>rate</strong>) at each timestep during the event.
In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'stochastic'</span></span></code> change events, K fluctuates
around the baseline value (i.e. the K-raster at the time that the change event
begins) at each required timestep during the event (where the sizes of the
fluctuations are drawn from the distribution indicated by
<strong>distr</strong>, the floor and ceiling on those sizes are set by
<strong>size_range</strong>, and the required timesteps are determined by <strong>interval</strong>).
In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cyclical'</span></span></code> change events, K undergoes a number (indicated
by <strong>n_cycles</strong>) of sinusoidal cycles between some minimum and maximum
values (indicated by <strong>size_range</strong>).
In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'custom'</span></span></code> change events, the baseline K is multiplied by a series
of particular factors (defined by <strong>sizes</strong>) at a series of particular
timesteps (defined by <strong>timesteps</strong>).</p>
<p><strong>start_t</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#starting timestep</span>
<span class="s1">&#39;start_t&#39;</span><span class="p">:</span>            <span class="mi">50</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 50</p>
<p>reset? P</p>
<p>This indicates the timestep at which the demographic change event
should start.</p>
<p><strong>end_t</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#ending timestep</span>
<span class="s1">&#39;end_t&#39;</span><span class="p">:</span>            <span class="mi">100</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 100</p>
<p>reset? P</p>
<p>This indicates the last timestep of the change event.</p>
<p><strong>rate</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#rate, for monotonic change</span>
<span class="s1">&#39;rate&#39;</span><span class="p">:</span>             <span class="mf">1.02</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1.02</p>
<p>reset? P</p>
<p>This indicates the rate at which a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'monotonic'</span></span></code> change event should occur.
At each timestep during the event, a new carrying capacity raster (K)
will be calculated by multiplying the previous step’s K by this factor.
Thus, values should be expressed relative to 1.0 indicating no change.</p>
<p><strong>interval</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#interval of changes, for stochastic change</span>
<span class="s1">&#39;interval&#39;</span><span class="p">:</span>         <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 1</p>
<p>reset? P</p>
<p>This indicates the interval at which fluctutations should occur during a
<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'stochastic'</span></span></code> change event (i.e. the number of timesteps to wait
between fluctuations).</p>
<p><strong>distr</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#distr, for stochastic change {&#39;uniform&#39;, &#39;normal&#39;}</span>
<span class="s1">&#39;distr&#39;</span><span class="p">:</span>            <span class="s1">&#39;uniform&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'uniform'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'normal'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'uniform'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the distribution from which to draw the sizes of
fluctuations in a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'stochastic'</span></span></code> change event. Valid options are
<cite>‘uniform’</cite> and <cite>‘normal’</cite>.</p>
<p><strong>n_cycles</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num cycles, for cyclical change</span>
<span class="s1">&#39;n_cycles&#39;</span><span class="p">:</span>         <span class="mi">10</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 10</p>
<p>reset? P</p>
<p>This indicates the number of cyclical fluctuations that should occur during
a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cyclical'</span></span></code> change event.</p>
<p><strong>size_range</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#min &amp; max sizes, for stochastic &amp; cyclical change</span>
<span class="s1">&#39;size_range&#39;</span><span class="p">:</span>       <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="literal number float"><span class="pre">0.5</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal number float"><span class="pre">1.5</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset? P</p>
<p>This defines the minimum and maximum sizes of fluctuations that can occur
during <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'stochastic'</span></span></code> and <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'cyclical'</span></span></code> change events.</p>
<p><strong>timesteps</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of timesteps, for custom change</span>
<span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">90</span><span class="p">,</span> <span class="mi">95</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: [50, 90, 95]</p>
<p>reset? P</p>
<p>This defines the series of particular timesteps at which fluctutations should
occur during a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'custom'</span></span></code> change event.</p>
<p><strong>sizes</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of sizes, for custom change</span>
<span class="s1">&#39;sizes&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: [2, 5, 0.5]</p>
<p>reset? P</p>
<p>This defines the series of particular fluctutations that should occur
during a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'custom'</span></span></code> change event.</p>
</section>
<hr class="docutils" />
<section id="life-history-change">
<h5>Life-history change<a class="headerlink" href="#life-history-change" title="Permalink to this headline"></a></h5>
<p><strong>&lt;life_hist_param&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#life-history parameter to change</span>
<span class="s1">&#39;&lt;life_hist_param&gt;&#39;</span><span class="p">:</span> <span class="p">{</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'&lt;life_hist_param&gt;'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the life-history parameter to be changed by this life-history
change event. (Note that unlike most parameters, this parameter is
a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key, the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters controlling how the life-history parameter that is named
will change.)</p>
<p><strong>timesteps</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of timesteps</span>
<span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the timesteps at which the life-history parameter being changed
should change (to the values indicated by <strong>vals</strong>).</p>
<p><strong>vals</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of values</span>
<span class="s1">&#39;vals&#39;</span><span class="p">:</span>        <span class="p">[],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the values to which the life-history parameter being changed
should change (at the timesteps indicated by <strong>timesteps</strong>).</p>
</section>
</section>
</section>
</section>
<hr class="docutils" />
<section id="other-parameters">
<h2>Other parameters<a class="headerlink" href="#other-parameters" title="Permalink to this headline"></a></h2>
<section id="id2">
<h3>Main<a class="headerlink" href="#id2" title="Permalink to this headline"></a></h3>
<p><strong>T</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#total Model runtime (in timesteps)</span>
<span class="s1">&#39;T&#39;</span><span class="p">:</span>            <span class="mi">100</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 100</p>
<p>reset? Y</p>
<p>This indicates the total number of timesteps for which the main portion of
a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> (i.e. the portion after the burn-in has completed) will be run
during each iteration.</p>
<p><strong>burn_T</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#min burn-in runtime (in timesteps)</span>
<span class="s1">&#39;burn_T&#39;</span><span class="p">:</span>       <span class="mi">30</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 30</p>
<p>reset? P</p>
<p>This indicates the minimum number of timesteps for which a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>’s
burn-in will run. (Note this is only a minimum because the test for
burn-in completion includes a check that at least this many timesteps have
elapsed, but also includes two statistical checks of stationarity of the
size of each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code>.)</p>
<p><strong>num</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#seed number</span>
<span class="s1">&#39;num&#39;</span><span class="p">:</span>          <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates whether or not to set the seeds of the random number
generators (by calling <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">random</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">seed</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">random</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">seed</span></span></code>)
before building and running a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>. If value is an integer, the seeds
will be set to that value. If value is <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, seeds will not be set.</p>
</section>
<hr class="docutils" />
<section id="iterations">
<h3>Iterations<a class="headerlink" href="#iterations" title="Permalink to this headline"></a></h3>
<p><strong>num_iterations</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num iterations</span>
<span class="s1">&#39;n_its&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 2</p>
<p>reset? Y</p>
<p>This indicates the number of iterations for which the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>
should be run. (Note that for each iteration a separate subdirectory of
data and stats will be written, if your <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> has parameterized data
and stats to be collected.)</p>
<p><strong>rand_landscape</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to randomize Landscape each iteration</span>
<span class="s1">&#39;rand_landscape&#39;</span><span class="p">:</span>    <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> should be randomized for each
iteration. If True, a new <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> will be generated at the start
of each iteration. If False, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> from iteration 0 will be
saved and reused for each subsequent iteration.</p>
<p><strong>rand_comm</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to randomize Community each iteration</span>
<span class="s1">&#39;rand_comm&#39;</span><span class="p">:</span>    <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> should be randomized for each
iteration. If True, a new <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> will be generated at the start
of each iteration. If False, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> from iteration 0 will be
saved and reused for each subsequent iteration (and whether that
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> is saved before or after being burned in will depend on
the value provided to the <strong>repeat_burn</strong> parameter).</p>
<p><strong>rand_genarch</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to randomize GenomicArchitectures each iteration</span>
<span class="s1">&#39;rand_genarch&#39;</span><span class="p">:</span>    <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? N</p>
<p>This indicates whether all <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> with genomes should have their
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitectures</span></span></code> randomized (and then genomes redrawn) for each
iteration.
Defaults to True because this, in combination with the other default settings
(<strong>rand_landscape</strong> False, <strong>rand_comm</strong> False, <strong>repeat_burn</strong> False) creates
a reasonable default behavior for the set of iterations in a run:
Each iteration will use the same landscape and the same number and spatial
distribution of starting individuals, thus obviating the need for the most
computationally costly components of a Geonomics model prior to the main phase,
but those individuals’ <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitectures</span></span></code> and genomes
will be drawn at random, providing independent instantiations
of the landscape genomic scenario being simulated.
Note, however, that this parameter is inconsequential for any <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
with a custom genomic architecture file; if a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ <strong>gen_arch_file</strong>
parameter is not None then the custom genomic architecture file will be used
to set the genomic architecture for that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> on each iteration,
regardless of whether <strong>rand_genarch</strong> is True or False.</p>
<p><strong>repeat_burn</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to burn in each iteration</span>
<span class="s1">&#39;repeat_burn&#39;</span><span class="p">:</span>  <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether a reused <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> should be burned in
separately for each iteration for which it is reused. If True, the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> from iteration 0 will be saved as soon as its instantiated,
but will have a new burn-in run for each iteration in which it is used. If
False, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> from iteration 0 will be saved after its burn-in
is complete, and then will only have the main portion of its <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> run
separately during each iteration. (Note that if <strong>rand_community</strong> is set to True then
the value of this parameter will not be used.)</p>
<hr class="docutils" />
<section id="data">
<h4>Data<a class="headerlink" href="#data" title="Permalink to this headline"></a></h4>
<section id="sampling">
<h5>Sampling<a class="headerlink" href="#sampling" title="Permalink to this headline"></a></h5>
<p><strong>scheme</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#sampling scheme {&#39;all&#39;, &#39;random&#39;, &#39;point&#39;, &#39;transect&#39;}</span>
<span class="s1">&#39;scheme&#39;</span><span class="p">:</span>               <span class="s1">&#39;random&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'all'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'random'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'random'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the sampling scheme to use when collecting data from a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>. Currently valid values include <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'all'</span></span></code>,
<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'random'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code>, and <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>.</p>
<p>With <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'all'</span></span></code>, data will be collected for all <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s
at each sampling timestep. With <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'random'</span></span></code>, data will be collected from a
random sample of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s (of size indicated by parameter <strong>n</strong>)
from anywhere on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>.
With <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code>, data will be collected from random samples of size <strong>n</strong>
within a certain distance (<strong>radius</strong>) of each of a set of particular
points (<strong>points</strong>). With <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>, a linear transect of some
number of points (<strong>n_transect_points</strong>) between some endpoints
(<strong>transect_endpoints</strong>) will be created, and then data will be collected
from random samples of size <strong>n</strong> with a certain distance (<strong>radius</strong>)
of each point along the transect.</p>
<p><strong>n</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#sample size at each point, for point &amp; transect sampling</span>
<span class="s1">&#39;n&#39;</span><span class="p">:</span>                    <span class="mi">250</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 250</p>
<p>reset? P</p>
<p>This indicates the total number of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s to sample each time
data is collected (if <strong>scheme</strong> is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'random'</span></span></code>), or the number of
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s to sample around each one of a set of points (if <strong>scheme</strong>
is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code> or <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>). This parameter will only be used if
<strong>scheme</strong> is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'random'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code>, or <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>; otherwise
it may be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.</p>
<p><strong>points</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#coords of collection points, for point sampling</span>
<span class="s1">&#39;points&#39;</span><span class="p">:</span>               <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code> of 2-<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code>s, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the points around which to sample <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s for data
collection. This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code>;
otherwise it may be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.</p>
<p><strong>transect_endpoints</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#coords of transect endpoints, for transect sampling</span>
<span class="s1">&#39;transect_endpoints&#39;</span><span class="p">:</span>   <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{2-<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code> of 2-<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code>s, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the endpoints between which to create a transect, along which
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s will be sampled for data collection.
This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>;
otherwise it may be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.</p>
<p><strong>n_transect_points</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num points along transect, for transect sampling</span>
<span class="s1">&#39;n_transect_points&#39;</span><span class="p">:</span>    <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the number of points to create on the transect along which
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s will be sampled for data collection.
This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>;
otherwise it may be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.</p>
<p><strong>radius</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#collection radius around points, for point &amp; transect sampling</span>
<span class="s1">&#39;radius&#39;</span><span class="p">:</span>               <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the radius around sampling points within which
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s may be sampled for data collection.
This parameter will only be used if <strong>scheme</strong> is <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'point'</span></span></code> or
<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'transect'</span></span></code>; otherwise it may be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.</p>
<p><strong>when</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#when to collect data</span>
<span class="s1">&#39;when&#39;</span><span class="p">:</span>                 <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the timesteps during main <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> iterations
at which data should be collected (in addition to after the final timestep
of each iteration, when data is always collected for any <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> for which
data collection is parameterized). If value is a non-zero <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>,
it will be treated as a frequency at which data should be collected (e.g.
a value of 5 will cause data to be collected every 5 timesteps). If value
is a list of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s, they will be treated as the particular timesteps
at which data should be collected. If value is 0 or <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>,
data will be collected only after the final timestep.</p>
<p><strong>include_landscape</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to save current Layers when data is collected</span>
<span class="s1">&#39;include_landscape&#39;</span><span class="p">:</span>         <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether to include the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s among the
data that is collected. If True, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> will be written to a raster
or array file (according to the format indicated by
<strong>geo_rast_format</strong>) each time data is collected.</p>
<p><strong>include_fixed_sites</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to include fixed loci in VCF files</span>
<span class="s1">&#39;include_fixed_sites&#39;</span><span class="p">:</span>  <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether fixed sites (i.e. loci which are fixed for either the
0 or 1 allele) should be included in any VCF files that are written. Thus,
this parameter is only relevant if <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'vcf'</span></span></code> is one of the genetic data
formats indicated by <strong>gen_format</strong>.</p>
</section>
<hr class="docutils" />
<section id="format">
<h5>Format<a class="headerlink" href="#format" title="Permalink to this headline"></a></h5>
<p><strong>gen_format</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#format for genetic data {&#39;vcf&#39;, &#39;fasta&#39;}</span>
<span class="s1">&#39;gen_format&#39;</span><span class="p">:</span>           <span class="p">[</span><span class="s1">&#39;vcf&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">],</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'vcf'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'fasta'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[</span></span><span class="literal string single"><span class="pre">'vcf'</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal string single"><span class="pre">'fasta'</span></span><span class="punctuation"><span class="pre">]</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[</span></span><span class="literal string single"><span class="pre">'vcf'</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal string single"><span class="pre">'fasta'</span></span><span class="punctuation"><span class="pre">]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format or formats to use for writing genetic data.
data. Currently valid formats include <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'vcf'</span></span></code> and <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'fasta'</span></span></code> formats.
Either or both formats may be specified; all formats that are specified will
be written each time data is collected.</p>
<p><strong>geo_vect_format</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#format for vector geodata {&#39;csv&#39;, &#39;shapefile&#39;, &#39;geojson&#39;}</span>
<span class="s1">&#39;geo_vect_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;csv&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'csv'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'shapefile'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'geojson'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'csv'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format to use for writing geographic vector data (i.e.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’ point locations).
Currently valid formats include <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'csv'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'shapefile'</span></span></code>,
and <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'geojson'</span></span></code>. Any one format may be specified.</p>
<p><strong>geo_rast_format</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#format for raster geodata {&#39;geotiff&#39;, &#39;txt&#39;}</span>
<span class="s1">&#39;geo_rast_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;geotiff&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'geotiff'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'txt'</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'geotiff'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format to use for writing geographic raster data (i.e.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> arrays). Currently valid formats include <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'geotiff'</span></span></code>
and <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'txt'</span></span></code>. Either format may be specified. Note that this parameter
will only be used if the <strong>include_landscape</strong> parameter is set to True.</p>
<p><strong>nonneut_loc_format</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#format for files containing non-neutral loci</span>
<span class="s1">&#39;nonneut_loc_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;csv&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'csv'</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'csv'</span></span></code></p>
<p>reset? P</p>
<p>This indicates the format to use for writing files
containing lists (in columns) of the non-neutral loci subtending
each of the traits (whose names are used as column names).
Currently valid values include <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">'csv'</span></span></code>,
which saves a simple CSV file, and <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>,
which tells Geonomics not to save these files as output.</p>
</section>
</section>
<hr class="docutils" />
<section id="stats">
<h4>Stats<a class="headerlink" href="#stats" title="Permalink to this headline"></a></h4>
<p>The stats parameters section has subsection for each statistc that Geonomics
can calculate. (Currently valid statistics include:
- <em>‘Nt’</em>: number of individuals at timestep t
- <em>‘het’</em>: heterozygosity
- <em>‘maf’</em>: minor allele frequency
- <em>‘mean_fit’</em>: mean fitness of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
- <em>‘ld’</em>: linkage disequilibrium</p>
<p>There are only a few parameters, which are shared across all
of those subsections, and each parameter always means the same thing despite
which statistic it is parameterizing. Thus, hereafter we provide a single of
each of those parameters are how it works, regardless of the statistic for
which it used:</p>
<p><strong>calc</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to calculate</span>
<span class="s1">&#39;calc&#39;</span><span class="p">:</span>     <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: (varies by statistic)</p>
<p>reset? P</p>
<p>This indicates whether or not a given statistic should be calculated. Thus,
only those statistics whose <strong>calc</strong> parameters are set to True will be
calculated and saved when their <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is run.</p>
<p><strong>freq</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#calculation frequency (in timesteps)</span>
<span class="s1">&#39;freq&#39;</span><span class="p">:</span>     <span class="mi">5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: (varies by statistic)</p>
<p>reset? P</p>
<p>This indicates the frequency with which a given statistic should be calculated
during each iteration (in timesteps). If set to 0, Geonomics will calculate
and save this statistic for only the first and last timesteps
of each iteration.</p>
<p><strong>mean</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to mean across sampled individs</span>
<span class="s1">&#39;mean&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: (varies by statistic, and only valid for certain statistics)</p>
<p>reset? P</p>
<p>For some statistics that produce a vector of values each timestep
when they are collected (containing one value per <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>),
such as heterozygosity, this indicates
whether those values should instead be meaned and saved as a
single value for each timestep.</p>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="viz.html" class="btn btn-neutral float-left" title="Visualization" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="modules.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Drew Hart.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>